{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load library and example datasets\n",
    "library(\"pacman\")\n",
    "\n",
    "pacman::p_load(\"pathview\", \"gage\", \"tidyverse\", \"MetaboAnalystR\", \"KEGGREST\")\n",
    "data(gse16873.d)\n",
    "# Load human pathways data\n",
    "data(paths.hsa)\n",
    "# load demo pathway-related data, including 3 pathway ids and related plotting params\n",
    "# this is in dictionary format\n",
    "data(demo.paths)\n",
    "\n",
    "source(\"~/maca-utils/maca-kegg-utils.R\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get DE metabs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"MetaboAnalyst R objects initialized ...\"\n",
      " [1] \"Successfully passed sanity check!\"                                                                    \n",
      " [2] \"Samples are not paired.\"                                                                              \n",
      " [3] \"2 groups were detected in samples.\"                                                                   \n",
      " [4] \"Only English letters, numbers, underscore, hyphen and forward slash (/) are allowed.\"                 \n",
      " [5] \"<font color=\\\"orange\\\">Other special characters or punctuations (if any) will be stripped off.</font>\"\n",
      " [6] \"All data values are numeric.\"                                                                         \n",
      " [7] \"A total of 2 (0.1%) missing values were detected.\"                                                    \n",
      " [8] \"<u>By default, these values will be replaced by a small value.</u>\"                                   \n",
      " [9] \"Click <b>Skip</b> button if you accept the default practice\"                                          \n",
      "[10] \"Or click <b>Missing value imputation</b> to use other methods\"                                        \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Duplicated column names deduplicated: 'Alanylglycine' => 'Alanylglycine_1' [59]”"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"MetaboAnalyst R objects initialized ...\"\n",
      "[1] \"Loaded files from MetaboAnalyst web-server.\"\n",
      "[1] \"Loaded files from MetaboAnalyst web-server.\"\n",
      "[1] \"Loaded files from MetaboAnalyst web-server.\"\n",
      "NULL\n"
     ]
    }
   ],
   "source": [
    "fn_auc_csv <- \"test_dme_data.csv\"\n",
    "tbl0 <- call_maca_normalization(fn_auc_csv)\n",
    "tbl0 <- get_de_metabs(tbl0, 0.01, \"treatment\", \"control\", 0.05)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get enriched pathways from MACA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## x <- call_maca_pw_analysis(\"test_dme_data.csv\", \"dme\")\n",
    "tbl1 <- x[[1]]\n",
    "pw_dict <- x[[2]]\n",
    "pw_names <- get_kegg_pw_ref_tbl(\"dme\")\n",
    "\n",
    "# Manually replace some pw_names\n",
    "tbl1$pw_name[tbl1$pw_name==\"Fatty acid elongation in mitochondria\"] <- \"Fatty acid elongation\"\n",
    "tbl1$pw_name[tbl1$pw_name==\"Glycolysis or Gluconeogenesis\"] <- \"Glycolysis / Gluconeogenesis\"\n",
    "\n",
    "# Join pw enrichment output with pw IDs\n",
    "tbl1 <- inner_join(tbl1, pw_names, by=\"pw_name\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pw_name_ls0 <- as.vector(unlist(tbl0[\"pw_name\"]))\n",
    "pw_name_ls1 <- as.vector(unlist(tbl1[\"pw_name\"]))\n",
    "\n",
    "for (pw in pw_name_ls0) {\n",
    "    if (pw %in% pw_name_ls1 == F) {\n",
    "        print(pw)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Call PathviewR\n",
    "\n",
    "* Use `pathviewR` to retrieve the sanitized KEGG graphs (which it claims to be able to)\n",
    "* Also do data mapping, with a single named list of logFC values (names are KEGG Ids). But these will only result in coloured nodes; barcharts are better. \n",
    "* Will return a lot of `pngs` and `xmls`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get input compound data\n",
    "# named list of logFCs, names are KEGG Id\n",
    "tbl_tmp <- tbl0 %>% dplyr::select(c(\"KEGG\", \"log2_fc\")) %>% filter(KEGG != \"undef\")\n",
    "cpd_data_ls <- unlist(tbl_tmp$log2_fc)\n",
    "names(cpd_data_ls) <- unlist(tbl_tmp$KEGG)\n",
    "\n",
    "# Get list of pathway numbers\n",
    "pw_num_ls <- lapply(unlist(tbl1$pw_id), function(x) {gsub(\"dme\", \"\", x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Info: Downloading xml files for dme00970, 1/1 pathways..\n",
      "Info: Downloading png files for dme00970, 1/1 pathways..\n",
      "Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Warning message in structure(x$children, class = \"XMLNodeList\"):\n",
      "“Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.\n",
      "  Consider 'structure(list(), *)' instead.”Info: Working in directory /Users/don/Documents/my_vignettes\n",
      "Info: Writing image file dme00970.dme00970.png\n"
     ]
    }
   ],
   "source": [
    "i <- 1\n",
    "suffix_i <- paste0(\"dme\", pw_num_ls[i])\n",
    "pathview(cpd.data = cpd_data_ls,\n",
    "         pathway.id = pw_num_ls[i], \n",
    "         species = \"dme\", \n",
    "         out.suffix = suffix_i,\n",
    "         keys.align = \"y\", \n",
    "         kegg.native = T\n",
    "         )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for pathview {pathview}\"><tr><td>pathview {pathview}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>\n",
       "Pathway based data integration and visualization\n",
       "</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Pathview is a tool set for pathway based data integration and\n",
       "visualization. It maps and renders user data on relevant pathway\n",
       "graphs. All users need is to supply their gene or compound data and\n",
       "specify the target pathway. Pathview automatically downloads the pathway\n",
       "graph data, parses the data file, maps user data to the pathway, and\n",
       "render pathway graph with the mapped data.\n",
       "Pathview generates both native KEGG view and Graphviz views for\n",
       "pathways. keggview.native and keggview.graph are the two viewer\n",
       "functions, and pathview is the main function providing a unified\n",
       "interface to downloader, parser, mapper and viewer functions.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "pathview(gene.data = NULL, cpd.data = NULL, pathway.id,\n",
       "species = \"hsa\", kegg.dir = \".\", cpd.idtype = \"kegg\", gene.idtype =\n",
       "\"entrez\", gene.annotpkg = NULL, min.nnodes = 3, kegg.native = TRUE,\n",
       "map.null = TRUE, expand.node = FALSE, split.group = FALSE, map.symbol =\n",
       "TRUE, map.cpdname = TRUE, node.sum = \"sum\", discrete=list(gene=FALSE,\n",
       "cpd=FALSE), limit = list(gene = 1, cpd = 1), bins = list(gene = 10, cpd\n",
       "= 10), both.dirs = list(gene = T, cpd = T), trans.fun = list(gene =\n",
       "NULL, cpd = NULL), low = list(gene = \"green\", cpd = \"blue\"), mid =\n",
       "list(gene = \"gray\", cpd = \"gray\"), high = list(gene = \"red\", cpd =\n",
       "\"yellow\"), na.col = \"transparent\", ...)\n",
       "\n",
       "keggview.native(plot.data.gene = NULL, plot.data.cpd = NULL,\n",
       "cols.ts.gene = NULL, cols.ts.cpd = NULL, node.data, pathway.name,\n",
       "out.suffix = \"pathview\", kegg.dir = \".\", multi.state=TRUE, match.data =\n",
       "TRUE, same.layer = TRUE, res = 300, cex = 0.25, discrete =\n",
       "list(gene=FALSE, cpd=FALSE), limit= list(gene = 1, cpd = 1), bins =\n",
       "list(gene = 10, cpd = 10), both.dirs =list(gene = T, cpd = T), low =\n",
       "list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd =\n",
       "\"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "\"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "\"x\", key.pos = \"topright\", ...)\n",
       "\n",
       "keggview.graph(plot.data.gene = NULL, plot.data.cpd = NULL, cols.ts.gene\n",
       "= NULL, cols.ts.cpd = NULL, node.data, path.graph, pathway.name,\n",
       "out.suffix = \"pathview\", pdf.size = c(7, 7), multi.state=TRUE,\n",
       "same.layer = TRUE, match.data = TRUE, rankdir = c(\"LR\", \"TB\")[1],\n",
       "is.signal = TRUE, split.group = F, afactor = 1, text.width = 15, cex =\n",
       "0.5, map.cpdname = FALSE, cpd.lab.offset = 1.0,\n",
       "discrete=list(gene=FALSE, cpd=FALSE), limit = list(gene = 1, cpd = 1),\n",
       "bins = list(gene = 10, cpd = 10), both.dirs = list(gene = T, cpd = T),\n",
       "low = list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd\n",
       "= \"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "\"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "\"x\", key.pos = \"topright\", sign.pos = \"bottomright\", ...)  </pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>gene.data</code></td>\n",
       "<td>\n",
       "\n",
       "<p>either vector (single sample) or a matrix-like data (multiple\n",
       "sample). Vector should be numeric with gene IDs as names or it may\n",
       "also be character of gene IDs. Character vector is treated as discrete\n",
       "or count data. Matrix-like data structure has genes as rows and\n",
       "samples as columns. Row names should be gene IDs. Here gene ID is a\n",
       "generic concepts, including multiple types of gene, transcript and\n",
       "protein uniquely mappable to KEGG gene IDs. KEGG ortholog IDs are also\n",
       "treated as gene IDs as to handle metagenomic data. Check details for\n",
       "mappable ID types. Default gene.data=NULL.\n",
       "</p>\n",
       "<p>numeric, character, continuous\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cpd.data</code></td>\n",
       "<td>\n",
       "\n",
       "<p>the same as gene.data, excpet named with IDs mappable to KEGG\n",
       "compound IDs. Over 20 types of IDs included in CHEMBL database can be\n",
       "used here. Check details for mappable ID types. Default\n",
       "cpd.data=NULL. Note that gene.data and cpd.data can't be NULL\n",
       "simultaneously.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>pathway.id</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character vector, the KEGG pathway ID(s), usually 5 digit, may also include the 3\n",
       "letter KEGG species code.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>species</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, either the kegg code, scientific name or the common name of\n",
       "the target species. This applies to both pathway and gene.data or\n",
       "cpd.data. When KEGG ortholog pathway is considered,\n",
       "species=&quot;ko&quot;. Default species=&quot;hsa&quot;, it is equivalent to use either\n",
       "&quot;Homo sapiens&quot; (scientific name) or &quot;human&quot; (common name). \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>kegg.dir</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, the directory of KEGG pathway data file (.xml) and image file\n",
       "(.png). Users may supply their own data files in the same format and\n",
       "naming convention of KEGG's (species code + pathway id,\n",
       "e.g. hsa04110.xml, hsa04110.png etc) in this directory. Default\n",
       "kegg.dir=&quot;.&quot; (current working directory). \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cpd.idtype</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, ID type used for the cpd.data. Default cpd.idtype=&quot;kegg&quot; (include\n",
       "compound, glycan and drug accessions). \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>gene.idtype</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, ID type used for the gene.data, case insensitive. Default\n",
       "gene.idtype=&quot;entrez&quot;, i.e. Entrez Gene, which are the primary KEGG gene ID\n",
       "for many common model organisms. For other species, gene.idtype should\n",
       "be set to &quot;KEGG&quot; as KEGG use other types of gene IDs. For the common\n",
       "model organisms (to check the list, do: <code>data(bods); bods</code>), you\n",
       "may also specify other types of valid IDs. To check the ID list, do:\n",
       "<code>data(gene.idtype.list); gene.idtype.list</code>. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>gene.annotpkg</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, the name of the annotation package to use for mapping between\n",
       "other gene ID types including symbols and Entrez gene ID. Default gene.annotpkg=NULL.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>min.nnodes</code></td>\n",
       "<td>\n",
       "\n",
       "<p>integer, minimal number of nodes of type &quot;gene&quot;,&quot;enzyme&quot;, &quot;compound&quot; or\n",
       "&quot;ortholog&quot; for a pathway to be considered. Default min.nnodes=3.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>kegg.native</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether to render pathway graph as native KEGG graph (.png) or\n",
       "using graphviz layout engine (.pdf). Default kegg.native=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>map.null</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether to map the NULL gene.data or cpd.data to pathway. When\n",
       "NULL data are mapped, the gene or compound nodes in the pathway will be\n",
       "rendered as actually mapped nodes, except with NA-valued color. When\n",
       "NULL data are not mapped, the nodes are rendered as unmapped nodes. This\n",
       "argument mainly affects native KEGG graph view, i.e. when\n",
       "kegg.native=TRUE. Default map.null=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>expand.node</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether the multiple-gene nodes are expanded into single-gene\n",
       "nodes. Each expanded single-gene nodes inherits all edges from the\n",
       "original multiple-gene node. This option only affects graphviz graph view, i.e. when\n",
       "kegg.native=FALSE. This option is not effective for most metabolic\n",
       "pathways where it conflits with  converting reactions to edges. Default expand.node=FLASE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>split.group</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether split node groups are split to individual nodes. Each\n",
       "split member nodes inherits all edges from the node group. This\n",
       "option only affects graphviz graph view, i.e. when\n",
       "kegg.native=FALSE. This option also effects most metabolic\n",
       "pathways even without group nodes defined orginally. For these pathways,\n",
       "genes involved in the same reaction are grouped automatically when\n",
       "converting reactions to edges unless split.group=TRUE. d\n",
       "split.group=FLASE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>map.symbol</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether map gene IDs to symbols for gene node labels or use the\n",
       "graphic name from the KGML file. This option is only effective for\n",
       "kegg.native=FALSE or same.layer=FALSE when kegg.native=TRUE. For\n",
       "same.layer=TRUE when kegg.native=TRUE, the native KEGG labels will be\n",
       "kept. Default map.symbol=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>map.cpdname</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether map compound IDs to formal names for compound node labels or use the\n",
       "graphic name from the KGML file (KEGG compound accessions). This option is only effective for\n",
       "kegg.native=FALSE. When kegg.native=TRUE, the native KEGG labels will be\n",
       "kept. Default map.cpdname=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>node.sum</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, the method name to calculate node summary given that\n",
       "multiple genes or compounds are mapped to it. Poential options include\n",
       "&quot;sum&quot;,&quot;mean&quot;, &quot;median&quot;, &quot;max&quot;, &quot;max.abs&quot; and &quot;random&quot;. Default node.sum=&quot;sum&quot;.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>discrete</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a list of two logical elements with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument tells whether gene.data or cpd.data should be treated as\n",
       "discrete. Default dsicrete=list(gene=FALSE, cpd=FALSE), i.e. both data should\n",
       "be treated as continuous.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>limit</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a list of two numeric elements with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument specifies the limit values for gene.data and cpd.data when converting\n",
       "them to pseudo colors. Each element of the list could be of length 1 or\n",
       "2. Length 1 suggests discrete data or 1 directional (positive-valued)\n",
       "data, or the absolute limit for 2 directional data. Length 2 suggests 2\n",
       "directional data. Default limit=list(gene=1, cpd=1).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>bins</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a list of two integer elements with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument specifies the number of levels or bins for gene.data and cpd.data when converting\n",
       "them to pseudo colors. Default limit=list(gene=10, cpd=10).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>both.dirs</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a list of two logical  elements with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument specifies whether gene.data and cpd.data are 1 directional or 2\n",
       "directional data when converting them to pseudo colors. Default limit=list(gene=TRUE, cpd=TRUE).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>trans.fun</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a list of two function (not character) elements with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument specifies whether and how gene.data and cpd.data are\n",
       "transformed. Examples are <code>log</code>, <code>abs</code> or users' own\n",
       "functions. Default limit=list(gene=NULL, cpd=NULL).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>low, mid, high</code></td>\n",
       "<td>\n",
       "\n",
       "<p>each is a list of two colors with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument specifies the color spectra to code gene.data and\n",
       "cpd.data. When data are 1 directional (TRUE value in both.dirs), only\n",
       "mid and high are used to specify the color spectra. Default spectra (low-mid-high)\n",
       "&quot;green&quot;-&quot;gray&quot;-&quot;red&quot; and &quot;blue&quot;-&quot;gray&quot;-&quot;yellow&quot; are used for gene.data\n",
       "and cpd.data respectively. \n",
       "The values for 'low, mid, high' can be given as color names\n",
       "('red'), plot color index (2=red), and HTML-style RGB,\n",
       "(&quot;\\#FF0000&quot;=red).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>na.col</code></td>\n",
       "<td>\n",
       "\n",
       "<p>color used for NA's or missing values in gene.data and cpd.data. d\n",
       "na.col=&quot;transparent&quot;. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>...</code></td>\n",
       "<td>\n",
       "\n",
       "<p>extra arguments passed to keggview.native or keggview.graph function.\n",
       "</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "<p>special arguments for keggview.native or keggview.graph function.\n",
       "</p>\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>plot.data.gene</code></td>\n",
       "<td>\n",
       "\n",
       "<p>data.frame returned by node.map function for rendering mapped gene\n",
       "nodes, including node name, type, positions (x, y), sizes (width,\n",
       "height), and mapped gene.data. This data is also used as input for\n",
       "pseduo-color coding through node.color function. Default plot.data.gene=NULL.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>plot.data.cpd</code></td>\n",
       "<td>\n",
       "\n",
       "<p>same as plot.data.gene function, except for mapped compound node data. d\n",
       "plot.data.cpd=NULL. Default plot.data.cpd=NULL. Note that plot.data.gene and\n",
       "plot.data.cpd can't be NULL simultaneously.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cols.ts.gene</code></td>\n",
       "<td>\n",
       "\n",
       "<p>vector or matrix of colors returned by node.color function for rendering\n",
       "gene.data. Dimensionality is the same as the latter. Default cols.ts.gene=NULL.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cols.ts.cpd</code></td>\n",
       "<td>\n",
       "\n",
       "<p>same as cols.ts.gene, except corresponding to cpd.data. d\n",
       "cols.ts.cpd=NULL. Note that cols.ts.gene and cols.ts.cpd\n",
       "plot.data.gene can't be NULL simultaneously. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>node.data</code></td>\n",
       "<td>\n",
       "\n",
       "<p>list returned by node.info function, which parse KGML file directly or\n",
       "indirectly, and extract the node data.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>pathway.name</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, the full KEGG pathway name in the format of 3-letter species\n",
       "code with 5-digit pathway id, eg &quot;hsa04612&quot;. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>out.suffix</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, the suffix to be added after the pathway name as part of the\n",
       "output graph file. Sample names or column names of the gene.data or\n",
       "cpd.data are also added when there are multiple samples. Default out.suffix=&quot;pathview&quot;.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>multi.state</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether multiple states (samples or columns) gene.data or\n",
       "cpd.data should be integrated and plotted in the same graph. Default\n",
       "match.data=TRUE. In other words, gene or compound nodes will be sliced\n",
       "into multiple pieces corresponding to the number of states in the\n",
       "data.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>match.data</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether the samples of gene.data and cpd.data are\n",
       "paired. Default match.data=TRUE. When let sample sizes of gene.data\n",
       "and cpd.data be m and n, when m&gt;n, extra columns of NA's (mapped to no\n",
       "color) will be added to cpd.data as to make the sample size the\n",
       "same. This will result in the smae number of slice in gene nodes and\n",
       "compound when multi.state=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>same.layer</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, control plotting layers:\n",
       "1) if node colors be plotted in the same layer as the pathway\n",
       "graph when kegg.native=TRUE, 2) if  edge/node type legend be plotted in\n",
       "the same page when kegg.native=FALSE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>res</code></td>\n",
       "<td>\n",
       "\n",
       "<p>The nominal resolution in ppi which will be recorded in the\n",
       "bitmap file, if a positive integer.  Also used for 'units'\n",
       "other than the default, and to convert points to pixels. This\n",
       "argument is only effective when kegg.native=TRUE. Default res=300.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cex</code></td>\n",
       "<td>\n",
       "\n",
       "<p>A numerical value giving the amount by which plotting text\n",
       "and symbols should be scaled relative to the default 1. Default cex=0.25\n",
       "when kegg.native=TRUE, cex=0.5 when kegg.native=FALSE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>new.signature</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether pathview signature is added to the pathway\n",
       "graphs. Default new.signature=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>plot.col.key</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether color key is added to the pathway\n",
       "graphs. Default plot.col.key= TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>key.align</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, controlling how the color keys are aligned when both\n",
       "gene.data and cpd.data are not NULL. Potential values are &quot;x&quot;, aligned\n",
       "by x coordinates, and &quot;y&quot;, aligned by y coordinates. Default key.align=&quot;x&quot;.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>key.pos</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, controlling the position of color key(s). Potentail values\n",
       "are &quot;bottomleft&quot;, &quot;bottomright&quot;, &quot;topleft&quot; and &quot;topright&quot;. d\n",
       "key.pos=&quot;topright&quot;. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>sign.pos</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, controlling the position of pathview signature. Only\n",
       "effective when kegg.native=FALSE, Signature position is fixed in place\n",
       "of the original KEGG signature when kegg.native=TRUE. Potentail values\n",
       "are &quot;bottomleft&quot;, &quot;bottomright&quot;, &quot;topleft&quot; and &quot;topright&quot;. d\n",
       "sign.pos=&quot;bottomright&quot;. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>path.graph</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a graph object parsed from KGML file, only effective when\n",
       "kegg.native=FALSE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>pdf.size</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a numeric vector of length 2, giving the width and height of the pathway\n",
       "graph pdf file. Note that pdf width increase by half when\n",
       "same.layer=TRUE to accommodate legends. Only effective when\n",
       "kegg.native=FALSE. Default pdf.size=c(7,7).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>rankdir</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, either &quot;LR&quot; (left to right) or &quot;TB&quot; (top to bottom),\n",
       "specifying the pathway graph layout direction. Only effective when\n",
       "kegg.native=FALSE. Default rank.dir=&quot;LR&quot;.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>is.signal</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, if the pathway is treated as a signaling pathway, where all the\n",
       "unconnected nodes are dropped. This argument also affect the graph\n",
       "layout type, i.e. &quot;dot&quot; for signals or &quot;neato&quot; otherwise. Only effective when\n",
       "kegg.native=FALSE. Default is.signal=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>afactor</code></td>\n",
       "<td>\n",
       "\n",
       "<p>numeric, node amplifying factor. This argument is for node size\n",
       "fine-tuning, its effect is subtler than expected. Only effective when\n",
       "kegg.native=FALSE. Default afctor=1.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>text.width</code></td>\n",
       "<td>\n",
       "\n",
       "<p>numeric, specifying the line width for text wrap. Only effective when\n",
       "kegg.native= FALSE. Default text.width=15 (characters).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cpd.lab.offset</code></td>\n",
       "<td>\n",
       "\n",
       "<p>numeric, specifying how much compound labels should be put above the\n",
       "default position or node center. This argument is useful when\n",
       "map.cpdname=TRUE, i.e. compounds are labelled by full name, which\n",
       "affects the look of compound nodes and color. Only effective when\n",
       "kegg.native=FALSE. Default cpd.lab.offset=1.0.\n",
       "</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>Pathview maps and renders user data on relevant pathway graphs. Pathview\n",
       "is a stand alone program for pathway based data integration and\n",
       "visualization. It also seamlessly integrates with pathway and functional\n",
       "analysis tools for large-scale and fully automated analysis.\n",
       "Pathview provides strong support for data Integration. It works with: 1)\n",
       "essentially all types of biological data mappable to pathways, 2) over\n",
       "10 types of gene or protein IDs, and 20 types of compound or metabolite\n",
       "IDs, 3) pathways for over 2000 species as well as KEGG orthology, 4)\n",
       "varoius data attributes and formats, i.e. continuous/discrete data,\n",
       "matrices/vectors, single/multiple samples etc.\n",
       "To see mappable external gene/protein IDs do:\n",
       "<code>data(gene.idtype.list)</code>, to see mappable external compound related\n",
       "IDs do: <code>data(rn.list)</code>; names(rn.list).\n",
       "Pathview generates both native KEGG view and Graphviz views for\n",
       "pathways. Currently only KEGG pathways are implemented. Hopefully, pathways from\n",
       "Reactome, NCI and other databases will be supported in the future.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>From viersion 1.9.3, pathview can accept either a single pathway or\n",
       "multiple pathway ids. The result returned by pathview function is a\n",
       "named list corresponding to the input pathway ids. Each element (for\n",
       "each pathway itself is a named list, with 2\n",
       "elements (&quot;plot.data.gene&quot;, &quot;plot.data.cpd&quot;). Both elements\n",
       "are data.frame or NULL depends on the corresponding input\n",
       "data gene.data and cpd.data. These data.frames record the plot data\n",
       "for mapped gene or compound nodes: rows are mapped genes/compounds,\n",
       "columns are:\n",
       "</p>\n",
       "<table summary=\"R valueblock\">\n",
       "<tr valign=\"top\"><td><code>kegg.names</code></td>\n",
       "<td>\n",
       "<p>standard KEGG IDs/Names for mapped nodes. It's\n",
       "Entrez Gene ID or KEGG Compound Accessions.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>labels</code></td>\n",
       "<td>\n",
       "<p>Node labels to be used when needed.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>all.mapped</code></td>\n",
       "<td>\n",
       "<p>All molecule (gene or compound) IDs mapped to this\n",
       "node.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>type</code></td>\n",
       "<td>\n",
       "<p>node type, currently 4 types are supported:\n",
       "&quot;gene&quot;,&quot;enzyme&quot;, &quot;compound&quot; and &quot;ortholog&quot;.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>x</code></td>\n",
       "<td>\n",
       "<p>x coordinate in the original KEGG pathway graph.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>y</code></td>\n",
       "<td>\n",
       "<p>y coordinate in the original KEGG pathway graph.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>width</code></td>\n",
       "<td>\n",
       "<p>node width in the original KEGG pathway graph.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>height</code></td>\n",
       "<td>\n",
       "<p>node height in the original KEGG pathway graph.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>other columns </code></td>\n",
       "<td>\n",
       "<p>columns of the mapped gene/compound data and\n",
       "corresponding pseudo-color codes for individual samples\n",
       "</p>\n",
       "</td></tr>      \n",
       "</table>\n",
       "<p>The results returned by <code>keggview.native</code> and\n",
       "codekeggview.graph are both a list of graph plotting\n",
       "parameters. These are not intended to be used externally.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Weijun Luo &lt;luo_weijun@yahoo.com&gt;\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>Luo, W. and Brouwer, C., Pathview: an R/Bioconductor package for\n",
       "pathway based data integration and visualization. Bioinformatics,\n",
       "2013, 29(14): 1830-1831, doi: 10.1093/bioinformatics/btt285\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>download.kegg</code> the downloader,\n",
       "<code>node.info</code> the parser,\n",
       "<code>node.map</code> and <code>node.color</code> the mapper. \n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "#load data\n",
       "data(gse16873.d)\n",
       "data(demo.paths)\n",
       "\n",
       "#KEGG view: gene data only\n",
       "i &lt;- 1\n",
       "pv.out &lt;- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "kegg.native = TRUE)\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.gene)\n",
       "#result PNG file in current directory\n",
       "\n",
       "#Graphviz view: gene data only\n",
       "pv.out &lt;- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "kegg.native = FALSE, sign.pos = demo.paths$spos[i])\n",
       "#result PDF file in current directory\n",
       "\n",
       "#KEGG view: both gene and compound data\n",
       "sim.cpd.data=sim.mol.data(mol.type=\"cpd\", nmol=3000)\n",
       "i &lt;- 3\n",
       "print(demo.paths$sel.paths[i])\n",
       "pv.out &lt;- pathview(gene.data = gse16873.d[, 1], cpd.data = sim.cpd.data,\n",
       "pathway.id = demo.paths$sel.paths[i], species = \"hsa\", out.suffix =\n",
       "\"gse16873.cpd\", keys.align = \"y\", kegg.native = TRUE, key.pos = demo.paths$kpos1[i])\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.cpd)\n",
       "\n",
       "#multiple states in one graph\n",
       "set.seed(10)\n",
       "sim.cpd.data2 = matrix(sample(sim.cpd.data, 18000, \n",
       "    replace = TRUE), ncol = 6)\n",
       "pv.out &lt;- pathview(gene.data = gse16873.d[, 1:3], \n",
       "    cpd.data = sim.cpd.data2[, 1:2], pathway.id = demo.paths$sel.paths[i], \n",
       "    species = \"hsa\", out.suffix = \"gse16873.cpd.3-2s\", keys.align = \"y\", \n",
       "    kegg.native = TRUE, match.data = FALSE, multi.state = TRUE, same.layer = TRUE)\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.cpd)\n",
       "\n",
       "#result PNG file in current directory\n",
       "\n",
       "##more examples of pathview usages are shown in the vignette.\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>pathview</em> version 1.24.0 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{pathview}{Pathway based data integration and visualization}{pathview}\n",
       "\\aliasA{keggview.graph}{pathview}{keggview.graph}\n",
       "\\aliasA{keggview.native}{pathview}{keggview.native}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Pathview is a tool set for pathway based data integration and\n",
       "visualization. It maps and renders user data on relevant pathway\n",
       "graphs. All users need is to supply their gene or compound data and\n",
       "specify the target pathway. Pathview automatically downloads the pathway\n",
       "graph data, parses the data file, maps user data to the pathway, and\n",
       "render pathway graph with the mapped data.\n",
       "Pathview generates both native KEGG view and Graphviz views for\n",
       "pathways. keggview.native and keggview.graph are the two viewer\n",
       "functions, and pathview is the main function providing a unified\n",
       "interface to downloader, parser, mapper and viewer functions.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "pathview(gene.data = NULL, cpd.data = NULL, pathway.id,\n",
       "species = \"hsa\", kegg.dir = \".\", cpd.idtype = \"kegg\", gene.idtype =\n",
       "\"entrez\", gene.annotpkg = NULL, min.nnodes = 3, kegg.native = TRUE,\n",
       "map.null = TRUE, expand.node = FALSE, split.group = FALSE, map.symbol =\n",
       "TRUE, map.cpdname = TRUE, node.sum = \"sum\", discrete=list(gene=FALSE,\n",
       "cpd=FALSE), limit = list(gene = 1, cpd = 1), bins = list(gene = 10, cpd\n",
       "= 10), both.dirs = list(gene = T, cpd = T), trans.fun = list(gene =\n",
       "NULL, cpd = NULL), low = list(gene = \"green\", cpd = \"blue\"), mid =\n",
       "list(gene = \"gray\", cpd = \"gray\"), high = list(gene = \"red\", cpd =\n",
       "\"yellow\"), na.col = \"transparent\", ...)\n",
       "\n",
       "keggview.native(plot.data.gene = NULL, plot.data.cpd = NULL,\n",
       "cols.ts.gene = NULL, cols.ts.cpd = NULL, node.data, pathway.name,\n",
       "out.suffix = \"pathview\", kegg.dir = \".\", multi.state=TRUE, match.data =\n",
       "TRUE, same.layer = TRUE, res = 300, cex = 0.25, discrete =\n",
       "list(gene=FALSE, cpd=FALSE), limit= list(gene = 1, cpd = 1), bins =\n",
       "list(gene = 10, cpd = 10), both.dirs =list(gene = T, cpd = T), low =\n",
       "list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd =\n",
       "\"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "\"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "\"x\", key.pos = \"topright\", ...)\n",
       "\n",
       "keggview.graph(plot.data.gene = NULL, plot.data.cpd = NULL, cols.ts.gene\n",
       "= NULL, cols.ts.cpd = NULL, node.data, path.graph, pathway.name,\n",
       "out.suffix = \"pathview\", pdf.size = c(7, 7), multi.state=TRUE,\n",
       "same.layer = TRUE, match.data = TRUE, rankdir = c(\"LR\", \"TB\")[1],\n",
       "is.signal = TRUE, split.group = F, afactor = 1, text.width = 15, cex =\n",
       "0.5, map.cpdname = FALSE, cpd.lab.offset = 1.0,\n",
       "discrete=list(gene=FALSE, cpd=FALSE), limit = list(gene = 1, cpd = 1),\n",
       "bins = list(gene = 10, cpd = 10), both.dirs = list(gene = T, cpd = T),\n",
       "low = list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd\n",
       "= \"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "\"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "\"x\", key.pos = \"topright\", sign.pos = \"bottomright\", ...)  \n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{gene.data}] \n",
       "either vector (single sample) or a matrix-like data (multiple\n",
       "sample). Vector should be numeric with gene IDs as names or it may\n",
       "also be character of gene IDs. Character vector is treated as discrete\n",
       "or count data. Matrix-like data structure has genes as rows and\n",
       "samples as columns. Row names should be gene IDs. Here gene ID is a\n",
       "generic concepts, including multiple types of gene, transcript and\n",
       "protein uniquely mappable to KEGG gene IDs. KEGG ortholog IDs are also\n",
       "treated as gene IDs as to handle metagenomic data. Check details for\n",
       "mappable ID types. Default gene.data=NULL.\n",
       "\n",
       "numeric, character, continuous\n",
       "\n",
       "\\item[\\code{cpd.data}] \n",
       "the same as gene.data, excpet named with IDs mappable to KEGG\n",
       "compound IDs. Over 20 types of IDs included in CHEMBL database can be\n",
       "used here. Check details for mappable ID types. Default\n",
       "cpd.data=NULL. Note that gene.data and cpd.data can't be NULL\n",
       "simultaneously.\n",
       "\n",
       "\\item[\\code{pathway.id}] \n",
       "character vector, the KEGG pathway ID(s), usually 5 digit, may also include the 3\n",
       "letter KEGG species code.\n",
       "\n",
       "\\item[\\code{species}] \n",
       "character, either the kegg code, scientific name or the common name of\n",
       "the target species. This applies to both pathway and gene.data or\n",
       "cpd.data. When KEGG ortholog pathway is considered,\n",
       "species=\"ko\". Default species=\"hsa\", it is equivalent to use either\n",
       "\"Homo sapiens\" (scientific name) or \"human\" (common name). \n",
       "\n",
       "\\item[\\code{kegg.dir}] \n",
       "character, the directory of KEGG pathway data file (.xml) and image file\n",
       "(.png). Users may supply their own data files in the same format and\n",
       "naming convention of KEGG's (species code + pathway id,\n",
       "e.g. hsa04110.xml, hsa04110.png etc) in this directory. Default\n",
       "kegg.dir=\".\" (current working directory). \n",
       "\n",
       "\\item[\\code{cpd.idtype}] \n",
       "character, ID type used for the cpd.data. Default cpd.idtype=\"kegg\" (include\n",
       "compound, glycan and drug accessions). \n",
       "\n",
       "\\item[\\code{gene.idtype}] \n",
       "character, ID type used for the gene.data, case insensitive. Default\n",
       "gene.idtype=\"entrez\", i.e. Entrez Gene, which are the primary KEGG gene ID\n",
       "for many common model organisms. For other species, gene.idtype should\n",
       "be set to \"KEGG\" as KEGG use other types of gene IDs. For the common\n",
       "model organisms (to check the list, do: \\code{data(bods); bods}), you\n",
       "may also specify other types of valid IDs. To check the ID list, do:\n",
       "\\code{data(gene.idtype.list); gene.idtype.list}. \n",
       "\n",
       "\\item[\\code{gene.annotpkg}] \n",
       "character, the name of the annotation package to use for mapping between\n",
       "other gene ID types including symbols and Entrez gene ID. Default gene.annotpkg=NULL.\n",
       "\n",
       "\\item[\\code{min.nnodes}] \n",
       "integer, minimal number of nodes of type \"gene\",\"enzyme\", \"compound\" or\n",
       "\"ortholog\" for a pathway to be considered. Default min.nnodes=3.\n",
       "\n",
       "\\item[\\code{kegg.native}] \n",
       "logical, whether to render pathway graph as native KEGG graph (.png) or\n",
       "using graphviz layout engine (.pdf). Default kegg.native=TRUE.\n",
       "\n",
       "\\item[\\code{map.null}] \n",
       "logical, whether to map the NULL gene.data or cpd.data to pathway. When\n",
       "NULL data are mapped, the gene or compound nodes in the pathway will be\n",
       "rendered as actually mapped nodes, except with NA-valued color. When\n",
       "NULL data are not mapped, the nodes are rendered as unmapped nodes. This\n",
       "argument mainly affects native KEGG graph view, i.e. when\n",
       "kegg.native=TRUE. Default map.null=TRUE.\n",
       "\n",
       "\\item[\\code{expand.node}] \n",
       "logical, whether the multiple-gene nodes are expanded into single-gene\n",
       "nodes. Each expanded single-gene nodes inherits all edges from the\n",
       "original multiple-gene node. This option only affects graphviz graph view, i.e. when\n",
       "kegg.native=FALSE. This option is not effective for most metabolic\n",
       "pathways where it conflits with  converting reactions to edges. Default expand.node=FLASE.\n",
       "\n",
       "\\item[\\code{split.group}] \n",
       "logical, whether split node groups are split to individual nodes. Each\n",
       "split member nodes inherits all edges from the node group. This\n",
       "option only affects graphviz graph view, i.e. when\n",
       "kegg.native=FALSE. This option also effects most metabolic\n",
       "pathways even without group nodes defined orginally. For these pathways,\n",
       "genes involved in the same reaction are grouped automatically when\n",
       "converting reactions to edges unless split.group=TRUE. d\n",
       "split.group=FLASE.\n",
       "\n",
       "\\item[\\code{map.symbol}] \n",
       "logical, whether map gene IDs to symbols for gene node labels or use the\n",
       "graphic name from the KGML file. This option is only effective for\n",
       "kegg.native=FALSE or same.layer=FALSE when kegg.native=TRUE. For\n",
       "same.layer=TRUE when kegg.native=TRUE, the native KEGG labels will be\n",
       "kept. Default map.symbol=TRUE.\n",
       "\n",
       "\\item[\\code{map.cpdname}] \n",
       "logical, whether map compound IDs to formal names for compound node labels or use the\n",
       "graphic name from the KGML file (KEGG compound accessions). This option is only effective for\n",
       "kegg.native=FALSE. When kegg.native=TRUE, the native KEGG labels will be\n",
       "kept. Default map.cpdname=TRUE.\n",
       "\n",
       "\\item[\\code{node.sum}] \n",
       "character, the method name to calculate node summary given that\n",
       "multiple genes or compounds are mapped to it. Poential options include\n",
       "\"sum\",\"mean\", \"median\", \"max\", \"max.abs\" and \"random\". Default node.sum=\"sum\".\n",
       "\n",
       "\\item[\\code{discrete}] \n",
       "a list of two logical elements with \"gene\" and \"cpd\" as the names. This\n",
       "argument tells whether gene.data or cpd.data should be treated as\n",
       "discrete. Default dsicrete=list(gene=FALSE, cpd=FALSE), i.e. both data should\n",
       "be treated as continuous.\n",
       "\n",
       "\\item[\\code{limit}] \n",
       "a list of two numeric elements with \"gene\" and \"cpd\" as the names. This\n",
       "argument specifies the limit values for gene.data and cpd.data when converting\n",
       "them to pseudo colors. Each element of the list could be of length 1 or\n",
       "2. Length 1 suggests discrete data or 1 directional (positive-valued)\n",
       "data, or the absolute limit for 2 directional data. Length 2 suggests 2\n",
       "directional data. Default limit=list(gene=1, cpd=1).\n",
       "\n",
       "\\item[\\code{bins}] \n",
       "a list of two integer elements with \"gene\" and \"cpd\" as the names. This\n",
       "argument specifies the number of levels or bins for gene.data and cpd.data when converting\n",
       "them to pseudo colors. Default limit=list(gene=10, cpd=10).\n",
       "\n",
       "\\item[\\code{both.dirs}] \n",
       "a list of two logical  elements with \"gene\" and \"cpd\" as the names. This\n",
       "argument specifies whether gene.data and cpd.data are 1 directional or 2\n",
       "directional data when converting them to pseudo colors. Default limit=list(gene=TRUE, cpd=TRUE).\n",
       "\n",
       "\\item[\\code{trans.fun}] \n",
       "a list of two function (not character) elements with \"gene\" and \"cpd\" as the names. This\n",
       "argument specifies whether and how gene.data and cpd.data are\n",
       "transformed. Examples are \\code{log}, \\code{abs} or users' own\n",
       "functions. Default limit=list(gene=NULL, cpd=NULL).\n",
       "\n",
       "\\item[\\code{low, mid, high}] \n",
       "each is a list of two colors with \"gene\" and \"cpd\" as the names. This\n",
       "argument specifies the color spectra to code gene.data and\n",
       "cpd.data. When data are 1 directional (TRUE value in both.dirs), only\n",
       "mid and high are used to specify the color spectra. Default spectra (low-mid-high)\n",
       "\"green\"-\"gray\"-\"red\" and \"blue\"-\"gray\"-\"yellow\" are used for gene.data\n",
       "and cpd.data respectively. \n",
       "The values for 'low, mid, high' can be given as color names\n",
       "('red'), plot color index (2=red), and HTML-style RGB,\n",
       "(\"\\bsl{}\\#FF0000\"=red).\n",
       "\n",
       "\\item[\\code{na.col}] \n",
       "color used for NA's or missing values in gene.data and cpd.data. d\n",
       "na.col=\"transparent\". \n",
       "\n",
       "\\item[\\code{...}] \n",
       "extra arguments passed to keggview.native or keggview.graph function.\n",
       "\n",
       "\n",
       "special arguments for keggview.native or keggview.graph function.\n",
       "\\item[\\code{plot.data.gene}] \n",
       "data.frame returned by node.map function for rendering mapped gene\n",
       "nodes, including node name, type, positions (x, y), sizes (width,\n",
       "height), and mapped gene.data. This data is also used as input for\n",
       "pseduo-color coding through node.color function. Default plot.data.gene=NULL.\n",
       "\n",
       "\\item[\\code{plot.data.cpd}] \n",
       "same as plot.data.gene function, except for mapped compound node data. d\n",
       "plot.data.cpd=NULL. Default plot.data.cpd=NULL. Note that plot.data.gene and\n",
       "plot.data.cpd can't be NULL simultaneously.\n",
       "\n",
       "\\item[\\code{cols.ts.gene}] \n",
       "vector or matrix of colors returned by node.color function for rendering\n",
       "gene.data. Dimensionality is the same as the latter. Default cols.ts.gene=NULL.\n",
       "\n",
       "\\item[\\code{cols.ts.cpd}] \n",
       "same as cols.ts.gene, except corresponding to cpd.data. d\n",
       "cols.ts.cpd=NULL. Note that cols.ts.gene and cols.ts.cpd\n",
       "plot.data.gene can't be NULL simultaneously. \n",
       "\n",
       "\\item[\\code{node.data}] \n",
       "list returned by node.info function, which parse KGML file directly or\n",
       "indirectly, and extract the node data.\n",
       "\n",
       "\\item[\\code{pathway.name}] \n",
       "character, the full KEGG pathway name in the format of 3-letter species\n",
       "code with 5-digit pathway id, eg \"hsa04612\". \n",
       "\n",
       "\\item[\\code{out.suffix}] \n",
       "character, the suffix to be added after the pathway name as part of the\n",
       "output graph file. Sample names or column names of the gene.data or\n",
       "cpd.data are also added when there are multiple samples. Default out.suffix=\"pathview\".\n",
       "\n",
       "\\item[\\code{multi.state}] \n",
       "logical, whether multiple states (samples or columns) gene.data or\n",
       "cpd.data should be integrated and plotted in the same graph. Default\n",
       "match.data=TRUE. In other words, gene or compound nodes will be sliced\n",
       "into multiple pieces corresponding to the number of states in the\n",
       "data.\n",
       "\n",
       "\\item[\\code{match.data}] \n",
       "logical, whether the samples of gene.data and cpd.data are\n",
       "paired. Default match.data=TRUE. When let sample sizes of gene.data\n",
       "and cpd.data be m and n, when m>n, extra columns of NA's (mapped to no\n",
       "color) will be added to cpd.data as to make the sample size the\n",
       "same. This will result in the smae number of slice in gene nodes and\n",
       "compound when multi.state=TRUE.\n",
       "\n",
       "\\item[\\code{same.layer}] \n",
       "logical, control plotting layers:\n",
       "1) if node colors be plotted in the same layer as the pathway\n",
       "graph when kegg.native=TRUE, 2) if  edge/node type legend be plotted in\n",
       "the same page when kegg.native=FALSE.\n",
       "\n",
       "\\item[\\code{res}] \n",
       "The nominal resolution in ppi which will be recorded in the\n",
       "bitmap file, if a positive integer.  Also used for 'units'\n",
       "other than the default, and to convert points to pixels. This\n",
       "argument is only effective when kegg.native=TRUE. Default res=300.\n",
       "\n",
       "\\item[\\code{cex}] \n",
       "A numerical value giving the amount by which plotting text\n",
       "and symbols should be scaled relative to the default 1. Default cex=0.25\n",
       "when kegg.native=TRUE, cex=0.5 when kegg.native=FALSE.\n",
       "\n",
       "\\item[\\code{new.signature}] \n",
       "logical, whether pathview signature is added to the pathway\n",
       "graphs. Default new.signature=TRUE.\n",
       "\n",
       "\\item[\\code{plot.col.key}] \n",
       "logical, whether color key is added to the pathway\n",
       "graphs. Default plot.col.key= TRUE.\n",
       "\n",
       "\n",
       "\\item[\\code{key.align}] \n",
       "character, controlling how the color keys are aligned when both\n",
       "gene.data and cpd.data are not NULL. Potential values are \"x\", aligned\n",
       "by x coordinates, and \"y\", aligned by y coordinates. Default key.align=\"x\".\n",
       "\n",
       "\\item[\\code{key.pos}] \n",
       "character, controlling the position of color key(s). Potentail values\n",
       "are \"bottomleft\", \"bottomright\", \"topleft\" and \"topright\". d\n",
       "key.pos=\"topright\". \n",
       "\n",
       "\\item[\\code{sign.pos}] \n",
       "character, controlling the position of pathview signature. Only\n",
       "effective when kegg.native=FALSE, Signature position is fixed in place\n",
       "of the original KEGG signature when kegg.native=TRUE. Potentail values\n",
       "are \"bottomleft\", \"bottomright\", \"topleft\" and \"topright\". d\n",
       "sign.pos=\"bottomright\". \n",
       "\n",
       "\n",
       "\\item[\\code{path.graph}] \n",
       "a graph object parsed from KGML file, only effective when\n",
       "kegg.native=FALSE.\n",
       "\n",
       "\\item[\\code{pdf.size}] \n",
       "a numeric vector of length 2, giving the width and height of the pathway\n",
       "graph pdf file. Note that pdf width increase by half when\n",
       "same.layer=TRUE to accommodate legends. Only effective when\n",
       "kegg.native=FALSE. Default pdf.size=c(7,7).\n",
       "\n",
       "\\item[\\code{rankdir}] \n",
       "character, either \"LR\" (left to right) or \"TB\" (top to bottom),\n",
       "specifying the pathway graph layout direction. Only effective when\n",
       "kegg.native=FALSE. Default rank.dir=\"LR\".\n",
       "\n",
       "\\item[\\code{is.signal}] \n",
       "logical, if the pathway is treated as a signaling pathway, where all the\n",
       "unconnected nodes are dropped. This argument also affect the graph\n",
       "layout type, i.e. \"dot\" for signals or \"neato\" otherwise. Only effective when\n",
       "kegg.native=FALSE. Default is.signal=TRUE.\n",
       "\n",
       "\\item[\\code{afactor}] \n",
       "numeric, node amplifying factor. This argument is for node size\n",
       "fine-tuning, its effect is subtler than expected. Only effective when\n",
       "kegg.native=FALSE. Default afctor=1.\n",
       "\n",
       "\\item[\\code{text.width}] \n",
       "numeric, specifying the line width for text wrap. Only effective when\n",
       "kegg.native= FALSE. Default text.width=15 (characters).\n",
       "\n",
       "\\item[\\code{cpd.lab.offset}] \n",
       "numeric, specifying how much compound labels should be put above the\n",
       "default position or node center. This argument is useful when\n",
       "map.cpdname=TRUE, i.e. compounds are labelled by full name, which\n",
       "affects the look of compound nodes and color. Only effective when\n",
       "kegg.native=FALSE. Default cpd.lab.offset=1.0.\n",
       "\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "Pathview maps and renders user data on relevant pathway graphs. Pathview\n",
       "is a stand alone program for pathway based data integration and\n",
       "visualization. It also seamlessly integrates with pathway and functional\n",
       "analysis tools for large-scale and fully automated analysis.\n",
       "Pathview provides strong support for data Integration. It works with: 1)\n",
       "essentially all types of biological data mappable to pathways, 2) over\n",
       "10 types of gene or protein IDs, and 20 types of compound or metabolite\n",
       "IDs, 3) pathways for over 2000 species as well as KEGG orthology, 4)\n",
       "varoius data attributes and formats, i.e. continuous/discrete data,\n",
       "matrices/vectors, single/multiple samples etc.\n",
       "To see mappable external gene/protein IDs do:\n",
       "\\code{data(gene.idtype.list)}, to see mappable external compound related\n",
       "IDs do: \\code{data(rn.list)}; names(rn.list).\n",
       "Pathview generates both native KEGG view and Graphviz views for\n",
       "pathways. Currently only KEGG pathways are implemented. Hopefully, pathways from\n",
       "Reactome, NCI and other databases will be supported in the future.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "From viersion 1.9.3, pathview can accept either a single pathway or\n",
       "multiple pathway ids. The result returned by pathview function is a\n",
       "named list corresponding to the input pathway ids. Each element (for\n",
       "each pathway itself is a named list, with 2\n",
       "elements (\"plot.data.gene\", \"plot.data.cpd\"). Both elements\n",
       "are data.frame or NULL depends on the corresponding input\n",
       "data gene.data and cpd.data. These data.frames record the plot data\n",
       "for mapped gene or compound nodes: rows are mapped genes/compounds,\n",
       "columns are:\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{kegg.names}] standard KEGG IDs/Names for mapped nodes. It's\n",
       "Entrez Gene ID or KEGG Compound Accessions.\n",
       "\\item[\\code{labels}] Node labels to be used when needed.\n",
       "\n",
       "\\item[\\code{all.mapped}] All molecule (gene or compound) IDs mapped to this\n",
       "node.\n",
       "\n",
       "\\item[\\code{type}] node type, currently 4 types are supported:\n",
       "\"gene\",\"enzyme\", \"compound\" and \"ortholog\".\n",
       "\n",
       "\\item[\\code{x}] x coordinate in the original KEGG pathway graph.\n",
       "\n",
       "\\item[\\code{y}] y coordinate in the original KEGG pathway graph.\n",
       "\n",
       "\\item[\\code{width}] node width in the original KEGG pathway graph.\n",
       "\n",
       "\\item[\\code{height}] node height in the original KEGG pathway graph.\n",
       "\n",
       "\\item[\\code{other columns }] columns of the mapped gene/compound data and\n",
       "corresponding pseudo-color codes for individual samples\n",
       "     \n",
       "\n",
       "\\end{ldescription}\n",
       "The results returned by \\code{keggview.native} and\n",
       "codekeggview.graph are both a list of graph plotting\n",
       "parameters. These are not intended to be used externally.\n",
       "\n",
       "\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "Weijun Luo <luo\\_weijun@yahoo.com>\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "Luo, W. and Brouwer, C., Pathview: an R/Bioconductor package for\n",
       "pathway based data integration and visualization. Bioinformatics,\n",
       "2013, 29(14): 1830-1831, doi: 10.1093/bioinformatics/btt285\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\code{\\LinkA{download.kegg}{download.kegg}} the downloader,\n",
       "\\code{\\LinkA{node.info}{node.info}} the parser,\n",
       "\\code{\\LinkA{node.map}{node.map}} and \\code{\\LinkA{node.color}{node.color}} the mapper. \n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "#load data\n",
       "data(gse16873.d)\n",
       "data(demo.paths)\n",
       "\n",
       "#KEGG view: gene data only\n",
       "i <- 1\n",
       "pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "kegg.native = TRUE)\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.gene)\n",
       "#result PNG file in current directory\n",
       "\n",
       "#Graphviz view: gene data only\n",
       "pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "kegg.native = FALSE, sign.pos = demo.paths$spos[i])\n",
       "#result PDF file in current directory\n",
       "\n",
       "#KEGG view: both gene and compound data\n",
       "sim.cpd.data=sim.mol.data(mol.type=\"cpd\", nmol=3000)\n",
       "i <- 3\n",
       "print(demo.paths$sel.paths[i])\n",
       "pv.out <- pathview(gene.data = gse16873.d[, 1], cpd.data = sim.cpd.data,\n",
       "pathway.id = demo.paths$sel.paths[i], species = \"hsa\", out.suffix =\n",
       "\"gse16873.cpd\", keys.align = \"y\", kegg.native = TRUE, key.pos = demo.paths$kpos1[i])\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.cpd)\n",
       "\n",
       "#multiple states in one graph\n",
       "set.seed(10)\n",
       "sim.cpd.data2 = matrix(sample(sim.cpd.data, 18000, \n",
       "    replace = TRUE), ncol = 6)\n",
       "pv.out <- pathview(gene.data = gse16873.d[, 1:3], \n",
       "    cpd.data = sim.cpd.data2[, 1:2], pathway.id = demo.paths$sel.paths[i], \n",
       "    species = \"hsa\", out.suffix = \"gse16873.cpd.3-2s\", keys.align = \"y\", \n",
       "    kegg.native = TRUE, match.data = FALSE, multi.state = TRUE, same.layer = TRUE)\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.cpd)\n",
       "\n",
       "#result PNG file in current directory\n",
       "\n",
       "##more examples of pathview usages are shown in the vignette.\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "pathview               package:pathview                R Documentation\n",
       "\n",
       "_\bP_\ba_\bt_\bh_\bw_\ba_\by _\bb_\ba_\bs_\be_\bd _\bd_\ba_\bt_\ba _\bi_\bn_\bt_\be_\bg_\br_\ba_\bt_\bi_\bo_\bn _\ba_\bn_\bd _\bv_\bi_\bs_\bu_\ba_\bl_\bi_\bz_\ba_\bt_\bi_\bo_\bn\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Pathview is a tool set for pathway based data integration and\n",
       "     visualization. It maps and renders user data on relevant pathway\n",
       "     graphs. All users need is to supply their gene or compound data\n",
       "     and specify the target pathway. Pathview automatically downloads\n",
       "     the pathway graph data, parses the data file, maps user data to\n",
       "     the pathway, and render pathway graph with the mapped data.\n",
       "     Pathview generates both native KEGG view and Graphviz views for\n",
       "     pathways. keggview.native and keggview.graph are the two viewer\n",
       "     functions, and pathview is the main function providing a unified\n",
       "     interface to downloader, parser, mapper and viewer functions.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     pathview(gene.data = NULL, cpd.data = NULL, pathway.id,\n",
       "     species = \"hsa\", kegg.dir = \".\", cpd.idtype = \"kegg\", gene.idtype =\n",
       "     \"entrez\", gene.annotpkg = NULL, min.nnodes = 3, kegg.native = TRUE,\n",
       "     map.null = TRUE, expand.node = FALSE, split.group = FALSE, map.symbol =\n",
       "     TRUE, map.cpdname = TRUE, node.sum = \"sum\", discrete=list(gene=FALSE,\n",
       "     cpd=FALSE), limit = list(gene = 1, cpd = 1), bins = list(gene = 10, cpd\n",
       "     = 10), both.dirs = list(gene = T, cpd = T), trans.fun = list(gene =\n",
       "     NULL, cpd = NULL), low = list(gene = \"green\", cpd = \"blue\"), mid =\n",
       "     list(gene = \"gray\", cpd = \"gray\"), high = list(gene = \"red\", cpd =\n",
       "     \"yellow\"), na.col = \"transparent\", ...)\n",
       "     \n",
       "     keggview.native(plot.data.gene = NULL, plot.data.cpd = NULL,\n",
       "     cols.ts.gene = NULL, cols.ts.cpd = NULL, node.data, pathway.name,\n",
       "     out.suffix = \"pathview\", kegg.dir = \".\", multi.state=TRUE, match.data =\n",
       "     TRUE, same.layer = TRUE, res = 300, cex = 0.25, discrete =\n",
       "     list(gene=FALSE, cpd=FALSE), limit= list(gene = 1, cpd = 1), bins =\n",
       "     list(gene = 10, cpd = 10), both.dirs =list(gene = T, cpd = T), low =\n",
       "     list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd =\n",
       "     \"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "     \"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "     \"x\", key.pos = \"topright\", ...)\n",
       "     \n",
       "     keggview.graph(plot.data.gene = NULL, plot.data.cpd = NULL, cols.ts.gene\n",
       "     = NULL, cols.ts.cpd = NULL, node.data, path.graph, pathway.name,\n",
       "     out.suffix = \"pathview\", pdf.size = c(7, 7), multi.state=TRUE,\n",
       "     same.layer = TRUE, match.data = TRUE, rankdir = c(\"LR\", \"TB\")[1],\n",
       "     is.signal = TRUE, split.group = F, afactor = 1, text.width = 15, cex =\n",
       "     0.5, map.cpdname = FALSE, cpd.lab.offset = 1.0,\n",
       "     discrete=list(gene=FALSE, cpd=FALSE), limit = list(gene = 1, cpd = 1),\n",
       "     bins = list(gene = 10, cpd = 10), both.dirs = list(gene = T, cpd = T),\n",
       "     low = list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd\n",
       "     = \"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "     \"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "     \"x\", key.pos = \"topright\", sign.pos = \"bottomright\", ...)  \n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "gene.data: either vector (single sample) or a matrix-like data\n",
       "          (multiple sample). Vector should be numeric with gene IDs as\n",
       "          names or it may also be character of gene IDs. Character\n",
       "          vector is treated as discrete or count data. Matrix-like data\n",
       "          structure has genes as rows and samples as columns. Row names\n",
       "          should be gene IDs. Here gene ID is a generic concepts,\n",
       "          including multiple types of gene, transcript and protein\n",
       "          uniquely mappable to KEGG gene IDs. KEGG ortholog IDs are\n",
       "          also treated as gene IDs as to handle metagenomic data. Check\n",
       "          details for mappable ID types. Default gene.data=NULL.\n",
       "\n",
       "          numeric, character, continuous\n",
       "\n",
       "cpd.data: the same as gene.data, excpet named with IDs mappable to KEGG\n",
       "          compound IDs. Over 20 types of IDs included in CHEMBL\n",
       "          database can be used here. Check details for mappable ID\n",
       "          types. Default cpd.data=NULL. Note that gene.data and\n",
       "          cpd.data can't be NULL simultaneously.\n",
       "\n",
       "pathway.id: character vector, the KEGG pathway ID(s), usually 5 digit,\n",
       "          may also include the 3 letter KEGG species code.\n",
       "\n",
       " species: character, either the kegg code, scientific name or the\n",
       "          common name of the target species. This applies to both\n",
       "          pathway and gene.data or cpd.data. When KEGG ortholog pathway\n",
       "          is considered, species=\"ko\". Default species=\"hsa\", it is\n",
       "          equivalent to use either \"Homo sapiens\" (scientific name) or\n",
       "          \"human\" (common name).\n",
       "\n",
       "kegg.dir: character, the directory of KEGG pathway data file (.xml) and\n",
       "          image file (.png). Users may supply their own data files in\n",
       "          the same format and naming convention of KEGG's (species code\n",
       "          + pathway id, e.g. hsa04110.xml, hsa04110.png etc) in this\n",
       "          directory. Default kegg.dir=\".\" (current working directory).\n",
       "\n",
       "cpd.idtype: character, ID type used for the cpd.data. Default\n",
       "          cpd.idtype=\"kegg\" (include compound, glycan and drug\n",
       "          accessions).\n",
       "\n",
       "gene.idtype: character, ID type used for the gene.data, case\n",
       "          insensitive. Default gene.idtype=\"entrez\", i.e. Entrez Gene,\n",
       "          which are the primary KEGG gene ID for many common model\n",
       "          organisms. For other species, gene.idtype should be set to\n",
       "          \"KEGG\" as KEGG use other types of gene IDs. For the common\n",
       "          model organisms (to check the list, do: ‘data(bods); bods’),\n",
       "          you may also specify other types of valid IDs. To check the\n",
       "          ID list, do: ‘data(gene.idtype.list); gene.idtype.list’.\n",
       "\n",
       "gene.annotpkg: character, the name of the annotation package to use for\n",
       "          mapping between other gene ID types including symbols and\n",
       "          Entrez gene ID. Default gene.annotpkg=NULL.\n",
       "\n",
       "min.nnodes: integer, minimal number of nodes of type \"gene\",\"enzyme\",\n",
       "          \"compound\" or \"ortholog\" for a pathway to be considered.\n",
       "          Default min.nnodes=3.\n",
       "\n",
       "kegg.native: logical, whether to render pathway graph as native KEGG\n",
       "          graph (.png) or using graphviz layout engine (.pdf). Default\n",
       "          kegg.native=TRUE.\n",
       "\n",
       "map.null: logical, whether to map the NULL gene.data or cpd.data to\n",
       "          pathway. When NULL data are mapped, the gene or compound\n",
       "          nodes in the pathway will be rendered as actually mapped\n",
       "          nodes, except with NA-valued color. When NULL data are not\n",
       "          mapped, the nodes are rendered as unmapped nodes. This\n",
       "          argument mainly affects native KEGG graph view, i.e. when\n",
       "          kegg.native=TRUE. Default map.null=TRUE.\n",
       "\n",
       "expand.node: logical, whether the multiple-gene nodes are expanded into\n",
       "          single-gene nodes. Each expanded single-gene nodes inherits\n",
       "          all edges from the original multiple-gene node. This option\n",
       "          only affects graphviz graph view, i.e. when\n",
       "          kegg.native=FALSE. This option is not effective for most\n",
       "          metabolic pathways where it conflits with converting\n",
       "          reactions to edges. Default expand.node=FLASE.\n",
       "\n",
       "split.group: logical, whether split node groups are split to individual\n",
       "          nodes. Each split member nodes inherits all edges from the\n",
       "          node group. This option only affects graphviz graph view,\n",
       "          i.e. when kegg.native=FALSE. This option also effects most\n",
       "          metabolic pathways even without group nodes defined\n",
       "          orginally. For these pathways, genes involved in the same\n",
       "          reaction are grouped automatically when converting reactions\n",
       "          to edges unless split.group=TRUE. d split.group=FLASE.\n",
       "\n",
       "map.symbol: logical, whether map gene IDs to symbols for gene node\n",
       "          labels or use the graphic name from the KGML file. This\n",
       "          option is only effective for kegg.native=FALSE or\n",
       "          same.layer=FALSE when kegg.native=TRUE. For same.layer=TRUE\n",
       "          when kegg.native=TRUE, the native KEGG labels will be kept.\n",
       "          Default map.symbol=TRUE.\n",
       "\n",
       "map.cpdname: logical, whether map compound IDs to formal names for\n",
       "          compound node labels or use the graphic name from the KGML\n",
       "          file (KEGG compound accessions). This option is only\n",
       "          effective for kegg.native=FALSE. When kegg.native=TRUE, the\n",
       "          native KEGG labels will be kept. Default map.cpdname=TRUE.\n",
       "\n",
       "node.sum: character, the method name to calculate node summary given\n",
       "          that multiple genes or compounds are mapped to it. Poential\n",
       "          options include \"sum\",\"mean\", \"median\", \"max\", \"max.abs\" and\n",
       "          \"random\". Default node.sum=\"sum\".\n",
       "\n",
       "discrete: a list of two logical elements with \"gene\" and \"cpd\" as the\n",
       "          names. This argument tells whether gene.data or cpd.data\n",
       "          should be treated as discrete. Default\n",
       "          dsicrete=list(gene=FALSE, cpd=FALSE), i.e. both data should\n",
       "          be treated as continuous.\n",
       "\n",
       "   limit: a list of two numeric elements with \"gene\" and \"cpd\" as the\n",
       "          names. This argument specifies the limit values for gene.data\n",
       "          and cpd.data when converting them to pseudo colors. Each\n",
       "          element of the list could be of length 1 or 2. Length 1\n",
       "          suggests discrete data or 1 directional (positive-valued)\n",
       "          data, or the absolute limit for 2 directional data. Length 2\n",
       "          suggests 2 directional data. Default limit=list(gene=1,\n",
       "          cpd=1).\n",
       "\n",
       "    bins: a list of two integer elements with \"gene\" and \"cpd\" as the\n",
       "          names. This argument specifies the number of levels or bins\n",
       "          for gene.data and cpd.data when converting them to pseudo\n",
       "          colors. Default limit=list(gene=10, cpd=10).\n",
       "\n",
       "both.dirs: a list of two logical elements with \"gene\" and \"cpd\" as the\n",
       "          names. This argument specifies whether gene.data and cpd.data\n",
       "          are 1 directional or 2 directional data when converting them\n",
       "          to pseudo colors. Default limit=list(gene=TRUE, cpd=TRUE).\n",
       "\n",
       "trans.fun: a list of two function (not character) elements with \"gene\"\n",
       "          and \"cpd\" as the names. This argument specifies whether and\n",
       "          how gene.data and cpd.data are transformed. Examples are\n",
       "          ‘log’, ‘abs’ or users' own functions. Default\n",
       "          limit=list(gene=NULL, cpd=NULL).\n",
       "\n",
       "low, mid, high: each is a list of two colors with \"gene\" and \"cpd\" as\n",
       "          the names. This argument specifies the color spectra to code\n",
       "          gene.data and cpd.data. When data are 1 directional (TRUE\n",
       "          value in both.dirs), only mid and high are used to specify\n",
       "          the color spectra. Default spectra (low-mid-high)\n",
       "          \"green\"-\"gray\"-\"red\" and \"blue\"-\"gray\"-\"yellow\" are used for\n",
       "          gene.data and cpd.data respectively.  The values for 'low,\n",
       "          mid, high' can be given as color names ('red'), plot color\n",
       "          index (2=red), and HTML-style RGB, (\"\\#FF0000\"=red).\n",
       "\n",
       "  na.col: color used for NA's or missing values in gene.data and\n",
       "          cpd.data. d na.col=\"transparent\".\n",
       "\n",
       "     ...: extra arguments passed to keggview.native or keggview.graph\n",
       "          function.\n",
       "     special arguments for keggview.native or keggview.graph function.\n",
       "\n",
       "plot.data.gene: data.frame returned by node.map function for rendering\n",
       "          mapped gene nodes, including node name, type, positions (x,\n",
       "          y), sizes (width, height), and mapped gene.data. This data is\n",
       "          also used as input for pseduo-color coding through node.color\n",
       "          function. Default plot.data.gene=NULL.\n",
       "\n",
       "plot.data.cpd: same as plot.data.gene function, except for mapped\n",
       "          compound node data. d plot.data.cpd=NULL. Default\n",
       "          plot.data.cpd=NULL. Note that plot.data.gene and\n",
       "          plot.data.cpd can't be NULL simultaneously.\n",
       "\n",
       "cols.ts.gene: vector or matrix of colors returned by node.color\n",
       "          function for rendering gene.data. Dimensionality is the same\n",
       "          as the latter. Default cols.ts.gene=NULL.\n",
       "\n",
       "cols.ts.cpd: same as cols.ts.gene, except corresponding to cpd.data. d\n",
       "          cols.ts.cpd=NULL. Note that cols.ts.gene and cols.ts.cpd\n",
       "          plot.data.gene can't be NULL simultaneously.\n",
       "\n",
       "node.data: list returned by node.info function, which parse KGML file\n",
       "          directly or indirectly, and extract the node data.\n",
       "\n",
       "pathway.name: character, the full KEGG pathway name in the format of\n",
       "          3-letter species code with 5-digit pathway id, eg \"hsa04612\".\n",
       "\n",
       "out.suffix: character, the suffix to be added after the pathway name as\n",
       "          part of the output graph file. Sample names or column names\n",
       "          of the gene.data or cpd.data are also added when there are\n",
       "          multiple samples. Default out.suffix=\"pathview\".\n",
       "\n",
       "multi.state: logical, whether multiple states (samples or columns)\n",
       "          gene.data or cpd.data should be integrated and plotted in the\n",
       "          same graph. Default match.data=TRUE. In other words, gene or\n",
       "          compound nodes will be sliced into multiple pieces\n",
       "          corresponding to the number of states in the data.\n",
       "\n",
       "match.data: logical, whether the samples of gene.data and cpd.data are\n",
       "          paired. Default match.data=TRUE. When let sample sizes of\n",
       "          gene.data and cpd.data be m and n, when m>n, extra columns of\n",
       "          NA's (mapped to no color) will be added to cpd.data as to\n",
       "          make the sample size the same. This will result in the smae\n",
       "          number of slice in gene nodes and compound when\n",
       "          multi.state=TRUE.\n",
       "\n",
       "same.layer: logical, control plotting layers: 1) if node colors be\n",
       "          plotted in the same layer as the pathway graph when\n",
       "          kegg.native=TRUE, 2) if edge/node type legend be plotted in\n",
       "          the same page when kegg.native=FALSE.\n",
       "\n",
       "     res: The nominal resolution in ppi which will be recorded in the\n",
       "          bitmap file, if a positive integer.  Also used for 'units'\n",
       "          other than the default, and to convert points to pixels. This\n",
       "          argument is only effective when kegg.native=TRUE. Default\n",
       "          res=300.\n",
       "\n",
       "     cex: A numerical value giving the amount by which plotting text\n",
       "          and symbols should be scaled relative to the default 1.\n",
       "          Default cex=0.25 when kegg.native=TRUE, cex=0.5 when\n",
       "          kegg.native=FALSE.\n",
       "\n",
       "new.signature: logical, whether pathview signature is added to the\n",
       "          pathway graphs. Default new.signature=TRUE.\n",
       "\n",
       "plot.col.key: logical, whether color key is added to the pathway\n",
       "          graphs. Default plot.col.key= TRUE.\n",
       "\n",
       "key.align: character, controlling how the color keys are aligned when\n",
       "          both gene.data and cpd.data are not NULL. Potential values\n",
       "          are \"x\", aligned by x coordinates, and \"y\", aligned by y\n",
       "          coordinates. Default key.align=\"x\".\n",
       "\n",
       " key.pos: character, controlling the position of color key(s).\n",
       "          Potentail values are \"bottomleft\", \"bottomright\", \"topleft\"\n",
       "          and \"topright\". d key.pos=\"topright\".\n",
       "\n",
       "sign.pos: character, controlling the position of pathview signature.\n",
       "          Only effective when kegg.native=FALSE, Signature position is\n",
       "          fixed in place of the original KEGG signature when\n",
       "          kegg.native=TRUE. Potentail values are \"bottomleft\",\n",
       "          \"bottomright\", \"topleft\" and \"topright\". d\n",
       "          sign.pos=\"bottomright\".\n",
       "\n",
       "path.graph: a graph object parsed from KGML file, only effective when\n",
       "          kegg.native=FALSE.\n",
       "\n",
       "pdf.size: a numeric vector of length 2, giving the width and height of\n",
       "          the pathway graph pdf file. Note that pdf width increase by\n",
       "          half when same.layer=TRUE to accommodate legends. Only\n",
       "          effective when kegg.native=FALSE. Default pdf.size=c(7,7).\n",
       "\n",
       " rankdir: character, either \"LR\" (left to right) or \"TB\" (top to\n",
       "          bottom), specifying the pathway graph layout direction. Only\n",
       "          effective when kegg.native=FALSE. Default rank.dir=\"LR\".\n",
       "\n",
       "is.signal: logical, if the pathway is treated as a signaling pathway,\n",
       "          where all the unconnected nodes are dropped. This argument\n",
       "          also affect the graph layout type, i.e. \"dot\" for signals or\n",
       "          \"neato\" otherwise. Only effective when kegg.native=FALSE.\n",
       "          Default is.signal=TRUE.\n",
       "\n",
       " afactor: numeric, node amplifying factor. This argument is for node\n",
       "          size fine-tuning, its effect is subtler than expected. Only\n",
       "          effective when kegg.native=FALSE. Default afctor=1.\n",
       "\n",
       "text.width: numeric, specifying the line width for text wrap. Only\n",
       "          effective when kegg.native= FALSE. Default text.width=15\n",
       "          (characters).\n",
       "\n",
       "cpd.lab.offset: numeric, specifying how much compound labels should be\n",
       "          put above the default position or node center. This argument\n",
       "          is useful when map.cpdname=TRUE, i.e. compounds are labelled\n",
       "          by full name, which affects the look of compound nodes and\n",
       "          color. Only effective when kegg.native=FALSE. Default\n",
       "          cpd.lab.offset=1.0.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     Pathview maps and renders user data on relevant pathway graphs.\n",
       "     Pathview is a stand alone program for pathway based data\n",
       "     integration and visualization. It also seamlessly integrates with\n",
       "     pathway and functional analysis tools for large-scale and fully\n",
       "     automated analysis. Pathview provides strong support for data\n",
       "     Integration. It works with: 1) essentially all types of biological\n",
       "     data mappable to pathways, 2) over 10 types of gene or protein\n",
       "     IDs, and 20 types of compound or metabolite IDs, 3) pathways for\n",
       "     over 2000 species as well as KEGG orthology, 4) varoius data\n",
       "     attributes and formats, i.e. continuous/discrete data,\n",
       "     matrices/vectors, single/multiple samples etc. To see mappable\n",
       "     external gene/protein IDs do: ‘data(gene.idtype.list)’, to see\n",
       "     mappable external compound related IDs do: ‘data(rn.list)’;\n",
       "     names(rn.list). Pathview generates both native KEGG view and\n",
       "     Graphviz views for pathways. Currently only KEGG pathways are\n",
       "     implemented. Hopefully, pathways from Reactome, NCI and other\n",
       "     databases will be supported in the future.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     From viersion 1.9.3, pathview can accept either a single pathway\n",
       "     or multiple pathway ids. The result returned by pathview function\n",
       "     is a named list corresponding to the input pathway ids. Each\n",
       "     element (for each pathway itself is a named list, with 2 elements\n",
       "     (\"plot.data.gene\", \"plot.data.cpd\"). Both elements are data.frame\n",
       "     or NULL depends on the corresponding input data gene.data and\n",
       "     cpd.data. These data.frames record the plot data for mapped gene\n",
       "     or compound nodes: rows are mapped genes/compounds, columns are:\n",
       "\n",
       "kegg.names: standard KEGG IDs/Names for mapped nodes. It's Entrez Gene\n",
       "          ID or KEGG Compound Accessions.\n",
       "\n",
       "  labels: Node labels to be used when needed.\n",
       "\n",
       "all.mapped: All molecule (gene or compound) IDs mapped to this node.\n",
       "\n",
       "    type: node type, currently 4 types are supported: \"gene\",\"enzyme\",\n",
       "          \"compound\" and \"ortholog\".\n",
       "\n",
       "       x: x coordinate in the original KEGG pathway graph.\n",
       "\n",
       "       y: y coordinate in the original KEGG pathway graph.\n",
       "\n",
       "   width: node width in the original KEGG pathway graph.\n",
       "\n",
       "  height: node height in the original KEGG pathway graph.\n",
       "\n",
       "other columns : columns of the mapped gene/compound data and\n",
       "          corresponding pseudo-color codes for individual samples\n",
       "     The results returned by ‘keggview.native’ and codekeggview.graph\n",
       "     are both a list of graph plotting parameters. These are not\n",
       "     intended to be used externally.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Weijun Luo <luo_weijun@yahoo.com>\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     Luo, W. and Brouwer, C., Pathview: an R/Bioconductor package for\n",
       "     pathway based data integration and visualization. Bioinformatics,\n",
       "     2013, 29(14): 1830-1831, doi: 10.1093/bioinformatics/btt285\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘download.kegg’ the downloader, ‘node.info’ the parser, ‘node.map’\n",
       "     and ‘node.color’ the mapper.\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     #load data\n",
       "     data(gse16873.d)\n",
       "     data(demo.paths)\n",
       "     \n",
       "     #KEGG view: gene data only\n",
       "     i <- 1\n",
       "     pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "     demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "     kegg.native = TRUE)\n",
       "     str(pv.out)\n",
       "     head(pv.out$plot.data.gene)\n",
       "     #result PNG file in current directory\n",
       "     \n",
       "     #Graphviz view: gene data only\n",
       "     pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "     demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "     kegg.native = FALSE, sign.pos = demo.paths$spos[i])\n",
       "     #result PDF file in current directory\n",
       "     \n",
       "     #KEGG view: both gene and compound data\n",
       "     sim.cpd.data=sim.mol.data(mol.type=\"cpd\", nmol=3000)\n",
       "     i <- 3\n",
       "     print(demo.paths$sel.paths[i])\n",
       "     pv.out <- pathview(gene.data = gse16873.d[, 1], cpd.data = sim.cpd.data,\n",
       "     pathway.id = demo.paths$sel.paths[i], species = \"hsa\", out.suffix =\n",
       "     \"gse16873.cpd\", keys.align = \"y\", kegg.native = TRUE, key.pos = demo.paths$kpos1[i])\n",
       "     str(pv.out)\n",
       "     head(pv.out$plot.data.cpd)\n",
       "     \n",
       "     #multiple states in one graph\n",
       "     set.seed(10)\n",
       "     sim.cpd.data2 = matrix(sample(sim.cpd.data, 18000, \n",
       "         replace = TRUE), ncol = 6)\n",
       "     pv.out <- pathview(gene.data = gse16873.d[, 1:3], \n",
       "         cpd.data = sim.cpd.data2[, 1:2], pathway.id = demo.paths$sel.paths[i], \n",
       "         species = \"hsa\", out.suffix = \"gse16873.cpd.3-2s\", keys.align = \"y\", \n",
       "         kegg.native = TRUE, match.data = FALSE, multi.state = TRUE, same.layer = TRUE)\n",
       "     str(pv.out)\n",
       "     head(pv.out$plot.data.cpd)\n",
       "     \n",
       "     #result PNG file in current directory\n",
       "     \n",
       "     ##more examples of pathview usages are shown in the vignette.\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?pathview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
