{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `pathview` Tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load library and example datasets\n",
    "library(\"pacman\")\n",
    "\n",
    "pacman::p_load(\"pathview\", \"gage\", \"tidyverse\")\n",
    "data(gse16873.d)\n",
    "# Load human pathways data\n",
    "data(paths.hsa)\n",
    "# load demo pathway-related data, including 3 pathway ids and related plotting params\n",
    "# this is in dictionary format\n",
    "data(demo.paths)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 11979 × 6 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>DCIS_1</th><th scope=col>DCIS_2</th><th scope=col>DCIS_3</th><th scope=col>DCIS_4</th><th scope=col>DCIS_5</th><th scope=col>DCIS_6</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>10000</th><td>-3.076448e-01</td><td>-1.472277e-01</td><td>-0.0237848080</td><td>-0.0705619313</td><td>-1.323087e-03</td><td>-0.1502681305</td></tr>\n",
       "\t<tr><th scope=row>10001</th><td> 4.158680e-01</td><td>-3.347726e-01</td><td>-0.5131369072</td><td>-0.1665371241</td><td> 1.111222e-01</td><td> 0.1340073370</td></tr>\n",
       "\t<tr><th scope=row>10002</th><td> 1.985493e-01</td><td> 3.789588e-02</td><td> 0.3418653408</td><td>-0.0852741998</td><td> 7.675593e-01</td><td> 0.1582860892</td></tr>\n",
       "\t<tr><th scope=row>10003</th><td>-2.315530e-01</td><td>-9.659311e-02</td><td>-0.1047272830</td><td>-0.0480140390</td><td>-2.080564e-01</td><td> 0.0334444824</td></tr>\n",
       "\t<tr><th scope=row>100048912</th><td>-4.490724e-02</td><td>-5.203146e-02</td><td> 0.0363903758</td><td> 0.0480782310</td><td> 2.720582e-02</td><td> 0.0544473911</td></tr>\n",
       "\t<tr><th scope=row>10004</th><td>-8.756237e-02</td><td>-5.027725e-02</td><td> 0.0018211331</td><td> 0.0302383471</td><td> 8.034394e-03</td><td>-0.0686074928</td></tr>\n",
       "\t<tr><th scope=row>10005</th><td>-1.262668e-01</td><td> 4.777874e-01</td><td>-0.1061224804</td><td> 0.2346640653</td><td> 9.676428e-02</td><td> 0.0635200439</td></tr>\n",
       "\t<tr><th scope=row>10006</th><td> 6.502821e-01</td><td> 1.950600e-01</td><td>-0.0053586469</td><td>-0.2263254562</td><td> 8.259053e-02</td><td>-0.0694435042</td></tr>\n",
       "\t<tr><th scope=row>10007</th><td> 1.566912e-01</td><td> 5.535167e-02</td><td> 0.0988623520</td><td>-0.0257075972</td><td> 2.126088e-01</td><td> 0.1593592082</td></tr>\n",
       "\t<tr><th scope=row>10009</th><td>-1.929644e-02</td><td> 5.141373e-03</td><td>-0.0196603248</td><td> 0.0058828281</td><td>-2.317358e-02</td><td>-0.0002808255</td></tr>\n",
       "\t<tr><th scope=row>1000</th><td>-7.689655e-03</td><td> 1.721621e-01</td><td>-0.1988314299</td><td>-0.2019850887</td><td>-7.995476e-02</td><td>-0.1728911097</td></tr>\n",
       "\t<tr><th scope=row>10010</th><td>-3.509999e-01</td><td>-3.596717e-01</td><td>-0.0540199940</td><td> 0.1954106017</td><td>-2.503424e-01</td><td>-0.1227903681</td></tr>\n",
       "\t<tr><th scope=row>100127886</th><td>-1.853968e-01</td><td>-1.012743e-02</td><td> 0.0557271185</td><td> 0.0028300388</td><td> 2.596228e-02</td><td> 0.0388121425</td></tr>\n",
       "\t<tr><th scope=row>100127972</th><td>-4.943871e-03</td><td>-2.280742e-03</td><td> 0.0031634175</td><td> 0.0035797150</td><td> 2.021840e-03</td><td>-0.0004505923</td></tr>\n",
       "\t<tr><th scope=row>100128008</th><td>-1.891914e-01</td><td> 4.913371e-02</td><td> 0.0002729363</td><td> 0.1219493187</td><td>-7.771520e-02</td><td>-0.1250310868</td></tr>\n",
       "\t<tr><th scope=row>100128062</th><td> 1.039215e+00</td><td> 3.279636e-02</td><td>-0.3429664266</td><td>-0.2767989481</td><td> 2.563069e-01</td><td> 0.4055939319</td></tr>\n",
       "\t<tr><th scope=row>100128414</th><td>-2.041448e-04</td><td> 1.271084e-05</td><td>-0.0000318674</td><td>-0.0000807691</td><td>-4.547224e-06</td><td>-0.0001340409</td></tr>\n",
       "\t<tr><th scope=row>100128919</th><td>-1.770150e-01</td><td>-6.929799e-02</td><td>-0.0048792855</td><td> 0.1586620449</td><td> 2.715102e-02</td><td>-0.0673444281</td></tr>\n",
       "\t<tr><th scope=row>100129015</th><td>-1.193076e-02</td><td>-9.064290e-05</td><td> 0.0015327461</td><td> 0.0061492361</td><td>-7.184290e-03</td><td>-0.0079698509</td></tr>\n",
       "\t<tr><th scope=row>100129128</th><td> 1.676678e-01</td><td>-5.315995e-02</td><td> 0.0065112112</td><td>-0.0759282591</td><td>-8.465508e-02</td><td>-0.0691177169</td></tr>\n",
       "\t<tr><th scope=row>100129250</th><td>-1.988180e-01</td><td> 5.939458e-02</td><td>-0.0472724175</td><td>-0.0776929725</td><td> 3.199363e-03</td><td>-0.0475972848</td></tr>\n",
       "\t<tr><th scope=row>100129271</th><td>-1.989387e-01</td><td> 1.851973e-02</td><td> 0.0038471059</td><td>-0.1748526017</td><td>-9.932469e-02</td><td>-0.0650394320</td></tr>\n",
       "\t<tr><th scope=row>100129361</th><td> 7.054649e-01</td><td> 3.681755e-01</td><td>-0.2185879706</td><td> 0.1598348804</td><td> 5.245252e-01</td><td>-0.0326498764</td></tr>\n",
       "\t<tr><th scope=row>100129410</th><td>-4.663317e-05</td><td>-3.954421e-05</td><td> 0.0003432414</td><td>-0.0000383419</td><td>-9.915890e-05</td><td>-0.0002412009</td></tr>\n",
       "\t<tr><th scope=row>100129482</th><td>-1.470412e-01</td><td>-1.576722e-01</td><td>-0.0694889809</td><td>-0.1512772982</td><td> 6.246949e-02</td><td> 0.1403982254</td></tr>\n",
       "\t<tr><th scope=row>100129500</th><td>-1.359120e-01</td><td>-6.243703e-02</td><td> 0.1730479819</td><td>-0.1527967107</td><td> 6.093229e-03</td><td> 0.0202833171</td></tr>\n",
       "\t<tr><th scope=row>100129503</th><td>-1.937177e-01</td><td>-1.160323e-01</td><td> 0.1601165395</td><td>-0.0086225219</td><td>-3.612119e-02</td><td> 0.0807876962</td></tr>\n",
       "\t<tr><th scope=row>100129624</th><td>-4.758874e-02</td><td> 1.357629e-02</td><td> 0.0368826756</td><td>-0.0908665392</td><td>-9.909635e-03</td><td> 0.0307345168</td></tr>\n",
       "\t<tr><th scope=row>100129656</th><td>-9.127286e-02</td><td> 6.218638e-02</td><td> 0.0045374444</td><td>-0.0631818164</td><td>-1.648619e-02</td><td> 0.0232970879</td></tr>\n",
       "\t<tr><th scope=row>100129668</th><td>-4.400424e-02</td><td> 8.941932e-02</td><td> 0.0709809030</td><td>-0.4645392185</td><td>-2.182906e-01</td><td>-0.0004430268</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Bs-lys-3</th><td>-9.037905e-03</td><td>-1.022145e-02</td><td> 7.565813e-03</td><td> 2.677956e-04</td><td> 5.806199e-03</td><td>-6.632238e-03</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Bs-lys-5</th><td>-2.950573e-02</td><td>-8.525517e-03</td><td>-3.494709e-03</td><td>-2.676942e-03</td><td> 1.608129e-02</td><td> 1.863921e-02</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Bs-lys-M</th><td>-8.573824e-03</td><td> 1.237125e-03</td><td> 1.022884e-02</td><td> 1.290452e-02</td><td>-1.985604e-03</td><td>-7.927933e-03</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Bs-phe-3</th><td> 3.900938e-04</td><td> 1.234103e-02</td><td> 1.145784e-02</td><td> 6.315344e-04</td><td>-1.903742e-04</td><td> 8.568381e-03</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Bs-phe-5</th><td>-4.505562e-02</td><td>-6.628921e-03</td><td> 1.155014e-02</td><td> 5.759616e-05</td><td> 2.513405e-02</td><td> 8.947731e-03</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Bs-phe-M</th><td>-1.036767e-01</td><td> 1.206104e-02</td><td> 3.633987e-02</td><td> 7.542170e-04</td><td> 5.619506e-02</td><td>-3.130054e-02</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Bs-thr-3_s</th><td>-4.972090e-03</td><td>-1.215819e-02</td><td>-1.200407e-02</td><td> 2.271640e-02</td><td> 7.068458e-02</td><td> 2.222268e-02</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Bs-thr-5_s</th><td>-1.533595e-02</td><td> 7.836306e-04</td><td> 6.008415e-02</td><td>-4.907319e-02</td><td>-3.792855e-02</td><td>-1.723455e-02</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Bs-thr-M_s</th><td>-1.412013e-01</td><td> 5.227123e-02</td><td> 1.396457e-01</td><td> 1.447931e-01</td><td>-7.243298e-02</td><td> 6.973207e-02</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Ec-bioB-3</th><td>-3.328436e-01</td><td> 9.237605e-02</td><td> 6.488547e-01</td><td>-1.402720e-01</td><td> 1.716021e-01</td><td>-3.793185e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Ec-bioB-5</th><td>-3.761715e-01</td><td> 6.691319e-02</td><td> 5.742376e-01</td><td>-6.975139e-03</td><td> 2.262841e-02</td><td>-4.970149e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Ec-bioB-M</th><td>-3.815584e-01</td><td> 2.899714e-01</td><td> 4.730499e-01</td><td>-9.867918e-02</td><td> 1.304209e-01</td><td>-3.593188e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Ec-bioC-3</th><td>-6.218098e-01</td><td> 5.902343e-02</td><td> 3.819226e-01</td><td>-3.129861e-02</td><td>-3.656372e-02</td><td>-3.152232e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Ec-bioC-5</th><td>-5.861536e-01</td><td> 1.490416e-01</td><td> 5.545161e-01</td><td>-1.534065e-01</td><td>-8.040619e-02</td><td>-3.312578e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Ec-bioD-3</th><td>-6.975061e-01</td><td> 6.147410e-02</td><td> 5.056704e-01</td><td>-1.811558e-01</td><td> 1.491453e-02</td><td>-5.085829e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Ec-bioD-5</th><td>-7.249099e-01</td><td> 1.868671e-01</td><td> 5.065900e-01</td><td>-1.608034e-01</td><td> 4.642487e-02</td><td>-3.912962e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Hs18SrRNA-3_s</th><td>-8.098065e-01</td><td>-3.456874e-02</td><td> 8.788347e-03</td><td>-1.123250e-01</td><td>-9.673530e-01</td><td>-2.039927e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Hs18SrRNA-5</th><td>-3.238156e-01</td><td> 6.472983e-01</td><td>-1.229498e-01</td><td> 1.907948e+00</td><td>-2.832521e-01</td><td> 2.156635e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Hs18SrRNA-M_x</th><td>-7.225679e-01</td><td> 4.694751e-02</td><td>-4.347568e-01</td><td> 7.607337e-01</td><td>-1.264574e+00</td><td> 1.949101e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Hs28SrRNA-3</th><td>-4.957211e-02</td><td>-4.684183e-01</td><td> 1.802405e-02</td><td> 3.051821e-01</td><td>-1.338035e-01</td><td> 7.951062e-02</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Hs28SrRNA-5</th><td>-4.142774e-01</td><td> 1.144169e-01</td><td>-5.898014e-02</td><td> 2.718412e-01</td><td>-1.214159e-02</td><td> 1.018798e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-Hs28SrRNA-M</th><td>-2.261485e-02</td><td>-2.155457e-02</td><td> 1.185648e-01</td><td>-2.712507e-02</td><td>-1.269695e-01</td><td> 9.544529e-02</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-P1-cre-3</th><td>-1.996843e-01</td><td>-2.772621e-03</td><td> 1.539262e-01</td><td>-1.169188e-01</td><td> 1.354375e-01</td><td>-1.755887e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-r2-P1-cre-5</th><td>-3.456983e-01</td><td> 7.307571e-02</td><td> 1.600307e-01</td><td>-4.319175e-01</td><td> 9.175283e-02</td><td>-4.990513e-01</td></tr>\n",
       "\t<tr><th scope=row>AFFX-ThrX-3</th><td>-5.663154e-03</td><td> 1.716721e-03</td><td> 8.447181e-03</td><td>-7.333537e-05</td><td> 4.770441e-03</td><td> 4.423160e-03</td></tr>\n",
       "\t<tr><th scope=row>AFFX-ThrX-5</th><td> 4.645983e-03</td><td>-3.107411e-02</td><td> 7.821647e-02</td><td>-1.565735e-02</td><td>-4.109157e-02</td><td> 7.462161e-03</td></tr>\n",
       "\t<tr><th scope=row>AFFX-ThrX-M</th><td>-8.246908e-04</td><td> 1.034649e-03</td><td>-1.682287e-03</td><td> 3.163011e-04</td><td>-4.460928e-04</td><td>-1.690070e-04</td></tr>\n",
       "\t<tr><th scope=row>AFFX-TrpnX-3</th><td>-2.300715e-05</td><td> 5.299686e-07</td><td> 1.554954e-05</td><td>-5.973189e-05</td><td> 3.760379e-05</td><td>-5.836678e-05</td></tr>\n",
       "\t<tr><th scope=row>AFFX-TrpnX-5</th><td>-2.931525e-02</td><td>-2.442613e-02</td><td> 6.048274e-03</td><td>-9.057592e-02</td><td>-2.445375e-03</td><td> 2.413393e-02</td></tr>\n",
       "\t<tr><th scope=row>AFFX-TrpnX-M</th><td>-1.411722e-02</td><td>-1.370392e-02</td><td> 5.094004e-02</td><td>-6.325498e-03</td><td>-6.110443e-04</td><td>-2.701850e-02</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 11979 × 6 of type dbl\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & DCIS\\_1 & DCIS\\_2 & DCIS\\_3 & DCIS\\_4 & DCIS\\_5 & DCIS\\_6\\\\\n",
       "\\hline\n",
       "\t10000 & -3.076448e-01 & -1.472277e-01 & -0.0237848080 & -0.0705619313 & -1.323087e-03 & -0.1502681305\\\\\n",
       "\t10001 &  4.158680e-01 & -3.347726e-01 & -0.5131369072 & -0.1665371241 &  1.111222e-01 &  0.1340073370\\\\\n",
       "\t10002 &  1.985493e-01 &  3.789588e-02 &  0.3418653408 & -0.0852741998 &  7.675593e-01 &  0.1582860892\\\\\n",
       "\t10003 & -2.315530e-01 & -9.659311e-02 & -0.1047272830 & -0.0480140390 & -2.080564e-01 &  0.0334444824\\\\\n",
       "\t100048912 & -4.490724e-02 & -5.203146e-02 &  0.0363903758 &  0.0480782310 &  2.720582e-02 &  0.0544473911\\\\\n",
       "\t10004 & -8.756237e-02 & -5.027725e-02 &  0.0018211331 &  0.0302383471 &  8.034394e-03 & -0.0686074928\\\\\n",
       "\t10005 & -1.262668e-01 &  4.777874e-01 & -0.1061224804 &  0.2346640653 &  9.676428e-02 &  0.0635200439\\\\\n",
       "\t10006 &  6.502821e-01 &  1.950600e-01 & -0.0053586469 & -0.2263254562 &  8.259053e-02 & -0.0694435042\\\\\n",
       "\t10007 &  1.566912e-01 &  5.535167e-02 &  0.0988623520 & -0.0257075972 &  2.126088e-01 &  0.1593592082\\\\\n",
       "\t10009 & -1.929644e-02 &  5.141373e-03 & -0.0196603248 &  0.0058828281 & -2.317358e-02 & -0.0002808255\\\\\n",
       "\t1000 & -7.689655e-03 &  1.721621e-01 & -0.1988314299 & -0.2019850887 & -7.995476e-02 & -0.1728911097\\\\\n",
       "\t10010 & -3.509999e-01 & -3.596717e-01 & -0.0540199940 &  0.1954106017 & -2.503424e-01 & -0.1227903681\\\\\n",
       "\t100127886 & -1.853968e-01 & -1.012743e-02 &  0.0557271185 &  0.0028300388 &  2.596228e-02 &  0.0388121425\\\\\n",
       "\t100127972 & -4.943871e-03 & -2.280742e-03 &  0.0031634175 &  0.0035797150 &  2.021840e-03 & -0.0004505923\\\\\n",
       "\t100128008 & -1.891914e-01 &  4.913371e-02 &  0.0002729363 &  0.1219493187 & -7.771520e-02 & -0.1250310868\\\\\n",
       "\t100128062 &  1.039215e+00 &  3.279636e-02 & -0.3429664266 & -0.2767989481 &  2.563069e-01 &  0.4055939319\\\\\n",
       "\t100128414 & -2.041448e-04 &  1.271084e-05 & -0.0000318674 & -0.0000807691 & -4.547224e-06 & -0.0001340409\\\\\n",
       "\t100128919 & -1.770150e-01 & -6.929799e-02 & -0.0048792855 &  0.1586620449 &  2.715102e-02 & -0.0673444281\\\\\n",
       "\t100129015 & -1.193076e-02 & -9.064290e-05 &  0.0015327461 &  0.0061492361 & -7.184290e-03 & -0.0079698509\\\\\n",
       "\t100129128 &  1.676678e-01 & -5.315995e-02 &  0.0065112112 & -0.0759282591 & -8.465508e-02 & -0.0691177169\\\\\n",
       "\t100129250 & -1.988180e-01 &  5.939458e-02 & -0.0472724175 & -0.0776929725 &  3.199363e-03 & -0.0475972848\\\\\n",
       "\t100129271 & -1.989387e-01 &  1.851973e-02 &  0.0038471059 & -0.1748526017 & -9.932469e-02 & -0.0650394320\\\\\n",
       "\t100129361 &  7.054649e-01 &  3.681755e-01 & -0.2185879706 &  0.1598348804 &  5.245252e-01 & -0.0326498764\\\\\n",
       "\t100129410 & -4.663317e-05 & -3.954421e-05 &  0.0003432414 & -0.0000383419 & -9.915890e-05 & -0.0002412009\\\\\n",
       "\t100129482 & -1.470412e-01 & -1.576722e-01 & -0.0694889809 & -0.1512772982 &  6.246949e-02 &  0.1403982254\\\\\n",
       "\t100129500 & -1.359120e-01 & -6.243703e-02 &  0.1730479819 & -0.1527967107 &  6.093229e-03 &  0.0202833171\\\\\n",
       "\t100129503 & -1.937177e-01 & -1.160323e-01 &  0.1601165395 & -0.0086225219 & -3.612119e-02 &  0.0807876962\\\\\n",
       "\t100129624 & -4.758874e-02 &  1.357629e-02 &  0.0368826756 & -0.0908665392 & -9.909635e-03 &  0.0307345168\\\\\n",
       "\t100129656 & -9.127286e-02 &  6.218638e-02 &  0.0045374444 & -0.0631818164 & -1.648619e-02 &  0.0232970879\\\\\n",
       "\t100129668 & -4.400424e-02 &  8.941932e-02 &  0.0709809030 & -0.4645392185 & -2.182906e-01 & -0.0004430268\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\tAFFX-r2-Bs-lys-3 & -9.037905e-03 & -1.022145e-02 &  7.565813e-03 &  2.677956e-04 &  5.806199e-03 & -6.632238e-03\\\\\n",
       "\tAFFX-r2-Bs-lys-5 & -2.950573e-02 & -8.525517e-03 & -3.494709e-03 & -2.676942e-03 &  1.608129e-02 &  1.863921e-02\\\\\n",
       "\tAFFX-r2-Bs-lys-M & -8.573824e-03 &  1.237125e-03 &  1.022884e-02 &  1.290452e-02 & -1.985604e-03 & -7.927933e-03\\\\\n",
       "\tAFFX-r2-Bs-phe-3 &  3.900938e-04 &  1.234103e-02 &  1.145784e-02 &  6.315344e-04 & -1.903742e-04 &  8.568381e-03\\\\\n",
       "\tAFFX-r2-Bs-phe-5 & -4.505562e-02 & -6.628921e-03 &  1.155014e-02 &  5.759616e-05 &  2.513405e-02 &  8.947731e-03\\\\\n",
       "\tAFFX-r2-Bs-phe-M & -1.036767e-01 &  1.206104e-02 &  3.633987e-02 &  7.542170e-04 &  5.619506e-02 & -3.130054e-02\\\\\n",
       "\tAFFX-r2-Bs-thr-3\\_s & -4.972090e-03 & -1.215819e-02 & -1.200407e-02 &  2.271640e-02 &  7.068458e-02 &  2.222268e-02\\\\\n",
       "\tAFFX-r2-Bs-thr-5\\_s & -1.533595e-02 &  7.836306e-04 &  6.008415e-02 & -4.907319e-02 & -3.792855e-02 & -1.723455e-02\\\\\n",
       "\tAFFX-r2-Bs-thr-M\\_s & -1.412013e-01 &  5.227123e-02 &  1.396457e-01 &  1.447931e-01 & -7.243298e-02 &  6.973207e-02\\\\\n",
       "\tAFFX-r2-Ec-bioB-3 & -3.328436e-01 &  9.237605e-02 &  6.488547e-01 & -1.402720e-01 &  1.716021e-01 & -3.793185e-01\\\\\n",
       "\tAFFX-r2-Ec-bioB-5 & -3.761715e-01 &  6.691319e-02 &  5.742376e-01 & -6.975139e-03 &  2.262841e-02 & -4.970149e-01\\\\\n",
       "\tAFFX-r2-Ec-bioB-M & -3.815584e-01 &  2.899714e-01 &  4.730499e-01 & -9.867918e-02 &  1.304209e-01 & -3.593188e-01\\\\\n",
       "\tAFFX-r2-Ec-bioC-3 & -6.218098e-01 &  5.902343e-02 &  3.819226e-01 & -3.129861e-02 & -3.656372e-02 & -3.152232e-01\\\\\n",
       "\tAFFX-r2-Ec-bioC-5 & -5.861536e-01 &  1.490416e-01 &  5.545161e-01 & -1.534065e-01 & -8.040619e-02 & -3.312578e-01\\\\\n",
       "\tAFFX-r2-Ec-bioD-3 & -6.975061e-01 &  6.147410e-02 &  5.056704e-01 & -1.811558e-01 &  1.491453e-02 & -5.085829e-01\\\\\n",
       "\tAFFX-r2-Ec-bioD-5 & -7.249099e-01 &  1.868671e-01 &  5.065900e-01 & -1.608034e-01 &  4.642487e-02 & -3.912962e-01\\\\\n",
       "\tAFFX-r2-Hs18SrRNA-3\\_s & -8.098065e-01 & -3.456874e-02 &  8.788347e-03 & -1.123250e-01 & -9.673530e-01 & -2.039927e-01\\\\\n",
       "\tAFFX-r2-Hs18SrRNA-5 & -3.238156e-01 &  6.472983e-01 & -1.229498e-01 &  1.907948e+00 & -2.832521e-01 &  2.156635e-01\\\\\n",
       "\tAFFX-r2-Hs18SrRNA-M\\_x & -7.225679e-01 &  4.694751e-02 & -4.347568e-01 &  7.607337e-01 & -1.264574e+00 &  1.949101e-01\\\\\n",
       "\tAFFX-r2-Hs28SrRNA-3 & -4.957211e-02 & -4.684183e-01 &  1.802405e-02 &  3.051821e-01 & -1.338035e-01 &  7.951062e-02\\\\\n",
       "\tAFFX-r2-Hs28SrRNA-5 & -4.142774e-01 &  1.144169e-01 & -5.898014e-02 &  2.718412e-01 & -1.214159e-02 &  1.018798e-01\\\\\n",
       "\tAFFX-r2-Hs28SrRNA-M & -2.261485e-02 & -2.155457e-02 &  1.185648e-01 & -2.712507e-02 & -1.269695e-01 &  9.544529e-02\\\\\n",
       "\tAFFX-r2-P1-cre-3 & -1.996843e-01 & -2.772621e-03 &  1.539262e-01 & -1.169188e-01 &  1.354375e-01 & -1.755887e-01\\\\\n",
       "\tAFFX-r2-P1-cre-5 & -3.456983e-01 &  7.307571e-02 &  1.600307e-01 & -4.319175e-01 &  9.175283e-02 & -4.990513e-01\\\\\n",
       "\tAFFX-ThrX-3 & -5.663154e-03 &  1.716721e-03 &  8.447181e-03 & -7.333537e-05 &  4.770441e-03 &  4.423160e-03\\\\\n",
       "\tAFFX-ThrX-5 &  4.645983e-03 & -3.107411e-02 &  7.821647e-02 & -1.565735e-02 & -4.109157e-02 &  7.462161e-03\\\\\n",
       "\tAFFX-ThrX-M & -8.246908e-04 &  1.034649e-03 & -1.682287e-03 &  3.163011e-04 & -4.460928e-04 & -1.690070e-04\\\\\n",
       "\tAFFX-TrpnX-3 & -2.300715e-05 &  5.299686e-07 &  1.554954e-05 & -5.973189e-05 &  3.760379e-05 & -5.836678e-05\\\\\n",
       "\tAFFX-TrpnX-5 & -2.931525e-02 & -2.442613e-02 &  6.048274e-03 & -9.057592e-02 & -2.445375e-03 &  2.413393e-02\\\\\n",
       "\tAFFX-TrpnX-M & -1.411722e-02 & -1.370392e-02 &  5.094004e-02 & -6.325498e-03 & -6.110443e-04 & -2.701850e-02\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 11979 × 6 of type dbl\n",
       "\n",
       "| <!--/--> | DCIS_1 | DCIS_2 | DCIS_3 | DCIS_4 | DCIS_5 | DCIS_6 |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 10000 | -3.076448e-01 | -1.472277e-01 | -0.0237848080 | -0.0705619313 | -1.323087e-03 | -0.1502681305 |\n",
       "| 10001 |  4.158680e-01 | -3.347726e-01 | -0.5131369072 | -0.1665371241 |  1.111222e-01 |  0.1340073370 |\n",
       "| 10002 |  1.985493e-01 |  3.789588e-02 |  0.3418653408 | -0.0852741998 |  7.675593e-01 |  0.1582860892 |\n",
       "| 10003 | -2.315530e-01 | -9.659311e-02 | -0.1047272830 | -0.0480140390 | -2.080564e-01 |  0.0334444824 |\n",
       "| 100048912 | -4.490724e-02 | -5.203146e-02 |  0.0363903758 |  0.0480782310 |  2.720582e-02 |  0.0544473911 |\n",
       "| 10004 | -8.756237e-02 | -5.027725e-02 |  0.0018211331 |  0.0302383471 |  8.034394e-03 | -0.0686074928 |\n",
       "| 10005 | -1.262668e-01 |  4.777874e-01 | -0.1061224804 |  0.2346640653 |  9.676428e-02 |  0.0635200439 |\n",
       "| 10006 |  6.502821e-01 |  1.950600e-01 | -0.0053586469 | -0.2263254562 |  8.259053e-02 | -0.0694435042 |\n",
       "| 10007 |  1.566912e-01 |  5.535167e-02 |  0.0988623520 | -0.0257075972 |  2.126088e-01 |  0.1593592082 |\n",
       "| 10009 | -1.929644e-02 |  5.141373e-03 | -0.0196603248 |  0.0058828281 | -2.317358e-02 | -0.0002808255 |\n",
       "| 1000 | -7.689655e-03 |  1.721621e-01 | -0.1988314299 | -0.2019850887 | -7.995476e-02 | -0.1728911097 |\n",
       "| 10010 | -3.509999e-01 | -3.596717e-01 | -0.0540199940 |  0.1954106017 | -2.503424e-01 | -0.1227903681 |\n",
       "| 100127886 | -1.853968e-01 | -1.012743e-02 |  0.0557271185 |  0.0028300388 |  2.596228e-02 |  0.0388121425 |\n",
       "| 100127972 | -4.943871e-03 | -2.280742e-03 |  0.0031634175 |  0.0035797150 |  2.021840e-03 | -0.0004505923 |\n",
       "| 100128008 | -1.891914e-01 |  4.913371e-02 |  0.0002729363 |  0.1219493187 | -7.771520e-02 | -0.1250310868 |\n",
       "| 100128062 |  1.039215e+00 |  3.279636e-02 | -0.3429664266 | -0.2767989481 |  2.563069e-01 |  0.4055939319 |\n",
       "| 100128414 | -2.041448e-04 |  1.271084e-05 | -0.0000318674 | -0.0000807691 | -4.547224e-06 | -0.0001340409 |\n",
       "| 100128919 | -1.770150e-01 | -6.929799e-02 | -0.0048792855 |  0.1586620449 |  2.715102e-02 | -0.0673444281 |\n",
       "| 100129015 | -1.193076e-02 | -9.064290e-05 |  0.0015327461 |  0.0061492361 | -7.184290e-03 | -0.0079698509 |\n",
       "| 100129128 |  1.676678e-01 | -5.315995e-02 |  0.0065112112 | -0.0759282591 | -8.465508e-02 | -0.0691177169 |\n",
       "| 100129250 | -1.988180e-01 |  5.939458e-02 | -0.0472724175 | -0.0776929725 |  3.199363e-03 | -0.0475972848 |\n",
       "| 100129271 | -1.989387e-01 |  1.851973e-02 |  0.0038471059 | -0.1748526017 | -9.932469e-02 | -0.0650394320 |\n",
       "| 100129361 |  7.054649e-01 |  3.681755e-01 | -0.2185879706 |  0.1598348804 |  5.245252e-01 | -0.0326498764 |\n",
       "| 100129410 | -4.663317e-05 | -3.954421e-05 |  0.0003432414 | -0.0000383419 | -9.915890e-05 | -0.0002412009 |\n",
       "| 100129482 | -1.470412e-01 | -1.576722e-01 | -0.0694889809 | -0.1512772982 |  6.246949e-02 |  0.1403982254 |\n",
       "| 100129500 | -1.359120e-01 | -6.243703e-02 |  0.1730479819 | -0.1527967107 |  6.093229e-03 |  0.0202833171 |\n",
       "| 100129503 | -1.937177e-01 | -1.160323e-01 |  0.1601165395 | -0.0086225219 | -3.612119e-02 |  0.0807876962 |\n",
       "| 100129624 | -4.758874e-02 |  1.357629e-02 |  0.0368826756 | -0.0908665392 | -9.909635e-03 |  0.0307345168 |\n",
       "| 100129656 | -9.127286e-02 |  6.218638e-02 |  0.0045374444 | -0.0631818164 | -1.648619e-02 |  0.0232970879 |\n",
       "| 100129668 | -4.400424e-02 |  8.941932e-02 |  0.0709809030 | -0.4645392185 | -2.182906e-01 | -0.0004430268 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| AFFX-r2-Bs-lys-3 | -9.037905e-03 | -1.022145e-02 |  7.565813e-03 |  2.677956e-04 |  5.806199e-03 | -6.632238e-03 |\n",
       "| AFFX-r2-Bs-lys-5 | -2.950573e-02 | -8.525517e-03 | -3.494709e-03 | -2.676942e-03 |  1.608129e-02 |  1.863921e-02 |\n",
       "| AFFX-r2-Bs-lys-M | -8.573824e-03 |  1.237125e-03 |  1.022884e-02 |  1.290452e-02 | -1.985604e-03 | -7.927933e-03 |\n",
       "| AFFX-r2-Bs-phe-3 |  3.900938e-04 |  1.234103e-02 |  1.145784e-02 |  6.315344e-04 | -1.903742e-04 |  8.568381e-03 |\n",
       "| AFFX-r2-Bs-phe-5 | -4.505562e-02 | -6.628921e-03 |  1.155014e-02 |  5.759616e-05 |  2.513405e-02 |  8.947731e-03 |\n",
       "| AFFX-r2-Bs-phe-M | -1.036767e-01 |  1.206104e-02 |  3.633987e-02 |  7.542170e-04 |  5.619506e-02 | -3.130054e-02 |\n",
       "| AFFX-r2-Bs-thr-3_s | -4.972090e-03 | -1.215819e-02 | -1.200407e-02 |  2.271640e-02 |  7.068458e-02 |  2.222268e-02 |\n",
       "| AFFX-r2-Bs-thr-5_s | -1.533595e-02 |  7.836306e-04 |  6.008415e-02 | -4.907319e-02 | -3.792855e-02 | -1.723455e-02 |\n",
       "| AFFX-r2-Bs-thr-M_s | -1.412013e-01 |  5.227123e-02 |  1.396457e-01 |  1.447931e-01 | -7.243298e-02 |  6.973207e-02 |\n",
       "| AFFX-r2-Ec-bioB-3 | -3.328436e-01 |  9.237605e-02 |  6.488547e-01 | -1.402720e-01 |  1.716021e-01 | -3.793185e-01 |\n",
       "| AFFX-r2-Ec-bioB-5 | -3.761715e-01 |  6.691319e-02 |  5.742376e-01 | -6.975139e-03 |  2.262841e-02 | -4.970149e-01 |\n",
       "| AFFX-r2-Ec-bioB-M | -3.815584e-01 |  2.899714e-01 |  4.730499e-01 | -9.867918e-02 |  1.304209e-01 | -3.593188e-01 |\n",
       "| AFFX-r2-Ec-bioC-3 | -6.218098e-01 |  5.902343e-02 |  3.819226e-01 | -3.129861e-02 | -3.656372e-02 | -3.152232e-01 |\n",
       "| AFFX-r2-Ec-bioC-5 | -5.861536e-01 |  1.490416e-01 |  5.545161e-01 | -1.534065e-01 | -8.040619e-02 | -3.312578e-01 |\n",
       "| AFFX-r2-Ec-bioD-3 | -6.975061e-01 |  6.147410e-02 |  5.056704e-01 | -1.811558e-01 |  1.491453e-02 | -5.085829e-01 |\n",
       "| AFFX-r2-Ec-bioD-5 | -7.249099e-01 |  1.868671e-01 |  5.065900e-01 | -1.608034e-01 |  4.642487e-02 | -3.912962e-01 |\n",
       "| AFFX-r2-Hs18SrRNA-3_s | -8.098065e-01 | -3.456874e-02 |  8.788347e-03 | -1.123250e-01 | -9.673530e-01 | -2.039927e-01 |\n",
       "| AFFX-r2-Hs18SrRNA-5 | -3.238156e-01 |  6.472983e-01 | -1.229498e-01 |  1.907948e+00 | -2.832521e-01 |  2.156635e-01 |\n",
       "| AFFX-r2-Hs18SrRNA-M_x | -7.225679e-01 |  4.694751e-02 | -4.347568e-01 |  7.607337e-01 | -1.264574e+00 |  1.949101e-01 |\n",
       "| AFFX-r2-Hs28SrRNA-3 | -4.957211e-02 | -4.684183e-01 |  1.802405e-02 |  3.051821e-01 | -1.338035e-01 |  7.951062e-02 |\n",
       "| AFFX-r2-Hs28SrRNA-5 | -4.142774e-01 |  1.144169e-01 | -5.898014e-02 |  2.718412e-01 | -1.214159e-02 |  1.018798e-01 |\n",
       "| AFFX-r2-Hs28SrRNA-M | -2.261485e-02 | -2.155457e-02 |  1.185648e-01 | -2.712507e-02 | -1.269695e-01 |  9.544529e-02 |\n",
       "| AFFX-r2-P1-cre-3 | -1.996843e-01 | -2.772621e-03 |  1.539262e-01 | -1.169188e-01 |  1.354375e-01 | -1.755887e-01 |\n",
       "| AFFX-r2-P1-cre-5 | -3.456983e-01 |  7.307571e-02 |  1.600307e-01 | -4.319175e-01 |  9.175283e-02 | -4.990513e-01 |\n",
       "| AFFX-ThrX-3 | -5.663154e-03 |  1.716721e-03 |  8.447181e-03 | -7.333537e-05 |  4.770441e-03 |  4.423160e-03 |\n",
       "| AFFX-ThrX-5 |  4.645983e-03 | -3.107411e-02 |  7.821647e-02 | -1.565735e-02 | -4.109157e-02 |  7.462161e-03 |\n",
       "| AFFX-ThrX-M | -8.246908e-04 |  1.034649e-03 | -1.682287e-03 |  3.163011e-04 | -4.460928e-04 | -1.690070e-04 |\n",
       "| AFFX-TrpnX-3 | -2.300715e-05 |  5.299686e-07 |  1.554954e-05 | -5.973189e-05 |  3.760379e-05 | -5.836678e-05 |\n",
       "| AFFX-TrpnX-5 | -2.931525e-02 | -2.442613e-02 |  6.048274e-03 | -9.057592e-02 | -2.445375e-03 |  2.413393e-02 |\n",
       "| AFFX-TrpnX-M | -1.411722e-02 | -1.370392e-02 |  5.094004e-02 | -6.325498e-03 | -6.110443e-04 | -2.701850e-02 |\n",
       "\n"
      ],
      "text/plain": [
       "                      DCIS_1        DCIS_2        DCIS_3        DCIS_4       \n",
       "10000                 -3.076448e-01 -1.472277e-01 -0.0237848080 -0.0705619313\n",
       "10001                  4.158680e-01 -3.347726e-01 -0.5131369072 -0.1665371241\n",
       "10002                  1.985493e-01  3.789588e-02  0.3418653408 -0.0852741998\n",
       "10003                 -2.315530e-01 -9.659311e-02 -0.1047272830 -0.0480140390\n",
       "100048912             -4.490724e-02 -5.203146e-02  0.0363903758  0.0480782310\n",
       "10004                 -8.756237e-02 -5.027725e-02  0.0018211331  0.0302383471\n",
       "10005                 -1.262668e-01  4.777874e-01 -0.1061224804  0.2346640653\n",
       "10006                  6.502821e-01  1.950600e-01 -0.0053586469 -0.2263254562\n",
       "10007                  1.566912e-01  5.535167e-02  0.0988623520 -0.0257075972\n",
       "10009                 -1.929644e-02  5.141373e-03 -0.0196603248  0.0058828281\n",
       "1000                  -7.689655e-03  1.721621e-01 -0.1988314299 -0.2019850887\n",
       "10010                 -3.509999e-01 -3.596717e-01 -0.0540199940  0.1954106017\n",
       "100127886             -1.853968e-01 -1.012743e-02  0.0557271185  0.0028300388\n",
       "100127972             -4.943871e-03 -2.280742e-03  0.0031634175  0.0035797150\n",
       "100128008             -1.891914e-01  4.913371e-02  0.0002729363  0.1219493187\n",
       "100128062              1.039215e+00  3.279636e-02 -0.3429664266 -0.2767989481\n",
       "100128414             -2.041448e-04  1.271084e-05 -0.0000318674 -0.0000807691\n",
       "100128919             -1.770150e-01 -6.929799e-02 -0.0048792855  0.1586620449\n",
       "100129015             -1.193076e-02 -9.064290e-05  0.0015327461  0.0061492361\n",
       "100129128              1.676678e-01 -5.315995e-02  0.0065112112 -0.0759282591\n",
       "100129250             -1.988180e-01  5.939458e-02 -0.0472724175 -0.0776929725\n",
       "100129271             -1.989387e-01  1.851973e-02  0.0038471059 -0.1748526017\n",
       "100129361              7.054649e-01  3.681755e-01 -0.2185879706  0.1598348804\n",
       "100129410             -4.663317e-05 -3.954421e-05  0.0003432414 -0.0000383419\n",
       "100129482             -1.470412e-01 -1.576722e-01 -0.0694889809 -0.1512772982\n",
       "100129500             -1.359120e-01 -6.243703e-02  0.1730479819 -0.1527967107\n",
       "100129503             -1.937177e-01 -1.160323e-01  0.1601165395 -0.0086225219\n",
       "100129624             -4.758874e-02  1.357629e-02  0.0368826756 -0.0908665392\n",
       "100129656             -9.127286e-02  6.218638e-02  0.0045374444 -0.0631818164\n",
       "100129668             -4.400424e-02  8.941932e-02  0.0709809030 -0.4645392185\n",
       "⋮                     ⋮             ⋮             ⋮             ⋮            \n",
       "AFFX-r2-Bs-lys-3      -9.037905e-03 -1.022145e-02  7.565813e-03  2.677956e-04\n",
       "AFFX-r2-Bs-lys-5      -2.950573e-02 -8.525517e-03 -3.494709e-03 -2.676942e-03\n",
       "AFFX-r2-Bs-lys-M      -8.573824e-03  1.237125e-03  1.022884e-02  1.290452e-02\n",
       "AFFX-r2-Bs-phe-3       3.900938e-04  1.234103e-02  1.145784e-02  6.315344e-04\n",
       "AFFX-r2-Bs-phe-5      -4.505562e-02 -6.628921e-03  1.155014e-02  5.759616e-05\n",
       "AFFX-r2-Bs-phe-M      -1.036767e-01  1.206104e-02  3.633987e-02  7.542170e-04\n",
       "AFFX-r2-Bs-thr-3_s    -4.972090e-03 -1.215819e-02 -1.200407e-02  2.271640e-02\n",
       "AFFX-r2-Bs-thr-5_s    -1.533595e-02  7.836306e-04  6.008415e-02 -4.907319e-02\n",
       "AFFX-r2-Bs-thr-M_s    -1.412013e-01  5.227123e-02  1.396457e-01  1.447931e-01\n",
       "AFFX-r2-Ec-bioB-3     -3.328436e-01  9.237605e-02  6.488547e-01 -1.402720e-01\n",
       "AFFX-r2-Ec-bioB-5     -3.761715e-01  6.691319e-02  5.742376e-01 -6.975139e-03\n",
       "AFFX-r2-Ec-bioB-M     -3.815584e-01  2.899714e-01  4.730499e-01 -9.867918e-02\n",
       "AFFX-r2-Ec-bioC-3     -6.218098e-01  5.902343e-02  3.819226e-01 -3.129861e-02\n",
       "AFFX-r2-Ec-bioC-5     -5.861536e-01  1.490416e-01  5.545161e-01 -1.534065e-01\n",
       "AFFX-r2-Ec-bioD-3     -6.975061e-01  6.147410e-02  5.056704e-01 -1.811558e-01\n",
       "AFFX-r2-Ec-bioD-5     -7.249099e-01  1.868671e-01  5.065900e-01 -1.608034e-01\n",
       "AFFX-r2-Hs18SrRNA-3_s -8.098065e-01 -3.456874e-02  8.788347e-03 -1.123250e-01\n",
       "AFFX-r2-Hs18SrRNA-5   -3.238156e-01  6.472983e-01 -1.229498e-01  1.907948e+00\n",
       "AFFX-r2-Hs18SrRNA-M_x -7.225679e-01  4.694751e-02 -4.347568e-01  7.607337e-01\n",
       "AFFX-r2-Hs28SrRNA-3   -4.957211e-02 -4.684183e-01  1.802405e-02  3.051821e-01\n",
       "AFFX-r2-Hs28SrRNA-5   -4.142774e-01  1.144169e-01 -5.898014e-02  2.718412e-01\n",
       "AFFX-r2-Hs28SrRNA-M   -2.261485e-02 -2.155457e-02  1.185648e-01 -2.712507e-02\n",
       "AFFX-r2-P1-cre-3      -1.996843e-01 -2.772621e-03  1.539262e-01 -1.169188e-01\n",
       "AFFX-r2-P1-cre-5      -3.456983e-01  7.307571e-02  1.600307e-01 -4.319175e-01\n",
       "AFFX-ThrX-3           -5.663154e-03  1.716721e-03  8.447181e-03 -7.333537e-05\n",
       "AFFX-ThrX-5            4.645983e-03 -3.107411e-02  7.821647e-02 -1.565735e-02\n",
       "AFFX-ThrX-M           -8.246908e-04  1.034649e-03 -1.682287e-03  3.163011e-04\n",
       "AFFX-TrpnX-3          -2.300715e-05  5.299686e-07  1.554954e-05 -5.973189e-05\n",
       "AFFX-TrpnX-5          -2.931525e-02 -2.442613e-02  6.048274e-03 -9.057592e-02\n",
       "AFFX-TrpnX-M          -1.411722e-02 -1.370392e-02  5.094004e-02 -6.325498e-03\n",
       "                      DCIS_5        DCIS_6       \n",
       "10000                 -1.323087e-03 -0.1502681305\n",
       "10001                  1.111222e-01  0.1340073370\n",
       "10002                  7.675593e-01  0.1582860892\n",
       "10003                 -2.080564e-01  0.0334444824\n",
       "100048912              2.720582e-02  0.0544473911\n",
       "10004                  8.034394e-03 -0.0686074928\n",
       "10005                  9.676428e-02  0.0635200439\n",
       "10006                  8.259053e-02 -0.0694435042\n",
       "10007                  2.126088e-01  0.1593592082\n",
       "10009                 -2.317358e-02 -0.0002808255\n",
       "1000                  -7.995476e-02 -0.1728911097\n",
       "10010                 -2.503424e-01 -0.1227903681\n",
       "100127886              2.596228e-02  0.0388121425\n",
       "100127972              2.021840e-03 -0.0004505923\n",
       "100128008             -7.771520e-02 -0.1250310868\n",
       "100128062              2.563069e-01  0.4055939319\n",
       "100128414             -4.547224e-06 -0.0001340409\n",
       "100128919              2.715102e-02 -0.0673444281\n",
       "100129015             -7.184290e-03 -0.0079698509\n",
       "100129128             -8.465508e-02 -0.0691177169\n",
       "100129250              3.199363e-03 -0.0475972848\n",
       "100129271             -9.932469e-02 -0.0650394320\n",
       "100129361              5.245252e-01 -0.0326498764\n",
       "100129410             -9.915890e-05 -0.0002412009\n",
       "100129482              6.246949e-02  0.1403982254\n",
       "100129500              6.093229e-03  0.0202833171\n",
       "100129503             -3.612119e-02  0.0807876962\n",
       "100129624             -9.909635e-03  0.0307345168\n",
       "100129656             -1.648619e-02  0.0232970879\n",
       "100129668             -2.182906e-01 -0.0004430268\n",
       "⋮                     ⋮             ⋮            \n",
       "AFFX-r2-Bs-lys-3       5.806199e-03 -6.632238e-03\n",
       "AFFX-r2-Bs-lys-5       1.608129e-02  1.863921e-02\n",
       "AFFX-r2-Bs-lys-M      -1.985604e-03 -7.927933e-03\n",
       "AFFX-r2-Bs-phe-3      -1.903742e-04  8.568381e-03\n",
       "AFFX-r2-Bs-phe-5       2.513405e-02  8.947731e-03\n",
       "AFFX-r2-Bs-phe-M       5.619506e-02 -3.130054e-02\n",
       "AFFX-r2-Bs-thr-3_s     7.068458e-02  2.222268e-02\n",
       "AFFX-r2-Bs-thr-5_s    -3.792855e-02 -1.723455e-02\n",
       "AFFX-r2-Bs-thr-M_s    -7.243298e-02  6.973207e-02\n",
       "AFFX-r2-Ec-bioB-3      1.716021e-01 -3.793185e-01\n",
       "AFFX-r2-Ec-bioB-5      2.262841e-02 -4.970149e-01\n",
       "AFFX-r2-Ec-bioB-M      1.304209e-01 -3.593188e-01\n",
       "AFFX-r2-Ec-bioC-3     -3.656372e-02 -3.152232e-01\n",
       "AFFX-r2-Ec-bioC-5     -8.040619e-02 -3.312578e-01\n",
       "AFFX-r2-Ec-bioD-3      1.491453e-02 -5.085829e-01\n",
       "AFFX-r2-Ec-bioD-5      4.642487e-02 -3.912962e-01\n",
       "AFFX-r2-Hs18SrRNA-3_s -9.673530e-01 -2.039927e-01\n",
       "AFFX-r2-Hs18SrRNA-5   -2.832521e-01  2.156635e-01\n",
       "AFFX-r2-Hs18SrRNA-M_x -1.264574e+00  1.949101e-01\n",
       "AFFX-r2-Hs28SrRNA-3   -1.338035e-01  7.951062e-02\n",
       "AFFX-r2-Hs28SrRNA-5   -1.214159e-02  1.018798e-01\n",
       "AFFX-r2-Hs28SrRNA-M   -1.269695e-01  9.544529e-02\n",
       "AFFX-r2-P1-cre-3       1.354375e-01 -1.755887e-01\n",
       "AFFX-r2-P1-cre-5       9.175283e-02 -4.990513e-01\n",
       "AFFX-ThrX-3            4.770441e-03  4.423160e-03\n",
       "AFFX-ThrX-5           -4.109157e-02  7.462161e-03\n",
       "AFFX-ThrX-M           -4.460928e-04 -1.690070e-04\n",
       "AFFX-TrpnX-3           3.760379e-05 -5.836678e-05\n",
       "AFFX-TrpnX-5          -2.445375e-03  2.413393e-02\n",
       "AFFX-TrpnX-M          -6.110443e-04 -2.701850e-02"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gse16873.d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate viz for only 1 column\n",
    "# Generate a single image file\n",
    "pv.out <- pathview(gene.data = gse16873.d[, 1], \n",
    "                   pathway.id = \"04110\",\n",
    "                   species = \"hsa\", \n",
    "                   out.suffix = \"gse16873\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i <- 1\n",
    "pv.out <- pathview(gene.data = gse16873.d[, 1], \n",
    "                   pathway.id = demo.paths$sel.paths[i],\n",
    "                   species = \"hsa\", \n",
    "                   out.suffix = \"gse16873\",\n",
    "                   kegg.native = T)\n",
    "list.files(pattern=\"hsa04110\", full.names=T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "str(pv.out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "head(pv.out$plot.data.gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id = demo.paths$sel.paths[i],\n",
    "                   species = \"hsa\", out.suffix = \"gse16873.2layer\", kegg.native = T,\n",
    "                   same.layer = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Integrating Cpd and Gene Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compound and gene data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# simulate cpd data\n",
    "sim.cpd.data = sim.mol.data(mol.type=\"cpd\", nmol=3000)\n",
    "data(cpd.simtypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify which pathway to retrieve\n",
    "i <- 3\n",
    "print(demo.paths$sel.paths[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pv.out <- suppressWarnings(pathview(gene.data = gse16873.d[, 1], \n",
    "                   cpd.data = sim.cpd.data,\n",
    "                   pathway.id = demo.paths$sel.paths[i], \n",
    "                   species = \"hsa\", \n",
    "                   out.suffix = \"gse16873.cpd\",\n",
    "                   keys.align = \"y\", \n",
    "                   kegg.native = T, \n",
    "                   key.pos = demo.paths$kpos1[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "head(pv.out$plot.data.cpd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pv.out <- suppressWarnings(pathview(gene.data = gse16873.d[, 1], \n",
    "                   cpd.data = sim.cpd.data, \n",
    "                   pathway.id = demo.paths$sel.paths[i], \n",
    "                   species = \"hsa\", \n",
    "                   out.suffix = \"gse16873.cpd\",\n",
    "                   keys.align = \"y\", \n",
    "                   kegg.native = F, \n",
    "                   key.pos = demo.paths$kpos2[i],\n",
    "                   sign.pos = demo.paths$spos[i], \n",
    "                   cpd.lab.offset = demo.paths$offs[i]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Multiple states or samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# simulate compound data with multiple replicate samples\n",
    "set.seed(10)\n",
    "sim.cpd.data2 = matrix(sample(sim.cpd.data, 18000,\n",
    "                              replace = T), ncol = 6)\n",
    "rownames(sim.cpd.data2) = names(sim.cpd.data)\n",
    "colnames(sim.cpd.data2) = paste(\"exp\", 1:6, sep = \"\")\n",
    "head(sim.cpd.data2, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# KEGG view\n",
    "pv.out <- suppressWarnings(pathview(gene.data = gse16873.d[, 1:3],\n",
    "                                    cpd.data = sim.cpd.data2[, 1:2], \n",
    "                                    pathway.id = demo.paths$sel.paths[i],\n",
    "                                    species = \"hsa\", \n",
    "                                    out.suffix = \"gse16873.cpd.3-2s\", \n",
    "                                    keys.align = \"y\",\n",
    "                                    kegg.native = T, \n",
    "                                    match.data = F, \n",
    "                                    multi.state = T, \n",
    "                                    same.layer = T))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# KEGG view with data match\n",
    "pv.out <- suppressWarnings(pathview(gene.data = gse16873.d[, 1:3],\n",
    "                                    cpd.data = sim.cpd.data2[, 1:2], \n",
    "                                    pathway.id = demo.paths$sel.paths[i],\n",
    "                                    species = \"hsa\", \n",
    "                                    out.suffix = \"gse16873.cpd.3-2s.match\",\n",
    "                                    keys.align = \"y\", \n",
    "                                    kegg.native = T, \n",
    "                                    match.data = T, \n",
    "                                    multi.state = T,\n",
    "                                    same.layer = T))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for pathview {pathview}\"><tr><td>pathview {pathview}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>\n",
       "Pathway based data integration and visualization\n",
       "</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Pathview is a tool set for pathway based data integration and\n",
       "visualization. It maps and renders user data on relevant pathway\n",
       "graphs. All users need is to supply their gene or compound data and\n",
       "specify the target pathway. Pathview automatically downloads the pathway\n",
       "graph data, parses the data file, maps user data to the pathway, and\n",
       "render pathway graph with the mapped data.\n",
       "Pathview generates both native KEGG view and Graphviz views for\n",
       "pathways. keggview.native and keggview.graph are the two viewer\n",
       "functions, and pathview is the main function providing a unified\n",
       "interface to downloader, parser, mapper and viewer functions.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "pathview(gene.data = NULL, cpd.data = NULL, pathway.id,\n",
       "species = \"hsa\", kegg.dir = \".\", cpd.idtype = \"kegg\", gene.idtype =\n",
       "\"entrez\", gene.annotpkg = NULL, min.nnodes = 3, kegg.native = TRUE,\n",
       "map.null = TRUE, expand.node = FALSE, split.group = FALSE, map.symbol =\n",
       "TRUE, map.cpdname = TRUE, node.sum = \"sum\", discrete=list(gene=FALSE,\n",
       "cpd=FALSE), limit = list(gene = 1, cpd = 1), bins = list(gene = 10, cpd\n",
       "= 10), both.dirs = list(gene = T, cpd = T), trans.fun = list(gene =\n",
       "NULL, cpd = NULL), low = list(gene = \"green\", cpd = \"blue\"), mid =\n",
       "list(gene = \"gray\", cpd = \"gray\"), high = list(gene = \"red\", cpd =\n",
       "\"yellow\"), na.col = \"transparent\", ...)\n",
       "\n",
       "keggview.native(plot.data.gene = NULL, plot.data.cpd = NULL,\n",
       "cols.ts.gene = NULL, cols.ts.cpd = NULL, node.data, pathway.name,\n",
       "out.suffix = \"pathview\", kegg.dir = \".\", multi.state=TRUE, match.data =\n",
       "TRUE, same.layer = TRUE, res = 300, cex = 0.25, discrete =\n",
       "list(gene=FALSE, cpd=FALSE), limit= list(gene = 1, cpd = 1), bins =\n",
       "list(gene = 10, cpd = 10), both.dirs =list(gene = T, cpd = T), low =\n",
       "list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd =\n",
       "\"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "\"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "\"x\", key.pos = \"topright\", ...)\n",
       "\n",
       "keggview.graph(plot.data.gene = NULL, plot.data.cpd = NULL, cols.ts.gene\n",
       "= NULL, cols.ts.cpd = NULL, node.data, path.graph, pathway.name,\n",
       "out.suffix = \"pathview\", pdf.size = c(7, 7), multi.state=TRUE,\n",
       "same.layer = TRUE, match.data = TRUE, rankdir = c(\"LR\", \"TB\")[1],\n",
       "is.signal = TRUE, split.group = F, afactor = 1, text.width = 15, cex =\n",
       "0.5, map.cpdname = FALSE, cpd.lab.offset = 1.0,\n",
       "discrete=list(gene=FALSE, cpd=FALSE), limit = list(gene = 1, cpd = 1),\n",
       "bins = list(gene = 10, cpd = 10), both.dirs = list(gene = T, cpd = T),\n",
       "low = list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd\n",
       "= \"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "\"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "\"x\", key.pos = \"topright\", sign.pos = \"bottomright\", ...)  </pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>gene.data</code></td>\n",
       "<td>\n",
       "\n",
       "<p>either vector (single sample) or a matrix-like data (multiple\n",
       "sample). Vector should be numeric with gene IDs as names or it may\n",
       "also be character of gene IDs. Character vector is treated as discrete\n",
       "or count data. Matrix-like data structure has genes as rows and\n",
       "samples as columns. Row names should be gene IDs. Here gene ID is a\n",
       "generic concepts, including multiple types of gene, transcript and\n",
       "protein uniquely mappable to KEGG gene IDs. KEGG ortholog IDs are also\n",
       "treated as gene IDs as to handle metagenomic data. Check details for\n",
       "mappable ID types. Default gene.data=NULL.\n",
       "</p>\n",
       "<p>numeric, character, continuous\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cpd.data</code></td>\n",
       "<td>\n",
       "\n",
       "<p>the same as gene.data, excpet named with IDs mappable to KEGG\n",
       "compound IDs. Over 20 types of IDs included in CHEMBL database can be\n",
       "used here. Check details for mappable ID types. Default\n",
       "cpd.data=NULL. Note that gene.data and cpd.data can't be NULL\n",
       "simultaneously.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>pathway.id</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character vector, the KEGG pathway ID(s), usually 5 digit, may also include the 3\n",
       "letter KEGG species code.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>species</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, either the kegg code, scientific name or the common name of\n",
       "the target species. This applies to both pathway and gene.data or\n",
       "cpd.data. When KEGG ortholog pathway is considered,\n",
       "species=&quot;ko&quot;. Default species=&quot;hsa&quot;, it is equivalent to use either\n",
       "&quot;Homo sapiens&quot; (scientific name) or &quot;human&quot; (common name). \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>kegg.dir</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, the directory of KEGG pathway data file (.xml) and image file\n",
       "(.png). Users may supply their own data files in the same format and\n",
       "naming convention of KEGG's (species code + pathway id,\n",
       "e.g. hsa04110.xml, hsa04110.png etc) in this directory. Default\n",
       "kegg.dir=&quot;.&quot; (current working directory). \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cpd.idtype</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, ID type used for the cpd.data. Default cpd.idtype=&quot;kegg&quot; (include\n",
       "compound, glycan and drug accessions). \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>gene.idtype</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, ID type used for the gene.data, case insensitive. Default\n",
       "gene.idtype=&quot;entrez&quot;, i.e. Entrez Gene, which are the primary KEGG gene ID\n",
       "for many common model organisms. For other species, gene.idtype should\n",
       "be set to &quot;KEGG&quot; as KEGG use other types of gene IDs. For the common\n",
       "model organisms (to check the list, do: <code>data(bods); bods</code>), you\n",
       "may also specify other types of valid IDs. To check the ID list, do:\n",
       "<code>data(gene.idtype.list); gene.idtype.list</code>. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>gene.annotpkg</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, the name of the annotation package to use for mapping between\n",
       "other gene ID types including symbols and Entrez gene ID. Default gene.annotpkg=NULL.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>min.nnodes</code></td>\n",
       "<td>\n",
       "\n",
       "<p>integer, minimal number of nodes of type &quot;gene&quot;,&quot;enzyme&quot;, &quot;compound&quot; or\n",
       "&quot;ortholog&quot; for a pathway to be considered. Default min.nnodes=3.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>kegg.native</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether to render pathway graph as native KEGG graph (.png) or\n",
       "using graphviz layout engine (.pdf). Default kegg.native=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>map.null</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether to map the NULL gene.data or cpd.data to pathway. When\n",
       "NULL data are mapped, the gene or compound nodes in the pathway will be\n",
       "rendered as actually mapped nodes, except with NA-valued color. When\n",
       "NULL data are not mapped, the nodes are rendered as unmapped nodes. This\n",
       "argument mainly affects native KEGG graph view, i.e. when\n",
       "kegg.native=TRUE. Default map.null=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>expand.node</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether the multiple-gene nodes are expanded into single-gene\n",
       "nodes. Each expanded single-gene nodes inherits all edges from the\n",
       "original multiple-gene node. This option only affects graphviz graph view, i.e. when\n",
       "kegg.native=FALSE. This option is not effective for most metabolic\n",
       "pathways where it conflits with  converting reactions to edges. Default expand.node=FLASE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>split.group</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether split node groups are split to individual nodes. Each\n",
       "split member nodes inherits all edges from the node group. This\n",
       "option only affects graphviz graph view, i.e. when\n",
       "kegg.native=FALSE. This option also effects most metabolic\n",
       "pathways even without group nodes defined orginally. For these pathways,\n",
       "genes involved in the same reaction are grouped automatically when\n",
       "converting reactions to edges unless split.group=TRUE. d\n",
       "split.group=FLASE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>map.symbol</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether map gene IDs to symbols for gene node labels or use the\n",
       "graphic name from the KGML file. This option is only effective for\n",
       "kegg.native=FALSE or same.layer=FALSE when kegg.native=TRUE. For\n",
       "same.layer=TRUE when kegg.native=TRUE, the native KEGG labels will be\n",
       "kept. Default map.symbol=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>map.cpdname</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether map compound IDs to formal names for compound node labels or use the\n",
       "graphic name from the KGML file (KEGG compound accessions). This option is only effective for\n",
       "kegg.native=FALSE. When kegg.native=TRUE, the native KEGG labels will be\n",
       "kept. Default map.cpdname=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>node.sum</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, the method name to calculate node summary given that\n",
       "multiple genes or compounds are mapped to it. Poential options include\n",
       "&quot;sum&quot;,&quot;mean&quot;, &quot;median&quot;, &quot;max&quot;, &quot;max.abs&quot; and &quot;random&quot;. Default node.sum=&quot;sum&quot;.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>discrete</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a list of two logical elements with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument tells whether gene.data or cpd.data should be treated as\n",
       "discrete. Default dsicrete=list(gene=FALSE, cpd=FALSE), i.e. both data should\n",
       "be treated as continuous.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>limit</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a list of two numeric elements with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument specifies the limit values for gene.data and cpd.data when converting\n",
       "them to pseudo colors. Each element of the list could be of length 1 or\n",
       "2. Length 1 suggests discrete data or 1 directional (positive-valued)\n",
       "data, or the absolute limit for 2 directional data. Length 2 suggests 2\n",
       "directional data. Default limit=list(gene=1, cpd=1).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>bins</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a list of two integer elements with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument specifies the number of levels or bins for gene.data and cpd.data when converting\n",
       "them to pseudo colors. Default limit=list(gene=10, cpd=10).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>both.dirs</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a list of two logical  elements with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument specifies whether gene.data and cpd.data are 1 directional or 2\n",
       "directional data when converting them to pseudo colors. Default limit=list(gene=TRUE, cpd=TRUE).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>trans.fun</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a list of two function (not character) elements with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument specifies whether and how gene.data and cpd.data are\n",
       "transformed. Examples are <code>log</code>, <code>abs</code> or users' own\n",
       "functions. Default limit=list(gene=NULL, cpd=NULL).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>low, mid, high</code></td>\n",
       "<td>\n",
       "\n",
       "<p>each is a list of two colors with &quot;gene&quot; and &quot;cpd&quot; as the names. This\n",
       "argument specifies the color spectra to code gene.data and\n",
       "cpd.data. When data are 1 directional (TRUE value in both.dirs), only\n",
       "mid and high are used to specify the color spectra. Default spectra (low-mid-high)\n",
       "&quot;green&quot;-&quot;gray&quot;-&quot;red&quot; and &quot;blue&quot;-&quot;gray&quot;-&quot;yellow&quot; are used for gene.data\n",
       "and cpd.data respectively. \n",
       "The values for 'low, mid, high' can be given as color names\n",
       "('red'), plot color index (2=red), and HTML-style RGB,\n",
       "(&quot;\\#FF0000&quot;=red).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>na.col</code></td>\n",
       "<td>\n",
       "\n",
       "<p>color used for NA's or missing values in gene.data and cpd.data. d\n",
       "na.col=&quot;transparent&quot;. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>...</code></td>\n",
       "<td>\n",
       "\n",
       "<p>extra arguments passed to keggview.native or keggview.graph function.\n",
       "</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "<p>special arguments for keggview.native or keggview.graph function.\n",
       "</p>\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>plot.data.gene</code></td>\n",
       "<td>\n",
       "\n",
       "<p>data.frame returned by node.map function for rendering mapped gene\n",
       "nodes, including node name, type, positions (x, y), sizes (width,\n",
       "height), and mapped gene.data. This data is also used as input for\n",
       "pseduo-color coding through node.color function. Default plot.data.gene=NULL.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>plot.data.cpd</code></td>\n",
       "<td>\n",
       "\n",
       "<p>same as plot.data.gene function, except for mapped compound node data. d\n",
       "plot.data.cpd=NULL. Default plot.data.cpd=NULL. Note that plot.data.gene and\n",
       "plot.data.cpd can't be NULL simultaneously.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cols.ts.gene</code></td>\n",
       "<td>\n",
       "\n",
       "<p>vector or matrix of colors returned by node.color function for rendering\n",
       "gene.data. Dimensionality is the same as the latter. Default cols.ts.gene=NULL.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cols.ts.cpd</code></td>\n",
       "<td>\n",
       "\n",
       "<p>same as cols.ts.gene, except corresponding to cpd.data. d\n",
       "cols.ts.cpd=NULL. Note that cols.ts.gene and cols.ts.cpd\n",
       "plot.data.gene can't be NULL simultaneously. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>node.data</code></td>\n",
       "<td>\n",
       "\n",
       "<p>list returned by node.info function, which parse KGML file directly or\n",
       "indirectly, and extract the node data.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>pathway.name</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, the full KEGG pathway name in the format of 3-letter species\n",
       "code with 5-digit pathway id, eg &quot;hsa04612&quot;. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>out.suffix</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, the suffix to be added after the pathway name as part of the\n",
       "output graph file. Sample names or column names of the gene.data or\n",
       "cpd.data are also added when there are multiple samples. Default out.suffix=&quot;pathview&quot;.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>multi.state</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether multiple states (samples or columns) gene.data or\n",
       "cpd.data should be integrated and plotted in the same graph. Default\n",
       "match.data=TRUE. In other words, gene or compound nodes will be sliced\n",
       "into multiple pieces corresponding to the number of states in the\n",
       "data.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>match.data</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether the samples of gene.data and cpd.data are\n",
       "paired. Default match.data=TRUE. When let sample sizes of gene.data\n",
       "and cpd.data be m and n, when m&gt;n, extra columns of NA's (mapped to no\n",
       "color) will be added to cpd.data as to make the sample size the\n",
       "same. This will result in the smae number of slice in gene nodes and\n",
       "compound when multi.state=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>same.layer</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, control plotting layers:\n",
       "1) if node colors be plotted in the same layer as the pathway\n",
       "graph when kegg.native=TRUE, 2) if  edge/node type legend be plotted in\n",
       "the same page when kegg.native=FALSE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>res</code></td>\n",
       "<td>\n",
       "\n",
       "<p>The nominal resolution in ppi which will be recorded in the\n",
       "bitmap file, if a positive integer.  Also used for 'units'\n",
       "other than the default, and to convert points to pixels. This\n",
       "argument is only effective when kegg.native=TRUE. Default res=300.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cex</code></td>\n",
       "<td>\n",
       "\n",
       "<p>A numerical value giving the amount by which plotting text\n",
       "and symbols should be scaled relative to the default 1. Default cex=0.25\n",
       "when kegg.native=TRUE, cex=0.5 when kegg.native=FALSE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>new.signature</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether pathview signature is added to the pathway\n",
       "graphs. Default new.signature=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>plot.col.key</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, whether color key is added to the pathway\n",
       "graphs. Default plot.col.key= TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>key.align</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, controlling how the color keys are aligned when both\n",
       "gene.data and cpd.data are not NULL. Potential values are &quot;x&quot;, aligned\n",
       "by x coordinates, and &quot;y&quot;, aligned by y coordinates. Default key.align=&quot;x&quot;.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>key.pos</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, controlling the position of color key(s). Potentail values\n",
       "are &quot;bottomleft&quot;, &quot;bottomright&quot;, &quot;topleft&quot; and &quot;topright&quot;. d\n",
       "key.pos=&quot;topright&quot;. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>sign.pos</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, controlling the position of pathview signature. Only\n",
       "effective when kegg.native=FALSE, Signature position is fixed in place\n",
       "of the original KEGG signature when kegg.native=TRUE. Potentail values\n",
       "are &quot;bottomleft&quot;, &quot;bottomright&quot;, &quot;topleft&quot; and &quot;topright&quot;. d\n",
       "sign.pos=&quot;bottomright&quot;. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>path.graph</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a graph object parsed from KGML file, only effective when\n",
       "kegg.native=FALSE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>pdf.size</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a numeric vector of length 2, giving the width and height of the pathway\n",
       "graph pdf file. Note that pdf width increase by half when\n",
       "same.layer=TRUE to accommodate legends. Only effective when\n",
       "kegg.native=FALSE. Default pdf.size=c(7,7).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>rankdir</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, either &quot;LR&quot; (left to right) or &quot;TB&quot; (top to bottom),\n",
       "specifying the pathway graph layout direction. Only effective when\n",
       "kegg.native=FALSE. Default rank.dir=&quot;LR&quot;.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>is.signal</code></td>\n",
       "<td>\n",
       "\n",
       "<p>logical, if the pathway is treated as a signaling pathway, where all the\n",
       "unconnected nodes are dropped. This argument also affect the graph\n",
       "layout type, i.e. &quot;dot&quot; for signals or &quot;neato&quot; otherwise. Only effective when\n",
       "kegg.native=FALSE. Default is.signal=TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>afactor</code></td>\n",
       "<td>\n",
       "\n",
       "<p>numeric, node amplifying factor. This argument is for node size\n",
       "fine-tuning, its effect is subtler than expected. Only effective when\n",
       "kegg.native=FALSE. Default afctor=1.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>text.width</code></td>\n",
       "<td>\n",
       "\n",
       "<p>numeric, specifying the line width for text wrap. Only effective when\n",
       "kegg.native= FALSE. Default text.width=15 (characters).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cpd.lab.offset</code></td>\n",
       "<td>\n",
       "\n",
       "<p>numeric, specifying how much compound labels should be put above the\n",
       "default position or node center. This argument is useful when\n",
       "map.cpdname=TRUE, i.e. compounds are labelled by full name, which\n",
       "affects the look of compound nodes and color. Only effective when\n",
       "kegg.native=FALSE. Default cpd.lab.offset=1.0.\n",
       "</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>Pathview maps and renders user data on relevant pathway graphs. Pathview\n",
       "is a stand alone program for pathway based data integration and\n",
       "visualization. It also seamlessly integrates with pathway and functional\n",
       "analysis tools for large-scale and fully automated analysis.\n",
       "Pathview provides strong support for data Integration. It works with: 1)\n",
       "essentially all types of biological data mappable to pathways, 2) over\n",
       "10 types of gene or protein IDs, and 20 types of compound or metabolite\n",
       "IDs, 3) pathways for over 2000 species as well as KEGG orthology, 4)\n",
       "varoius data attributes and formats, i.e. continuous/discrete data,\n",
       "matrices/vectors, single/multiple samples etc.\n",
       "To see mappable external gene/protein IDs do:\n",
       "<code>data(gene.idtype.list)</code>, to see mappable external compound related\n",
       "IDs do: <code>data(rn.list)</code>; names(rn.list).\n",
       "Pathview generates both native KEGG view and Graphviz views for\n",
       "pathways. Currently only KEGG pathways are implemented. Hopefully, pathways from\n",
       "Reactome, NCI and other databases will be supported in the future.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>From viersion 1.9.3, pathview can accept either a single pathway or\n",
       "multiple pathway ids. The result returned by pathview function is a\n",
       "named list corresponding to the input pathway ids. Each element (for\n",
       "each pathway itself is a named list, with 2\n",
       "elements (&quot;plot.data.gene&quot;, &quot;plot.data.cpd&quot;). Both elements\n",
       "are data.frame or NULL depends on the corresponding input\n",
       "data gene.data and cpd.data. These data.frames record the plot data\n",
       "for mapped gene or compound nodes: rows are mapped genes/compounds,\n",
       "columns are:\n",
       "</p>\n",
       "<table summary=\"R valueblock\">\n",
       "<tr valign=\"top\"><td><code>kegg.names</code></td>\n",
       "<td>\n",
       "<p>standard KEGG IDs/Names for mapped nodes. It's\n",
       "Entrez Gene ID or KEGG Compound Accessions.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>labels</code></td>\n",
       "<td>\n",
       "<p>Node labels to be used when needed.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>all.mapped</code></td>\n",
       "<td>\n",
       "<p>All molecule (gene or compound) IDs mapped to this\n",
       "node.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>type</code></td>\n",
       "<td>\n",
       "<p>node type, currently 4 types are supported:\n",
       "&quot;gene&quot;,&quot;enzyme&quot;, &quot;compound&quot; and &quot;ortholog&quot;.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>x</code></td>\n",
       "<td>\n",
       "<p>x coordinate in the original KEGG pathway graph.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>y</code></td>\n",
       "<td>\n",
       "<p>y coordinate in the original KEGG pathway graph.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>width</code></td>\n",
       "<td>\n",
       "<p>node width in the original KEGG pathway graph.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>height</code></td>\n",
       "<td>\n",
       "<p>node height in the original KEGG pathway graph.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>other columns </code></td>\n",
       "<td>\n",
       "<p>columns of the mapped gene/compound data and\n",
       "corresponding pseudo-color codes for individual samples\n",
       "</p>\n",
       "</td></tr>      \n",
       "</table>\n",
       "<p>The results returned by <code>keggview.native</code> and\n",
       "codekeggview.graph are both a list of graph plotting\n",
       "parameters. These are not intended to be used externally.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Weijun Luo &lt;luo_weijun@yahoo.com&gt;\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>Luo, W. and Brouwer, C., Pathview: an R/Bioconductor package for\n",
       "pathway based data integration and visualization. Bioinformatics,\n",
       "2013, 29(14): 1830-1831, doi: 10.1093/bioinformatics/btt285\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>download.kegg</code> the downloader,\n",
       "<code>node.info</code> the parser,\n",
       "<code>node.map</code> and <code>node.color</code> the mapper. \n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "#load data\n",
       "data(gse16873.d)\n",
       "data(demo.paths)\n",
       "\n",
       "#KEGG view: gene data only\n",
       "i &lt;- 1\n",
       "pv.out &lt;- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "kegg.native = TRUE)\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.gene)\n",
       "#result PNG file in current directory\n",
       "\n",
       "#Graphviz view: gene data only\n",
       "pv.out &lt;- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "kegg.native = FALSE, sign.pos = demo.paths$spos[i])\n",
       "#result PDF file in current directory\n",
       "\n",
       "#KEGG view: both gene and compound data\n",
       "sim.cpd.data=sim.mol.data(mol.type=\"cpd\", nmol=3000)\n",
       "i &lt;- 3\n",
       "print(demo.paths$sel.paths[i])\n",
       "pv.out &lt;- pathview(gene.data = gse16873.d[, 1], cpd.data = sim.cpd.data,\n",
       "pathway.id = demo.paths$sel.paths[i], species = \"hsa\", out.suffix =\n",
       "\"gse16873.cpd\", keys.align = \"y\", kegg.native = TRUE, key.pos = demo.paths$kpos1[i])\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.cpd)\n",
       "\n",
       "#multiple states in one graph\n",
       "set.seed(10)\n",
       "sim.cpd.data2 = matrix(sample(sim.cpd.data, 18000, \n",
       "    replace = TRUE), ncol = 6)\n",
       "pv.out &lt;- pathview(gene.data = gse16873.d[, 1:3], \n",
       "    cpd.data = sim.cpd.data2[, 1:2], pathway.id = demo.paths$sel.paths[i], \n",
       "    species = \"hsa\", out.suffix = \"gse16873.cpd.3-2s\", keys.align = \"y\", \n",
       "    kegg.native = TRUE, match.data = FALSE, multi.state = TRUE, same.layer = TRUE)\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.cpd)\n",
       "\n",
       "#result PNG file in current directory\n",
       "\n",
       "##more examples of pathview usages are shown in the vignette.\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>pathview</em> version 1.24.0 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{pathview}{Pathway based data integration and visualization}{pathview}\n",
       "\\aliasA{keggview.graph}{pathview}{keggview.graph}\n",
       "\\aliasA{keggview.native}{pathview}{keggview.native}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Pathview is a tool set for pathway based data integration and\n",
       "visualization. It maps and renders user data on relevant pathway\n",
       "graphs. All users need is to supply their gene or compound data and\n",
       "specify the target pathway. Pathview automatically downloads the pathway\n",
       "graph data, parses the data file, maps user data to the pathway, and\n",
       "render pathway graph with the mapped data.\n",
       "Pathview generates both native KEGG view and Graphviz views for\n",
       "pathways. keggview.native and keggview.graph are the two viewer\n",
       "functions, and pathview is the main function providing a unified\n",
       "interface to downloader, parser, mapper and viewer functions.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "pathview(gene.data = NULL, cpd.data = NULL, pathway.id,\n",
       "species = \"hsa\", kegg.dir = \".\", cpd.idtype = \"kegg\", gene.idtype =\n",
       "\"entrez\", gene.annotpkg = NULL, min.nnodes = 3, kegg.native = TRUE,\n",
       "map.null = TRUE, expand.node = FALSE, split.group = FALSE, map.symbol =\n",
       "TRUE, map.cpdname = TRUE, node.sum = \"sum\", discrete=list(gene=FALSE,\n",
       "cpd=FALSE), limit = list(gene = 1, cpd = 1), bins = list(gene = 10, cpd\n",
       "= 10), both.dirs = list(gene = T, cpd = T), trans.fun = list(gene =\n",
       "NULL, cpd = NULL), low = list(gene = \"green\", cpd = \"blue\"), mid =\n",
       "list(gene = \"gray\", cpd = \"gray\"), high = list(gene = \"red\", cpd =\n",
       "\"yellow\"), na.col = \"transparent\", ...)\n",
       "\n",
       "keggview.native(plot.data.gene = NULL, plot.data.cpd = NULL,\n",
       "cols.ts.gene = NULL, cols.ts.cpd = NULL, node.data, pathway.name,\n",
       "out.suffix = \"pathview\", kegg.dir = \".\", multi.state=TRUE, match.data =\n",
       "TRUE, same.layer = TRUE, res = 300, cex = 0.25, discrete =\n",
       "list(gene=FALSE, cpd=FALSE), limit= list(gene = 1, cpd = 1), bins =\n",
       "list(gene = 10, cpd = 10), both.dirs =list(gene = T, cpd = T), low =\n",
       "list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd =\n",
       "\"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "\"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "\"x\", key.pos = \"topright\", ...)\n",
       "\n",
       "keggview.graph(plot.data.gene = NULL, plot.data.cpd = NULL, cols.ts.gene\n",
       "= NULL, cols.ts.cpd = NULL, node.data, path.graph, pathway.name,\n",
       "out.suffix = \"pathview\", pdf.size = c(7, 7), multi.state=TRUE,\n",
       "same.layer = TRUE, match.data = TRUE, rankdir = c(\"LR\", \"TB\")[1],\n",
       "is.signal = TRUE, split.group = F, afactor = 1, text.width = 15, cex =\n",
       "0.5, map.cpdname = FALSE, cpd.lab.offset = 1.0,\n",
       "discrete=list(gene=FALSE, cpd=FALSE), limit = list(gene = 1, cpd = 1),\n",
       "bins = list(gene = 10, cpd = 10), both.dirs = list(gene = T, cpd = T),\n",
       "low = list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd\n",
       "= \"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "\"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "\"x\", key.pos = \"topright\", sign.pos = \"bottomright\", ...)  \n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{gene.data}] \n",
       "either vector (single sample) or a matrix-like data (multiple\n",
       "sample). Vector should be numeric with gene IDs as names or it may\n",
       "also be character of gene IDs. Character vector is treated as discrete\n",
       "or count data. Matrix-like data structure has genes as rows and\n",
       "samples as columns. Row names should be gene IDs. Here gene ID is a\n",
       "generic concepts, including multiple types of gene, transcript and\n",
       "protein uniquely mappable to KEGG gene IDs. KEGG ortholog IDs are also\n",
       "treated as gene IDs as to handle metagenomic data. Check details for\n",
       "mappable ID types. Default gene.data=NULL.\n",
       "\n",
       "numeric, character, continuous\n",
       "\n",
       "\\item[\\code{cpd.data}] \n",
       "the same as gene.data, excpet named with IDs mappable to KEGG\n",
       "compound IDs. Over 20 types of IDs included in CHEMBL database can be\n",
       "used here. Check details for mappable ID types. Default\n",
       "cpd.data=NULL. Note that gene.data and cpd.data can't be NULL\n",
       "simultaneously.\n",
       "\n",
       "\\item[\\code{pathway.id}] \n",
       "character vector, the KEGG pathway ID(s), usually 5 digit, may also include the 3\n",
       "letter KEGG species code.\n",
       "\n",
       "\\item[\\code{species}] \n",
       "character, either the kegg code, scientific name or the common name of\n",
       "the target species. This applies to both pathway and gene.data or\n",
       "cpd.data. When KEGG ortholog pathway is considered,\n",
       "species=\"ko\". Default species=\"hsa\", it is equivalent to use either\n",
       "\"Homo sapiens\" (scientific name) or \"human\" (common name). \n",
       "\n",
       "\\item[\\code{kegg.dir}] \n",
       "character, the directory of KEGG pathway data file (.xml) and image file\n",
       "(.png). Users may supply their own data files in the same format and\n",
       "naming convention of KEGG's (species code + pathway id,\n",
       "e.g. hsa04110.xml, hsa04110.png etc) in this directory. Default\n",
       "kegg.dir=\".\" (current working directory). \n",
       "\n",
       "\\item[\\code{cpd.idtype}] \n",
       "character, ID type used for the cpd.data. Default cpd.idtype=\"kegg\" (include\n",
       "compound, glycan and drug accessions). \n",
       "\n",
       "\\item[\\code{gene.idtype}] \n",
       "character, ID type used for the gene.data, case insensitive. Default\n",
       "gene.idtype=\"entrez\", i.e. Entrez Gene, which are the primary KEGG gene ID\n",
       "for many common model organisms. For other species, gene.idtype should\n",
       "be set to \"KEGG\" as KEGG use other types of gene IDs. For the common\n",
       "model organisms (to check the list, do: \\code{data(bods); bods}), you\n",
       "may also specify other types of valid IDs. To check the ID list, do:\n",
       "\\code{data(gene.idtype.list); gene.idtype.list}. \n",
       "\n",
       "\\item[\\code{gene.annotpkg}] \n",
       "character, the name of the annotation package to use for mapping between\n",
       "other gene ID types including symbols and Entrez gene ID. Default gene.annotpkg=NULL.\n",
       "\n",
       "\\item[\\code{min.nnodes}] \n",
       "integer, minimal number of nodes of type \"gene\",\"enzyme\", \"compound\" or\n",
       "\"ortholog\" for a pathway to be considered. Default min.nnodes=3.\n",
       "\n",
       "\\item[\\code{kegg.native}] \n",
       "logical, whether to render pathway graph as native KEGG graph (.png) or\n",
       "using graphviz layout engine (.pdf). Default kegg.native=TRUE.\n",
       "\n",
       "\\item[\\code{map.null}] \n",
       "logical, whether to map the NULL gene.data or cpd.data to pathway. When\n",
       "NULL data are mapped, the gene or compound nodes in the pathway will be\n",
       "rendered as actually mapped nodes, except with NA-valued color. When\n",
       "NULL data are not mapped, the nodes are rendered as unmapped nodes. This\n",
       "argument mainly affects native KEGG graph view, i.e. when\n",
       "kegg.native=TRUE. Default map.null=TRUE.\n",
       "\n",
       "\\item[\\code{expand.node}] \n",
       "logical, whether the multiple-gene nodes are expanded into single-gene\n",
       "nodes. Each expanded single-gene nodes inherits all edges from the\n",
       "original multiple-gene node. This option only affects graphviz graph view, i.e. when\n",
       "kegg.native=FALSE. This option is not effective for most metabolic\n",
       "pathways where it conflits with  converting reactions to edges. Default expand.node=FLASE.\n",
       "\n",
       "\\item[\\code{split.group}] \n",
       "logical, whether split node groups are split to individual nodes. Each\n",
       "split member nodes inherits all edges from the node group. This\n",
       "option only affects graphviz graph view, i.e. when\n",
       "kegg.native=FALSE. This option also effects most metabolic\n",
       "pathways even without group nodes defined orginally. For these pathways,\n",
       "genes involved in the same reaction are grouped automatically when\n",
       "converting reactions to edges unless split.group=TRUE. d\n",
       "split.group=FLASE.\n",
       "\n",
       "\\item[\\code{map.symbol}] \n",
       "logical, whether map gene IDs to symbols for gene node labels or use the\n",
       "graphic name from the KGML file. This option is only effective for\n",
       "kegg.native=FALSE or same.layer=FALSE when kegg.native=TRUE. For\n",
       "same.layer=TRUE when kegg.native=TRUE, the native KEGG labels will be\n",
       "kept. Default map.symbol=TRUE.\n",
       "\n",
       "\\item[\\code{map.cpdname}] \n",
       "logical, whether map compound IDs to formal names for compound node labels or use the\n",
       "graphic name from the KGML file (KEGG compound accessions). This option is only effective for\n",
       "kegg.native=FALSE. When kegg.native=TRUE, the native KEGG labels will be\n",
       "kept. Default map.cpdname=TRUE.\n",
       "\n",
       "\\item[\\code{node.sum}] \n",
       "character, the method name to calculate node summary given that\n",
       "multiple genes or compounds are mapped to it. Poential options include\n",
       "\"sum\",\"mean\", \"median\", \"max\", \"max.abs\" and \"random\". Default node.sum=\"sum\".\n",
       "\n",
       "\\item[\\code{discrete}] \n",
       "a list of two logical elements with \"gene\" and \"cpd\" as the names. This\n",
       "argument tells whether gene.data or cpd.data should be treated as\n",
       "discrete. Default dsicrete=list(gene=FALSE, cpd=FALSE), i.e. both data should\n",
       "be treated as continuous.\n",
       "\n",
       "\\item[\\code{limit}] \n",
       "a list of two numeric elements with \"gene\" and \"cpd\" as the names. This\n",
       "argument specifies the limit values for gene.data and cpd.data when converting\n",
       "them to pseudo colors. Each element of the list could be of length 1 or\n",
       "2. Length 1 suggests discrete data or 1 directional (positive-valued)\n",
       "data, or the absolute limit for 2 directional data. Length 2 suggests 2\n",
       "directional data. Default limit=list(gene=1, cpd=1).\n",
       "\n",
       "\\item[\\code{bins}] \n",
       "a list of two integer elements with \"gene\" and \"cpd\" as the names. This\n",
       "argument specifies the number of levels or bins for gene.data and cpd.data when converting\n",
       "them to pseudo colors. Default limit=list(gene=10, cpd=10).\n",
       "\n",
       "\\item[\\code{both.dirs}] \n",
       "a list of two logical  elements with \"gene\" and \"cpd\" as the names. This\n",
       "argument specifies whether gene.data and cpd.data are 1 directional or 2\n",
       "directional data when converting them to pseudo colors. Default limit=list(gene=TRUE, cpd=TRUE).\n",
       "\n",
       "\\item[\\code{trans.fun}] \n",
       "a list of two function (not character) elements with \"gene\" and \"cpd\" as the names. This\n",
       "argument specifies whether and how gene.data and cpd.data are\n",
       "transformed. Examples are \\code{log}, \\code{abs} or users' own\n",
       "functions. Default limit=list(gene=NULL, cpd=NULL).\n",
       "\n",
       "\\item[\\code{low, mid, high}] \n",
       "each is a list of two colors with \"gene\" and \"cpd\" as the names. This\n",
       "argument specifies the color spectra to code gene.data and\n",
       "cpd.data. When data are 1 directional (TRUE value in both.dirs), only\n",
       "mid and high are used to specify the color spectra. Default spectra (low-mid-high)\n",
       "\"green\"-\"gray\"-\"red\" and \"blue\"-\"gray\"-\"yellow\" are used for gene.data\n",
       "and cpd.data respectively. \n",
       "The values for 'low, mid, high' can be given as color names\n",
       "('red'), plot color index (2=red), and HTML-style RGB,\n",
       "(\"\\bsl{}\\#FF0000\"=red).\n",
       "\n",
       "\\item[\\code{na.col}] \n",
       "color used for NA's or missing values in gene.data and cpd.data. d\n",
       "na.col=\"transparent\". \n",
       "\n",
       "\\item[\\code{...}] \n",
       "extra arguments passed to keggview.native or keggview.graph function.\n",
       "\n",
       "\n",
       "special arguments for keggview.native or keggview.graph function.\n",
       "\\item[\\code{plot.data.gene}] \n",
       "data.frame returned by node.map function for rendering mapped gene\n",
       "nodes, including node name, type, positions (x, y), sizes (width,\n",
       "height), and mapped gene.data. This data is also used as input for\n",
       "pseduo-color coding through node.color function. Default plot.data.gene=NULL.\n",
       "\n",
       "\\item[\\code{plot.data.cpd}] \n",
       "same as plot.data.gene function, except for mapped compound node data. d\n",
       "plot.data.cpd=NULL. Default plot.data.cpd=NULL. Note that plot.data.gene and\n",
       "plot.data.cpd can't be NULL simultaneously.\n",
       "\n",
       "\\item[\\code{cols.ts.gene}] \n",
       "vector or matrix of colors returned by node.color function for rendering\n",
       "gene.data. Dimensionality is the same as the latter. Default cols.ts.gene=NULL.\n",
       "\n",
       "\\item[\\code{cols.ts.cpd}] \n",
       "same as cols.ts.gene, except corresponding to cpd.data. d\n",
       "cols.ts.cpd=NULL. Note that cols.ts.gene and cols.ts.cpd\n",
       "plot.data.gene can't be NULL simultaneously. \n",
       "\n",
       "\\item[\\code{node.data}] \n",
       "list returned by node.info function, which parse KGML file directly or\n",
       "indirectly, and extract the node data.\n",
       "\n",
       "\\item[\\code{pathway.name}] \n",
       "character, the full KEGG pathway name in the format of 3-letter species\n",
       "code with 5-digit pathway id, eg \"hsa04612\". \n",
       "\n",
       "\\item[\\code{out.suffix}] \n",
       "character, the suffix to be added after the pathway name as part of the\n",
       "output graph file. Sample names or column names of the gene.data or\n",
       "cpd.data are also added when there are multiple samples. Default out.suffix=\"pathview\".\n",
       "\n",
       "\\item[\\code{multi.state}] \n",
       "logical, whether multiple states (samples or columns) gene.data or\n",
       "cpd.data should be integrated and plotted in the same graph. Default\n",
       "match.data=TRUE. In other words, gene or compound nodes will be sliced\n",
       "into multiple pieces corresponding to the number of states in the\n",
       "data.\n",
       "\n",
       "\\item[\\code{match.data}] \n",
       "logical, whether the samples of gene.data and cpd.data are\n",
       "paired. Default match.data=TRUE. When let sample sizes of gene.data\n",
       "and cpd.data be m and n, when m>n, extra columns of NA's (mapped to no\n",
       "color) will be added to cpd.data as to make the sample size the\n",
       "same. This will result in the smae number of slice in gene nodes and\n",
       "compound when multi.state=TRUE.\n",
       "\n",
       "\\item[\\code{same.layer}] \n",
       "logical, control plotting layers:\n",
       "1) if node colors be plotted in the same layer as the pathway\n",
       "graph when kegg.native=TRUE, 2) if  edge/node type legend be plotted in\n",
       "the same page when kegg.native=FALSE.\n",
       "\n",
       "\\item[\\code{res}] \n",
       "The nominal resolution in ppi which will be recorded in the\n",
       "bitmap file, if a positive integer.  Also used for 'units'\n",
       "other than the default, and to convert points to pixels. This\n",
       "argument is only effective when kegg.native=TRUE. Default res=300.\n",
       "\n",
       "\\item[\\code{cex}] \n",
       "A numerical value giving the amount by which plotting text\n",
       "and symbols should be scaled relative to the default 1. Default cex=0.25\n",
       "when kegg.native=TRUE, cex=0.5 when kegg.native=FALSE.\n",
       "\n",
       "\\item[\\code{new.signature}] \n",
       "logical, whether pathview signature is added to the pathway\n",
       "graphs. Default new.signature=TRUE.\n",
       "\n",
       "\\item[\\code{plot.col.key}] \n",
       "logical, whether color key is added to the pathway\n",
       "graphs. Default plot.col.key= TRUE.\n",
       "\n",
       "\n",
       "\\item[\\code{key.align}] \n",
       "character, controlling how the color keys are aligned when both\n",
       "gene.data and cpd.data are not NULL. Potential values are \"x\", aligned\n",
       "by x coordinates, and \"y\", aligned by y coordinates. Default key.align=\"x\".\n",
       "\n",
       "\\item[\\code{key.pos}] \n",
       "character, controlling the position of color key(s). Potentail values\n",
       "are \"bottomleft\", \"bottomright\", \"topleft\" and \"topright\". d\n",
       "key.pos=\"topright\". \n",
       "\n",
       "\\item[\\code{sign.pos}] \n",
       "character, controlling the position of pathview signature. Only\n",
       "effective when kegg.native=FALSE, Signature position is fixed in place\n",
       "of the original KEGG signature when kegg.native=TRUE. Potentail values\n",
       "are \"bottomleft\", \"bottomright\", \"topleft\" and \"topright\". d\n",
       "sign.pos=\"bottomright\". \n",
       "\n",
       "\n",
       "\\item[\\code{path.graph}] \n",
       "a graph object parsed from KGML file, only effective when\n",
       "kegg.native=FALSE.\n",
       "\n",
       "\\item[\\code{pdf.size}] \n",
       "a numeric vector of length 2, giving the width and height of the pathway\n",
       "graph pdf file. Note that pdf width increase by half when\n",
       "same.layer=TRUE to accommodate legends. Only effective when\n",
       "kegg.native=FALSE. Default pdf.size=c(7,7).\n",
       "\n",
       "\\item[\\code{rankdir}] \n",
       "character, either \"LR\" (left to right) or \"TB\" (top to bottom),\n",
       "specifying the pathway graph layout direction. Only effective when\n",
       "kegg.native=FALSE. Default rank.dir=\"LR\".\n",
       "\n",
       "\\item[\\code{is.signal}] \n",
       "logical, if the pathway is treated as a signaling pathway, where all the\n",
       "unconnected nodes are dropped. This argument also affect the graph\n",
       "layout type, i.e. \"dot\" for signals or \"neato\" otherwise. Only effective when\n",
       "kegg.native=FALSE. Default is.signal=TRUE.\n",
       "\n",
       "\\item[\\code{afactor}] \n",
       "numeric, node amplifying factor. This argument is for node size\n",
       "fine-tuning, its effect is subtler than expected. Only effective when\n",
       "kegg.native=FALSE. Default afctor=1.\n",
       "\n",
       "\\item[\\code{text.width}] \n",
       "numeric, specifying the line width for text wrap. Only effective when\n",
       "kegg.native= FALSE. Default text.width=15 (characters).\n",
       "\n",
       "\\item[\\code{cpd.lab.offset}] \n",
       "numeric, specifying how much compound labels should be put above the\n",
       "default position or node center. This argument is useful when\n",
       "map.cpdname=TRUE, i.e. compounds are labelled by full name, which\n",
       "affects the look of compound nodes and color. Only effective when\n",
       "kegg.native=FALSE. Default cpd.lab.offset=1.0.\n",
       "\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "Pathview maps and renders user data on relevant pathway graphs. Pathview\n",
       "is a stand alone program for pathway based data integration and\n",
       "visualization. It also seamlessly integrates with pathway and functional\n",
       "analysis tools for large-scale and fully automated analysis.\n",
       "Pathview provides strong support for data Integration. It works with: 1)\n",
       "essentially all types of biological data mappable to pathways, 2) over\n",
       "10 types of gene or protein IDs, and 20 types of compound or metabolite\n",
       "IDs, 3) pathways for over 2000 species as well as KEGG orthology, 4)\n",
       "varoius data attributes and formats, i.e. continuous/discrete data,\n",
       "matrices/vectors, single/multiple samples etc.\n",
       "To see mappable external gene/protein IDs do:\n",
       "\\code{data(gene.idtype.list)}, to see mappable external compound related\n",
       "IDs do: \\code{data(rn.list)}; names(rn.list).\n",
       "Pathview generates both native KEGG view and Graphviz views for\n",
       "pathways. Currently only KEGG pathways are implemented. Hopefully, pathways from\n",
       "Reactome, NCI and other databases will be supported in the future.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "From viersion 1.9.3, pathview can accept either a single pathway or\n",
       "multiple pathway ids. The result returned by pathview function is a\n",
       "named list corresponding to the input pathway ids. Each element (for\n",
       "each pathway itself is a named list, with 2\n",
       "elements (\"plot.data.gene\", \"plot.data.cpd\"). Both elements\n",
       "are data.frame or NULL depends on the corresponding input\n",
       "data gene.data and cpd.data. These data.frames record the plot data\n",
       "for mapped gene or compound nodes: rows are mapped genes/compounds,\n",
       "columns are:\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{kegg.names}] standard KEGG IDs/Names for mapped nodes. It's\n",
       "Entrez Gene ID or KEGG Compound Accessions.\n",
       "\\item[\\code{labels}] Node labels to be used when needed.\n",
       "\n",
       "\\item[\\code{all.mapped}] All molecule (gene or compound) IDs mapped to this\n",
       "node.\n",
       "\n",
       "\\item[\\code{type}] node type, currently 4 types are supported:\n",
       "\"gene\",\"enzyme\", \"compound\" and \"ortholog\".\n",
       "\n",
       "\\item[\\code{x}] x coordinate in the original KEGG pathway graph.\n",
       "\n",
       "\\item[\\code{y}] y coordinate in the original KEGG pathway graph.\n",
       "\n",
       "\\item[\\code{width}] node width in the original KEGG pathway graph.\n",
       "\n",
       "\\item[\\code{height}] node height in the original KEGG pathway graph.\n",
       "\n",
       "\\item[\\code{other columns }] columns of the mapped gene/compound data and\n",
       "corresponding pseudo-color codes for individual samples\n",
       "     \n",
       "\n",
       "\\end{ldescription}\n",
       "The results returned by \\code{keggview.native} and\n",
       "codekeggview.graph are both a list of graph plotting\n",
       "parameters. These are not intended to be used externally.\n",
       "\n",
       "\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "Weijun Luo <luo\\_weijun@yahoo.com>\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "Luo, W. and Brouwer, C., Pathview: an R/Bioconductor package for\n",
       "pathway based data integration and visualization. Bioinformatics,\n",
       "2013, 29(14): 1830-1831, doi: 10.1093/bioinformatics/btt285\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\code{\\LinkA{download.kegg}{download.kegg}} the downloader,\n",
       "\\code{\\LinkA{node.info}{node.info}} the parser,\n",
       "\\code{\\LinkA{node.map}{node.map}} and \\code{\\LinkA{node.color}{node.color}} the mapper. \n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "#load data\n",
       "data(gse16873.d)\n",
       "data(demo.paths)\n",
       "\n",
       "#KEGG view: gene data only\n",
       "i <- 1\n",
       "pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "kegg.native = TRUE)\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.gene)\n",
       "#result PNG file in current directory\n",
       "\n",
       "#Graphviz view: gene data only\n",
       "pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "kegg.native = FALSE, sign.pos = demo.paths$spos[i])\n",
       "#result PDF file in current directory\n",
       "\n",
       "#KEGG view: both gene and compound data\n",
       "sim.cpd.data=sim.mol.data(mol.type=\"cpd\", nmol=3000)\n",
       "i <- 3\n",
       "print(demo.paths$sel.paths[i])\n",
       "pv.out <- pathview(gene.data = gse16873.d[, 1], cpd.data = sim.cpd.data,\n",
       "pathway.id = demo.paths$sel.paths[i], species = \"hsa\", out.suffix =\n",
       "\"gse16873.cpd\", keys.align = \"y\", kegg.native = TRUE, key.pos = demo.paths$kpos1[i])\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.cpd)\n",
       "\n",
       "#multiple states in one graph\n",
       "set.seed(10)\n",
       "sim.cpd.data2 = matrix(sample(sim.cpd.data, 18000, \n",
       "    replace = TRUE), ncol = 6)\n",
       "pv.out <- pathview(gene.data = gse16873.d[, 1:3], \n",
       "    cpd.data = sim.cpd.data2[, 1:2], pathway.id = demo.paths$sel.paths[i], \n",
       "    species = \"hsa\", out.suffix = \"gse16873.cpd.3-2s\", keys.align = \"y\", \n",
       "    kegg.native = TRUE, match.data = FALSE, multi.state = TRUE, same.layer = TRUE)\n",
       "str(pv.out)\n",
       "head(pv.out$plot.data.cpd)\n",
       "\n",
       "#result PNG file in current directory\n",
       "\n",
       "##more examples of pathview usages are shown in the vignette.\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "pathview               package:pathview                R Documentation\n",
       "\n",
       "_\bP_\ba_\bt_\bh_\bw_\ba_\by _\bb_\ba_\bs_\be_\bd _\bd_\ba_\bt_\ba _\bi_\bn_\bt_\be_\bg_\br_\ba_\bt_\bi_\bo_\bn _\ba_\bn_\bd _\bv_\bi_\bs_\bu_\ba_\bl_\bi_\bz_\ba_\bt_\bi_\bo_\bn\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Pathview is a tool set for pathway based data integration and\n",
       "     visualization. It maps and renders user data on relevant pathway\n",
       "     graphs. All users need is to supply their gene or compound data\n",
       "     and specify the target pathway. Pathview automatically downloads\n",
       "     the pathway graph data, parses the data file, maps user data to\n",
       "     the pathway, and render pathway graph with the mapped data.\n",
       "     Pathview generates both native KEGG view and Graphviz views for\n",
       "     pathways. keggview.native and keggview.graph are the two viewer\n",
       "     functions, and pathview is the main function providing a unified\n",
       "     interface to downloader, parser, mapper and viewer functions.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     pathview(gene.data = NULL, cpd.data = NULL, pathway.id,\n",
       "     species = \"hsa\", kegg.dir = \".\", cpd.idtype = \"kegg\", gene.idtype =\n",
       "     \"entrez\", gene.annotpkg = NULL, min.nnodes = 3, kegg.native = TRUE,\n",
       "     map.null = TRUE, expand.node = FALSE, split.group = FALSE, map.symbol =\n",
       "     TRUE, map.cpdname = TRUE, node.sum = \"sum\", discrete=list(gene=FALSE,\n",
       "     cpd=FALSE), limit = list(gene = 1, cpd = 1), bins = list(gene = 10, cpd\n",
       "     = 10), both.dirs = list(gene = T, cpd = T), trans.fun = list(gene =\n",
       "     NULL, cpd = NULL), low = list(gene = \"green\", cpd = \"blue\"), mid =\n",
       "     list(gene = \"gray\", cpd = \"gray\"), high = list(gene = \"red\", cpd =\n",
       "     \"yellow\"), na.col = \"transparent\", ...)\n",
       "     \n",
       "     keggview.native(plot.data.gene = NULL, plot.data.cpd = NULL,\n",
       "     cols.ts.gene = NULL, cols.ts.cpd = NULL, node.data, pathway.name,\n",
       "     out.suffix = \"pathview\", kegg.dir = \".\", multi.state=TRUE, match.data =\n",
       "     TRUE, same.layer = TRUE, res = 300, cex = 0.25, discrete =\n",
       "     list(gene=FALSE, cpd=FALSE), limit= list(gene = 1, cpd = 1), bins =\n",
       "     list(gene = 10, cpd = 10), both.dirs =list(gene = T, cpd = T), low =\n",
       "     list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd =\n",
       "     \"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "     \"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "     \"x\", key.pos = \"topright\", ...)\n",
       "     \n",
       "     keggview.graph(plot.data.gene = NULL, plot.data.cpd = NULL, cols.ts.gene\n",
       "     = NULL, cols.ts.cpd = NULL, node.data, path.graph, pathway.name,\n",
       "     out.suffix = \"pathview\", pdf.size = c(7, 7), multi.state=TRUE,\n",
       "     same.layer = TRUE, match.data = TRUE, rankdir = c(\"LR\", \"TB\")[1],\n",
       "     is.signal = TRUE, split.group = F, afactor = 1, text.width = 15, cex =\n",
       "     0.5, map.cpdname = FALSE, cpd.lab.offset = 1.0,\n",
       "     discrete=list(gene=FALSE, cpd=FALSE), limit = list(gene = 1, cpd = 1),\n",
       "     bins = list(gene = 10, cpd = 10), both.dirs = list(gene = T, cpd = T),\n",
       "     low = list(gene = \"green\", cpd = \"blue\"), mid = list(gene = \"gray\", cpd\n",
       "     = \"gray\"), high = list(gene = \"red\", cpd = \"yellow\"), na.col =\n",
       "     \"transparent\", new.signature = TRUE, plot.col.key = TRUE, key.align =\n",
       "     \"x\", key.pos = \"topright\", sign.pos = \"bottomright\", ...)  \n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "gene.data: either vector (single sample) or a matrix-like data\n",
       "          (multiple sample). Vector should be numeric with gene IDs as\n",
       "          names or it may also be character of gene IDs. Character\n",
       "          vector is treated as discrete or count data. Matrix-like data\n",
       "          structure has genes as rows and samples as columns. Row names\n",
       "          should be gene IDs. Here gene ID is a generic concepts,\n",
       "          including multiple types of gene, transcript and protein\n",
       "          uniquely mappable to KEGG gene IDs. KEGG ortholog IDs are\n",
       "          also treated as gene IDs as to handle metagenomic data. Check\n",
       "          details for mappable ID types. Default gene.data=NULL.\n",
       "\n",
       "          numeric, character, continuous\n",
       "\n",
       "cpd.data: the same as gene.data, excpet named with IDs mappable to KEGG\n",
       "          compound IDs. Over 20 types of IDs included in CHEMBL\n",
       "          database can be used here. Check details for mappable ID\n",
       "          types. Default cpd.data=NULL. Note that gene.data and\n",
       "          cpd.data can't be NULL simultaneously.\n",
       "\n",
       "pathway.id: character vector, the KEGG pathway ID(s), usually 5 digit,\n",
       "          may also include the 3 letter KEGG species code.\n",
       "\n",
       " species: character, either the kegg code, scientific name or the\n",
       "          common name of the target species. This applies to both\n",
       "          pathway and gene.data or cpd.data. When KEGG ortholog pathway\n",
       "          is considered, species=\"ko\". Default species=\"hsa\", it is\n",
       "          equivalent to use either \"Homo sapiens\" (scientific name) or\n",
       "          \"human\" (common name).\n",
       "\n",
       "kegg.dir: character, the directory of KEGG pathway data file (.xml) and\n",
       "          image file (.png). Users may supply their own data files in\n",
       "          the same format and naming convention of KEGG's (species code\n",
       "          + pathway id, e.g. hsa04110.xml, hsa04110.png etc) in this\n",
       "          directory. Default kegg.dir=\".\" (current working directory).\n",
       "\n",
       "cpd.idtype: character, ID type used for the cpd.data. Default\n",
       "          cpd.idtype=\"kegg\" (include compound, glycan and drug\n",
       "          accessions).\n",
       "\n",
       "gene.idtype: character, ID type used for the gene.data, case\n",
       "          insensitive. Default gene.idtype=\"entrez\", i.e. Entrez Gene,\n",
       "          which are the primary KEGG gene ID for many common model\n",
       "          organisms. For other species, gene.idtype should be set to\n",
       "          \"KEGG\" as KEGG use other types of gene IDs. For the common\n",
       "          model organisms (to check the list, do: ‘data(bods); bods’),\n",
       "          you may also specify other types of valid IDs. To check the\n",
       "          ID list, do: ‘data(gene.idtype.list); gene.idtype.list’.\n",
       "\n",
       "gene.annotpkg: character, the name of the annotation package to use for\n",
       "          mapping between other gene ID types including symbols and\n",
       "          Entrez gene ID. Default gene.annotpkg=NULL.\n",
       "\n",
       "min.nnodes: integer, minimal number of nodes of type \"gene\",\"enzyme\",\n",
       "          \"compound\" or \"ortholog\" for a pathway to be considered.\n",
       "          Default min.nnodes=3.\n",
       "\n",
       "kegg.native: logical, whether to render pathway graph as native KEGG\n",
       "          graph (.png) or using graphviz layout engine (.pdf). Default\n",
       "          kegg.native=TRUE.\n",
       "\n",
       "map.null: logical, whether to map the NULL gene.data or cpd.data to\n",
       "          pathway. When NULL data are mapped, the gene or compound\n",
       "          nodes in the pathway will be rendered as actually mapped\n",
       "          nodes, except with NA-valued color. When NULL data are not\n",
       "          mapped, the nodes are rendered as unmapped nodes. This\n",
       "          argument mainly affects native KEGG graph view, i.e. when\n",
       "          kegg.native=TRUE. Default map.null=TRUE.\n",
       "\n",
       "expand.node: logical, whether the multiple-gene nodes are expanded into\n",
       "          single-gene nodes. Each expanded single-gene nodes inherits\n",
       "          all edges from the original multiple-gene node. This option\n",
       "          only affects graphviz graph view, i.e. when\n",
       "          kegg.native=FALSE. This option is not effective for most\n",
       "          metabolic pathways where it conflits with converting\n",
       "          reactions to edges. Default expand.node=FLASE.\n",
       "\n",
       "split.group: logical, whether split node groups are split to individual\n",
       "          nodes. Each split member nodes inherits all edges from the\n",
       "          node group. This option only affects graphviz graph view,\n",
       "          i.e. when kegg.native=FALSE. This option also effects most\n",
       "          metabolic pathways even without group nodes defined\n",
       "          orginally. For these pathways, genes involved in the same\n",
       "          reaction are grouped automatically when converting reactions\n",
       "          to edges unless split.group=TRUE. d split.group=FLASE.\n",
       "\n",
       "map.symbol: logical, whether map gene IDs to symbols for gene node\n",
       "          labels or use the graphic name from the KGML file. This\n",
       "          option is only effective for kegg.native=FALSE or\n",
       "          same.layer=FALSE when kegg.native=TRUE. For same.layer=TRUE\n",
       "          when kegg.native=TRUE, the native KEGG labels will be kept.\n",
       "          Default map.symbol=TRUE.\n",
       "\n",
       "map.cpdname: logical, whether map compound IDs to formal names for\n",
       "          compound node labels or use the graphic name from the KGML\n",
       "          file (KEGG compound accessions). This option is only\n",
       "          effective for kegg.native=FALSE. When kegg.native=TRUE, the\n",
       "          native KEGG labels will be kept. Default map.cpdname=TRUE.\n",
       "\n",
       "node.sum: character, the method name to calculate node summary given\n",
       "          that multiple genes or compounds are mapped to it. Poential\n",
       "          options include \"sum\",\"mean\", \"median\", \"max\", \"max.abs\" and\n",
       "          \"random\". Default node.sum=\"sum\".\n",
       "\n",
       "discrete: a list of two logical elements with \"gene\" and \"cpd\" as the\n",
       "          names. This argument tells whether gene.data or cpd.data\n",
       "          should be treated as discrete. Default\n",
       "          dsicrete=list(gene=FALSE, cpd=FALSE), i.e. both data should\n",
       "          be treated as continuous.\n",
       "\n",
       "   limit: a list of two numeric elements with \"gene\" and \"cpd\" as the\n",
       "          names. This argument specifies the limit values for gene.data\n",
       "          and cpd.data when converting them to pseudo colors. Each\n",
       "          element of the list could be of length 1 or 2. Length 1\n",
       "          suggests discrete data or 1 directional (positive-valued)\n",
       "          data, or the absolute limit for 2 directional data. Length 2\n",
       "          suggests 2 directional data. Default limit=list(gene=1,\n",
       "          cpd=1).\n",
       "\n",
       "    bins: a list of two integer elements with \"gene\" and \"cpd\" as the\n",
       "          names. This argument specifies the number of levels or bins\n",
       "          for gene.data and cpd.data when converting them to pseudo\n",
       "          colors. Default limit=list(gene=10, cpd=10).\n",
       "\n",
       "both.dirs: a list of two logical elements with \"gene\" and \"cpd\" as the\n",
       "          names. This argument specifies whether gene.data and cpd.data\n",
       "          are 1 directional or 2 directional data when converting them\n",
       "          to pseudo colors. Default limit=list(gene=TRUE, cpd=TRUE).\n",
       "\n",
       "trans.fun: a list of two function (not character) elements with \"gene\"\n",
       "          and \"cpd\" as the names. This argument specifies whether and\n",
       "          how gene.data and cpd.data are transformed. Examples are\n",
       "          ‘log’, ‘abs’ or users' own functions. Default\n",
       "          limit=list(gene=NULL, cpd=NULL).\n",
       "\n",
       "low, mid, high: each is a list of two colors with \"gene\" and \"cpd\" as\n",
       "          the names. This argument specifies the color spectra to code\n",
       "          gene.data and cpd.data. When data are 1 directional (TRUE\n",
       "          value in both.dirs), only mid and high are used to specify\n",
       "          the color spectra. Default spectra (low-mid-high)\n",
       "          \"green\"-\"gray\"-\"red\" and \"blue\"-\"gray\"-\"yellow\" are used for\n",
       "          gene.data and cpd.data respectively.  The values for 'low,\n",
       "          mid, high' can be given as color names ('red'), plot color\n",
       "          index (2=red), and HTML-style RGB, (\"\\#FF0000\"=red).\n",
       "\n",
       "  na.col: color used for NA's or missing values in gene.data and\n",
       "          cpd.data. d na.col=\"transparent\".\n",
       "\n",
       "     ...: extra arguments passed to keggview.native or keggview.graph\n",
       "          function.\n",
       "     special arguments for keggview.native or keggview.graph function.\n",
       "\n",
       "plot.data.gene: data.frame returned by node.map function for rendering\n",
       "          mapped gene nodes, including node name, type, positions (x,\n",
       "          y), sizes (width, height), and mapped gene.data. This data is\n",
       "          also used as input for pseduo-color coding through node.color\n",
       "          function. Default plot.data.gene=NULL.\n",
       "\n",
       "plot.data.cpd: same as plot.data.gene function, except for mapped\n",
       "          compound node data. d plot.data.cpd=NULL. Default\n",
       "          plot.data.cpd=NULL. Note that plot.data.gene and\n",
       "          plot.data.cpd can't be NULL simultaneously.\n",
       "\n",
       "cols.ts.gene: vector or matrix of colors returned by node.color\n",
       "          function for rendering gene.data. Dimensionality is the same\n",
       "          as the latter. Default cols.ts.gene=NULL.\n",
       "\n",
       "cols.ts.cpd: same as cols.ts.gene, except corresponding to cpd.data. d\n",
       "          cols.ts.cpd=NULL. Note that cols.ts.gene and cols.ts.cpd\n",
       "          plot.data.gene can't be NULL simultaneously.\n",
       "\n",
       "node.data: list returned by node.info function, which parse KGML file\n",
       "          directly or indirectly, and extract the node data.\n",
       "\n",
       "pathway.name: character, the full KEGG pathway name in the format of\n",
       "          3-letter species code with 5-digit pathway id, eg \"hsa04612\".\n",
       "\n",
       "out.suffix: character, the suffix to be added after the pathway name as\n",
       "          part of the output graph file. Sample names or column names\n",
       "          of the gene.data or cpd.data are also added when there are\n",
       "          multiple samples. Default out.suffix=\"pathview\".\n",
       "\n",
       "multi.state: logical, whether multiple states (samples or columns)\n",
       "          gene.data or cpd.data should be integrated and plotted in the\n",
       "          same graph. Default match.data=TRUE. In other words, gene or\n",
       "          compound nodes will be sliced into multiple pieces\n",
       "          corresponding to the number of states in the data.\n",
       "\n",
       "match.data: logical, whether the samples of gene.data and cpd.data are\n",
       "          paired. Default match.data=TRUE. When let sample sizes of\n",
       "          gene.data and cpd.data be m and n, when m>n, extra columns of\n",
       "          NA's (mapped to no color) will be added to cpd.data as to\n",
       "          make the sample size the same. This will result in the smae\n",
       "          number of slice in gene nodes and compound when\n",
       "          multi.state=TRUE.\n",
       "\n",
       "same.layer: logical, control plotting layers: 1) if node colors be\n",
       "          plotted in the same layer as the pathway graph when\n",
       "          kegg.native=TRUE, 2) if edge/node type legend be plotted in\n",
       "          the same page when kegg.native=FALSE.\n",
       "\n",
       "     res: The nominal resolution in ppi which will be recorded in the\n",
       "          bitmap file, if a positive integer.  Also used for 'units'\n",
       "          other than the default, and to convert points to pixels. This\n",
       "          argument is only effective when kegg.native=TRUE. Default\n",
       "          res=300.\n",
       "\n",
       "     cex: A numerical value giving the amount by which plotting text\n",
       "          and symbols should be scaled relative to the default 1.\n",
       "          Default cex=0.25 when kegg.native=TRUE, cex=0.5 when\n",
       "          kegg.native=FALSE.\n",
       "\n",
       "new.signature: logical, whether pathview signature is added to the\n",
       "          pathway graphs. Default new.signature=TRUE.\n",
       "\n",
       "plot.col.key: logical, whether color key is added to the pathway\n",
       "          graphs. Default plot.col.key= TRUE.\n",
       "\n",
       "key.align: character, controlling how the color keys are aligned when\n",
       "          both gene.data and cpd.data are not NULL. Potential values\n",
       "          are \"x\", aligned by x coordinates, and \"y\", aligned by y\n",
       "          coordinates. Default key.align=\"x\".\n",
       "\n",
       " key.pos: character, controlling the position of color key(s).\n",
       "          Potentail values are \"bottomleft\", \"bottomright\", \"topleft\"\n",
       "          and \"topright\". d key.pos=\"topright\".\n",
       "\n",
       "sign.pos: character, controlling the position of pathview signature.\n",
       "          Only effective when kegg.native=FALSE, Signature position is\n",
       "          fixed in place of the original KEGG signature when\n",
       "          kegg.native=TRUE. Potentail values are \"bottomleft\",\n",
       "          \"bottomright\", \"topleft\" and \"topright\". d\n",
       "          sign.pos=\"bottomright\".\n",
       "\n",
       "path.graph: a graph object parsed from KGML file, only effective when\n",
       "          kegg.native=FALSE.\n",
       "\n",
       "pdf.size: a numeric vector of length 2, giving the width and height of\n",
       "          the pathway graph pdf file. Note that pdf width increase by\n",
       "          half when same.layer=TRUE to accommodate legends. Only\n",
       "          effective when kegg.native=FALSE. Default pdf.size=c(7,7).\n",
       "\n",
       " rankdir: character, either \"LR\" (left to right) or \"TB\" (top to\n",
       "          bottom), specifying the pathway graph layout direction. Only\n",
       "          effective when kegg.native=FALSE. Default rank.dir=\"LR\".\n",
       "\n",
       "is.signal: logical, if the pathway is treated as a signaling pathway,\n",
       "          where all the unconnected nodes are dropped. This argument\n",
       "          also affect the graph layout type, i.e. \"dot\" for signals or\n",
       "          \"neato\" otherwise. Only effective when kegg.native=FALSE.\n",
       "          Default is.signal=TRUE.\n",
       "\n",
       " afactor: numeric, node amplifying factor. This argument is for node\n",
       "          size fine-tuning, its effect is subtler than expected. Only\n",
       "          effective when kegg.native=FALSE. Default afctor=1.\n",
       "\n",
       "text.width: numeric, specifying the line width for text wrap. Only\n",
       "          effective when kegg.native= FALSE. Default text.width=15\n",
       "          (characters).\n",
       "\n",
       "cpd.lab.offset: numeric, specifying how much compound labels should be\n",
       "          put above the default position or node center. This argument\n",
       "          is useful when map.cpdname=TRUE, i.e. compounds are labelled\n",
       "          by full name, which affects the look of compound nodes and\n",
       "          color. Only effective when kegg.native=FALSE. Default\n",
       "          cpd.lab.offset=1.0.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     Pathview maps and renders user data on relevant pathway graphs.\n",
       "     Pathview is a stand alone program for pathway based data\n",
       "     integration and visualization. It also seamlessly integrates with\n",
       "     pathway and functional analysis tools for large-scale and fully\n",
       "     automated analysis. Pathview provides strong support for data\n",
       "     Integration. It works with: 1) essentially all types of biological\n",
       "     data mappable to pathways, 2) over 10 types of gene or protein\n",
       "     IDs, and 20 types of compound or metabolite IDs, 3) pathways for\n",
       "     over 2000 species as well as KEGG orthology, 4) varoius data\n",
       "     attributes and formats, i.e. continuous/discrete data,\n",
       "     matrices/vectors, single/multiple samples etc. To see mappable\n",
       "     external gene/protein IDs do: ‘data(gene.idtype.list)’, to see\n",
       "     mappable external compound related IDs do: ‘data(rn.list)’;\n",
       "     names(rn.list). Pathview generates both native KEGG view and\n",
       "     Graphviz views for pathways. Currently only KEGG pathways are\n",
       "     implemented. Hopefully, pathways from Reactome, NCI and other\n",
       "     databases will be supported in the future.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     From viersion 1.9.3, pathview can accept either a single pathway\n",
       "     or multiple pathway ids. The result returned by pathview function\n",
       "     is a named list corresponding to the input pathway ids. Each\n",
       "     element (for each pathway itself is a named list, with 2 elements\n",
       "     (\"plot.data.gene\", \"plot.data.cpd\"). Both elements are data.frame\n",
       "     or NULL depends on the corresponding input data gene.data and\n",
       "     cpd.data. These data.frames record the plot data for mapped gene\n",
       "     or compound nodes: rows are mapped genes/compounds, columns are:\n",
       "\n",
       "kegg.names: standard KEGG IDs/Names for mapped nodes. It's Entrez Gene\n",
       "          ID or KEGG Compound Accessions.\n",
       "\n",
       "  labels: Node labels to be used when needed.\n",
       "\n",
       "all.mapped: All molecule (gene or compound) IDs mapped to this node.\n",
       "\n",
       "    type: node type, currently 4 types are supported: \"gene\",\"enzyme\",\n",
       "          \"compound\" and \"ortholog\".\n",
       "\n",
       "       x: x coordinate in the original KEGG pathway graph.\n",
       "\n",
       "       y: y coordinate in the original KEGG pathway graph.\n",
       "\n",
       "   width: node width in the original KEGG pathway graph.\n",
       "\n",
       "  height: node height in the original KEGG pathway graph.\n",
       "\n",
       "other columns : columns of the mapped gene/compound data and\n",
       "          corresponding pseudo-color codes for individual samples\n",
       "     The results returned by ‘keggview.native’ and codekeggview.graph\n",
       "     are both a list of graph plotting parameters. These are not\n",
       "     intended to be used externally.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Weijun Luo <luo_weijun@yahoo.com>\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     Luo, W. and Brouwer, C., Pathview: an R/Bioconductor package for\n",
       "     pathway based data integration and visualization. Bioinformatics,\n",
       "     2013, 29(14): 1830-1831, doi: 10.1093/bioinformatics/btt285\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘download.kegg’ the downloader, ‘node.info’ the parser, ‘node.map’\n",
       "     and ‘node.color’ the mapper.\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     #load data\n",
       "     data(gse16873.d)\n",
       "     data(demo.paths)\n",
       "     \n",
       "     #KEGG view: gene data only\n",
       "     i <- 1\n",
       "     pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "     demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "     kegg.native = TRUE)\n",
       "     str(pv.out)\n",
       "     head(pv.out$plot.data.gene)\n",
       "     #result PNG file in current directory\n",
       "     \n",
       "     #Graphviz view: gene data only\n",
       "     pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id =\n",
       "     demo.paths$sel.paths[i], species = \"hsa\", out.suffix = \"gse16873\",\n",
       "     kegg.native = FALSE, sign.pos = demo.paths$spos[i])\n",
       "     #result PDF file in current directory\n",
       "     \n",
       "     #KEGG view: both gene and compound data\n",
       "     sim.cpd.data=sim.mol.data(mol.type=\"cpd\", nmol=3000)\n",
       "     i <- 3\n",
       "     print(demo.paths$sel.paths[i])\n",
       "     pv.out <- pathview(gene.data = gse16873.d[, 1], cpd.data = sim.cpd.data,\n",
       "     pathway.id = demo.paths$sel.paths[i], species = \"hsa\", out.suffix =\n",
       "     \"gse16873.cpd\", keys.align = \"y\", kegg.native = TRUE, key.pos = demo.paths$kpos1[i])\n",
       "     str(pv.out)\n",
       "     head(pv.out$plot.data.cpd)\n",
       "     \n",
       "     #multiple states in one graph\n",
       "     set.seed(10)\n",
       "     sim.cpd.data2 = matrix(sample(sim.cpd.data, 18000, \n",
       "         replace = TRUE), ncol = 6)\n",
       "     pv.out <- pathview(gene.data = gse16873.d[, 1:3], \n",
       "         cpd.data = sim.cpd.data2[, 1:2], pathway.id = demo.paths$sel.paths[i], \n",
       "         species = \"hsa\", out.suffix = \"gse16873.cpd.3-2s\", keys.align = \"y\", \n",
       "         kegg.native = TRUE, match.data = FALSE, multi.state = TRUE, same.layer = TRUE)\n",
       "     str(pv.out)\n",
       "     head(pv.out$plot.data.cpd)\n",
       "     \n",
       "     #result PNG file in current directory\n",
       "     \n",
       "     ##more examples of pathview usages are shown in the vignette.\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?pathview::pathview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# graphviz view\n",
    "pv.out <- suppressWarnings(pathview(gene.data = gse16873.d[, 1:3],\n",
    "                                    cpd.data = sim.cpd.data2[, 1:2], \n",
    "                                    pathway.id = demo.paths$sel.paths[i],\n",
    "                                    species = \"hsa\", \n",
    "                                    out.suffix = \"gse16873.cpd.3-2s\", \n",
    "                                    keys.align = \"y\",\n",
    "                                    kegg.native = F, \n",
    "                                    match.data = F, \n",
    "                                    multi.state = T, \n",
    "                                    same.layer = T,\n",
    "                                    key.pos = demo.paths$kpos2[i], \n",
    "                                    sign.pos = demo.paths$spos[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot samples/states separately\n",
    "# Doesn't seem to print out images well\n",
    "pv.out <- suppressWarnings(pathview(gene.data = gse16873.d[, 1:3],\n",
    "                                    cpd.data = sim.cpd.data2[, 1:2], \n",
    "                                    pathway.id = demo.paths$sel.paths[i],\n",
    "                                    species = \"hsa\", \n",
    "                                    out.suffix = \"gse16873.cpd.3-2s\", \n",
    "                                    keys.align = \"y\",\n",
    "                                    kegg.native = T, \n",
    "                                    match.data = F, \n",
    "                                    multi.state = F, \n",
    "                                    same.layer = T))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# KEGG layer with 2 views. Loses the original KEGG gene labels (or EC numbers)\n",
    "pv.out <- suppressWarnings(pathview(gene.data = gse16873.d[, 1:3],\n",
    "                                    cpd.data = sim.cpd.data2[, 1:2], \n",
    "                                    pathway.id = demo.paths$sel.paths[i],\n",
    "                                    species = \"hsa\", \n",
    "                                    out.suffix = \"gse16873.cpd.3-2s.2layer\",\n",
    "                                    keys.align = \"y\", \n",
    "                                    kegg.native = T, \n",
    "                                    match.data = F, \n",
    "                                    multi.state = T,\n",
    "                                    same.layer = F))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feat. `GAGE`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load some datasets\n",
    "data(gse16873)\n",
    "hn <- grep('HN', colnames(gse16873), ignore.case =TRUEs) # indices of HN samples in colnames\n",
    "dcis <- grep('DCIS', colnames(gse16873), ignore.case =TRUE) # indice of DCIS samples in colnames\n",
    "data(kegg.gs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pw analysis with gage, gene data only\n",
    "gse16873.kegg.p <- gage(gse16873, \n",
    "                        gsets = kegg.gs, \n",
    "                        ref = hn, \n",
    "                        samp = dcis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare the differential expression data\n",
    "gse16873.d <- gagePrep(gse16873, ref = hn, samp = dcis)\n",
    "\n",
    "#equivalently, you can do simple subtraction for paired samples\n",
    "gse16873.d <- gse16873[,dcis]-gse16873[,hn]\n",
    "\n",
    "#select significant pathways and extract their IDs\n",
    "sel <- gse16873.kegg.p$greater[, \"q.val\"] < 0.1 & !is.na(gse16873.kegg.p$greater[,\"q.val\"])\n",
    "\n",
    "path.ids <- rownames(gse16873.kegg.p$greater)[sel]\n",
    "path.ids2 <- substr(path.ids[c(1, 2, 7)], 1, 8) # Grab paths with indices 1, 2 and 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pathview visualization\n",
    "pv.out.list <- sapply(path.ids2, function(pid) pathview(gene.data = gse16873.d[,1:2], \n",
    "                                                        pathway.id = pid, \n",
    "                                                        species = \"hsa\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x <- as_tibble(gse16873.kegg.p$greater, rownames = \"pw_name\") %>% drop_na() %>% filter(q.val<0.1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for gage {gage}\"><tr><td>gage {gage}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>\n",
       "GAGE (Generally Applicable Gene-set Enrichment) analysis\n",
       "</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Run GAGE analysis to infer gene sets (or pathways, functional groups\n",
       "etc) that are signficantly perturbed relative to all genes\n",
       "considered. GAGE is generally applicable to essentially all microarray\n",
       "dta independent of data attributes including sample size, experimental\n",
       "layout, study design, and all types of heterogeneity in data\n",
       "generation.\n",
       "</p>\n",
       "<p>gage is the main function; gagePrep is the functions for the initial\n",
       "data preparation, especially sample pairing; gageSum carries out the\n",
       "final meta-test summarization.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "gage(exprs, gsets, ref = NULL, samp = NULL, set.size = c(10, 500),\n",
       "same.dir = TRUE, compare = \"paired\", rank.test = FALSE, use.fold = TRUE,\n",
       "FDR.adj = TRUE, weights = NULL, full.table = FALSE, saaPrep = gagePrep,\n",
       "saaTest = gs.tTest, saaSum = gageSum, use.stouffer=TRUE, ...)\n",
       "\n",
       "gagePrep(exprs, ref = NULL, samp = NULL, same.dir = TRUE, compare =\n",
       "\"paired\", rank.test = FALSE, use.fold = TRUE, weights = NULL, full.table =\n",
       "FALSE, ...)\n",
       "\n",
       "gageSum(rawRes, ref = NULL, test4up = TRUE, same.dir =\n",
       "TRUE, compare = \"paired\", use.fold = TRUE, weights = NULL, full.table =\n",
       "FALSE, use.stouffer=TRUE, ...)\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>exprs</code></td>\n",
       "<td>\n",
       "\n",
       "<p>an expression matrix or matrix-like data structure, with genes as\n",
       "rows and samples as columns.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>gsets</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a named list, each element contains a gene set that is a character\n",
       "vector of gene IDs or symbols. For example, type <code>head(kegg.gs)</code>. A\n",
       "gene set can also be a &quot;smc&quot; object defined in PGSEA package. Please\n",
       "make sure that the same gene ID system is used for both <code>gsets</code>\n",
       "and <code>exprs</code>.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>ref</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a numeric vector of column numbers for the reference condition or\n",
       "phenotype (i.e. the control group) in the exprs data matrix. Default\n",
       "ref = NULL, all columns are considered as target experiments.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>samp</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a numeric vector of column numbers for the target condition or\n",
       "phenotype (i.e. the experiment group) in the exprs data\n",
       "matrix. Default samp = NULL, all columns other than ref are\n",
       "considered as target experiments.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>set.size</code></td>\n",
       "<td>\n",
       "\n",
       "<p>gene set size (number of genes) range to be considered for\n",
       "enrichment test. Tests for too small or too big gene sets are not\n",
       "robust statistically or informative biologically. Default to be\n",
       "set.size = c(10, 500).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>same.dir</code></td>\n",
       "<td>\n",
       "\n",
       "<p>boolean, whether to test for changes in a gene set toward a single direction\n",
       "(all genes up or down regulated) or changes towards both directions\n",
       "simultaneously. For experimentally derived gene sets, GO term\n",
       "groups, etc, coregulation is commonly the case, hence same.dir =\n",
       "TRUE (default); In KEGG, BioCarta pathways, genes frequently are not\n",
       "coregulated, hence it could be informative to let same.dir =\n",
       "FALSE. Although same.dir = TRUE could also be interesting for\n",
       "pathways.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>compare</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, which comparison scheme to be used: 'paired', 'unpaired',\n",
       "'1ongroup', 'as.group'. 'paired' is the default, ref and samp are of\n",
       "equal length and one-on-one paired by the original experimental\n",
       "design; 'as.group', group-on-group comparison between ref and samp;\n",
       "'unpaired' (used to be '1on1'), one-on-one comparison between all\n",
       "possible ref and samp combinations, although the original\n",
       "experimental design may not be one-on-one paired; '1ongroup',\n",
       "comparison between one samp column at a time vs the average of all\n",
       "ref columns.\n",
       "</p>\n",
       "<p>For PAGE-like analysis, the default is compare='as.group', which is\n",
       "the only option provided in the original PAGE method. All other\n",
       "comparison schemas are set here for direct comparison to gage.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>rank.test</code></td>\n",
       "<td>\n",
       "\n",
       "<p>rank.test: Boolean, whether do the optional rank based two-sample\n",
       "t-test (equivalent to the non-parametric Wilcoxon Mann-Whitney test)\n",
       "instead of parametric two-sample t-test. Default rank.test =\n",
       "FALSE. This argument should be used with respect to argument\n",
       "saaTest.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>use.fold</code></td>\n",
       "<td>\n",
       "\n",
       "<p>Boolean, whether to use fold changes or t-test statistics as per\n",
       "gene statistics. Default use.fold= TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>FDR.adj</code></td>\n",
       "<td>\n",
       "\n",
       "<p>Boolean, whether to do adjust for multiple testing as to control FDR\n",
       "(False dicovery rate). Default to be TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>weights</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a numeric vector to specify the weights assigned to pairs of\n",
       "ref-samp. This is needed for data with both technical replicates and\n",
       "biological replicates as to count for the different contributions\n",
       "from the two types of replicates. This argument is also useful in\n",
       "manually paring ref-samp for unpaired data, as in pairData function.\n",
       "function. Default to be NULL.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>full.table</code></td>\n",
       "<td>\n",
       "\n",
       "<p>This option is obsolete. Boolean, whether to output the full table of all individual p-values\n",
       "from the pairwise comparisons of ref and samp. Default to\n",
       "be FALSE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>saaPrep</code></td>\n",
       "<td>\n",
       "\n",
       "<p>function used for data preparation for single array based analysis,\n",
       "including sanity check, sample pairing, per gene statistics\n",
       "calculation etc. Default to be gagePrep, i.e. the default data\n",
       "preparation routine for gage analysis.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>saaTest</code></td>\n",
       "<td>\n",
       "\n",
       "<p>function used for gene set tests for single array based\n",
       "analysis. Default to be gs.tTest, which features a two-sample t-test\n",
       "for differential expression of gene sets. Other options includes:\n",
       "gs.zTest, using one-sample z-test as in PAGE, or gs.KSTest, using\n",
       "the non-parametric Kolmogorov-Smirnov tests as in GSEA. The two\n",
       "non-default options should only be used when rank.test = FALSE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>saaSum</code></td>\n",
       "<td>\n",
       "\n",
       "<p>function used for summarization of the results from single array\n",
       "analysis (i.e. pairwise comparison between ref and samp). This\n",
       "function should include a meta-test for a global p-value or summary\n",
       "statistis and a FDR adjustment for multi-testing issue. Default to\n",
       "be gageSum, i.e. the default data summarization routine for gage\n",
       "analysis.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>rawRes</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a named list, the raw results of gene set tests. Check the help\n",
       "information of gene set test functions like <code>gs.tTest</code> for\n",
       "details. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>test4up</code></td>\n",
       "<td>\n",
       "\n",
       "<p>boolean, whether to summarize the p-value results for up-regulation\n",
       "test (p.results) or not (ps.results for down-regulation).\n",
       "This argument is only needed when\n",
       "the argument same.dir=TRUE in the main <code>gage</code> function,\n",
       "i.e. when test for one-directional changes.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>use.stouffer</code></td>\n",
       "<td>\n",
       "\n",
       "<p>Boolean, whether to use Stouffer's method when summarizing\n",
       "individual p-values into a global p-value. Default to\n",
       "be TRUE. This default setting is recommended as to avoid the &quot;dual\n",
       "significance&quot;, i.e. a gene set is significant for both up-regulation\n",
       "and down-regulation tests simultaneously. Dual signficance  occurs\n",
       "sometimes for data with large sample size, due to extremely small\n",
       "p-values in a few pair-wise comparison. More robust p-value\n",
       "summarization using Stouffer's method is a\n",
       "important new feature added to GAGE since version 2.2.0\n",
       "(Bioconductor 2.8). This new argument is set as to provide a option\n",
       "to the original summarization based on Gamma distribution (FALSE).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>...</code></td>\n",
       "<td>\n",
       "\n",
       "<p>other arguments to be passed into the optional functions for\n",
       "saaPrep, saaTest and saaSum.\n",
       "</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>We proposed a single array analysis (i.e. the one-on-one comparison)\n",
       "approach with GAGE. Here we made single array analysis a general\n",
       "workflow for gene set analysis.  Single array analysis has 4 major\n",
       "steps: Step 1 sample pairing, Step 2 per gene tests, Step 3 gene set\n",
       "tests and Step 4 meta-test summarization. Correspondingly, this new main\n",
       "function, gage, is divided into 3 relatively independent modules. Module\n",
       "1 input preparation covers step 1-2 of single array analysis. Module 2\n",
       "corresponds to step 3 gene set test, and module 3 to step 4 meta-test\n",
       "summarization. These 3 modules become 3 argument functions to gage,\n",
       "saaPrep, saaTest and saaSum. The modulization made gage open to\n",
       "customization or plug-in routines at each steps and fully realize the\n",
       "general applicability of single array analysis. More examples will be\n",
       "included in a second vignette to demonstrate the customization with\n",
       "these modules.\n",
       "</p>\n",
       "<p>some important updates has been made to gage package since version 2.2.0\n",
       "(Bioconductor 2.8):\n",
       "First,  more robust p-value summarization using Stouffer's method\n",
       "through argument use.stouffer=TRUE. The original p-value\n",
       "summarization, i.e. negative log sum following a Gamma distribution as\n",
       "the Null hypothesis, may produce less stable global p-values for large\n",
       "or heterogenous datasets. In other words, the global p-value could be\n",
       "heavily affected by a small subset of extremely small individual\n",
       "p-values from pair-wise comparisons. Such sensitive global p-value\n",
       "leads to the &quot;dual signficance&quot; phenomenon. Dual-signficant means a gene set is called\n",
       "significant simultaneously in both 1-direction tests (up- and\n",
       "down-regulated). &quot;Dual\n",
       "signficance&quot; could be informative in revealing the sub-types or\n",
       "sub-classes in big clinical or disease studies, but may not be\n",
       "desirable in other cases.\n",
       "Second, output of gage function now includes the gene set test\n",
       "statistics from pair-wise comparisons for all proper gene sets. The\n",
       "output is always a named list now, with either 3 elements\n",
       "(&quot;greater&quot;, &quot;less&quot;, &quot;stats&quot;) for one-directional test or 2 elements\n",
       "(&quot;greater&quot;, &quot;stats&quot;) for two-directional test. \n",
       "Third, the individual p-value (and test statistics)from dependent pair-wise\n",
       "comparisions, i.e. comparisions between the same experiment vs\n",
       "different controls, are now summarized into a single value. In other\n",
       "words, the column number of individual p-values or statistics is\n",
       "always the same as the sample number in the experiment (or disease)\n",
       "group. This change made the argument value compare=&quot;1ongroup&quot;\n",
       "and argument full.table less useful. It also became easier to check the\n",
       "perturbations at gene-set level for individual samples. \n",
       "Fourth, whole gene-set level changes (either p-values or statistics)\n",
       "can now be visualized using heatmaps due to the third change above.\n",
       "Correspondingly, functions <code>sigGeneSet</code> and <code>gagePipe</code> have\n",
       "been revised to plot heatmaps for whole gene sets. \n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>The result returned by gage function is a named list, with either 3 elements\n",
       "(&quot;greater&quot;, &quot;less&quot;, &quot;stats&quot;) for one-directional test (same.dir =\n",
       "TRUE) or 2 elements (&quot;greater&quot;, &quot;stats&quot;) for two-directional test\n",
       "(same.dir = FALSE). Elements &quot;greater&quot; and &quot;less&quot; are two data\n",
       "matrices of the same structure, mainly the p-values, element &quot;stats&quot;\n",
       "contains the test statistics. Each data matrix here has gene sets as rows sorted by\n",
       "global p- or q-values. Test signficance or statistics columns include:\n",
       "</p>\n",
       "<table summary=\"R valueblock\">\n",
       "<tr valign=\"top\"><td><code>p.geomean </code></td>\n",
       "<td>\n",
       "<p>geometric mean of the individual p-values from\n",
       "multiple single array based gene set tests</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>stat.mean </code></td>\n",
       "<td>\n",
       "<p>mean of the individual statistics from multiple\n",
       "single array based gene set tests. Normally, its absoluate value\n",
       "measures the magnitude of gene-set level changes, and its sign\n",
       "indicates direction of the changes. When saaTest=gs.KSTest,\n",
       "stat.mean is always positive.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>p.val </code></td>\n",
       "<td>\n",
       "<p>gloal p-value or summary of the individual p-values from\n",
       "multiple single array based gene set tests. This is the default\n",
       "p-value being used.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>q.val </code></td>\n",
       "<td>\n",
       "<p>FDR q-value adjustment of the global p-value using\n",
       "the Benjamini &amp; Hochberg procedure implemented in multtest\n",
       "package. This is the default q-value being used.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>set.size </code></td>\n",
       "<td>\n",
       "<p>the effective gene set size, i.e. the number of genes\n",
       "included in the gene set test</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>other columns </code></td>\n",
       "<td>\n",
       "<p>columns of the individual p-values or\n",
       "statistics, each measures the gene set perturbation in a single\n",
       "experiment (vs its control or all controls, depends on the\n",
       "&quot;compare argument value)</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "<p>The result returned by <code>gagePrep</code> is a data matrix derived from \n",
       "<code>exprs</code>, but ready for column-wise gene est tests. In the\n",
       "matrix, genes are rows, and columns are the per gene test\n",
       "statistics from the ref-samp pairwise comparison.\n",
       "</p>\n",
       "<p>The result returned by <code>gageSum</code> is almost identical to the results\n",
       "of <code>gage</code> function, it is also a named list but has only 2\n",
       "elements, &quot;p.glob&quot; and &quot;results&quot;, with one round of test results. \n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Weijun Luo &lt;luo_weijun@yahoo.com&gt;\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>Luo, W., Friedman, M., Shedden K., Hankenson, K. and Woolf, P GAGE:\n",
       "Generally Applicable Gene Set Enrichment for Pathways Analysis. BMC\n",
       "Bioinformatics 2009, 10:161\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>gs.tTest</code>, <code>gs.zTest</code>, and\n",
       "<code>gs.KSTest</code> functions used for gene set test;\n",
       "<code>gagePipe</code> and <code>heter.gage</code> function used for\n",
       "multiple GAGE analysis in a batch or combined GAGE analysis on\n",
       "heterogeneous data\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "data(gse16873)\n",
       "cn=colnames(gse16873)\n",
       "hn=grep('HN',cn, ignore.case =TRUE)\n",
       "dcis=grep('DCIS',cn, ignore.case =TRUE)\n",
       "data(kegg.gs)\n",
       "data(go.gs)\n",
       "\n",
       "#kegg test for 1-directional changes\n",
       "gse16873.kegg.p &lt;- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis)\n",
       "#go.gs with the first 1000 entries as a fast example.\n",
       "gse16873.go.p &lt;- gage(gse16873, gsets = go.gs, \n",
       "    ref = hn, samp = dcis)\n",
       "str(gse16873.kegg.p)\n",
       "head(gse16873.kegg.p$greater)\n",
       "head(gse16873.kegg.p$less)\n",
       "head(gse16873.kegg.p$stats)\n",
       "#kegg test for 2-directional changes\n",
       "gse16873.kegg.2d.p &lt;- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis, same.dir = FALSE)\n",
       "head(gse16873.kegg.2d.p$greater)\n",
       "head(gse16873.kegg.2d.p$stats)\n",
       "\n",
       "###alternative ways to do GAGE analysis###\n",
       "#with unpaired samples\n",
       "gse16873.kegg.unpaired.p &lt;- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis, compare = \"unpaired\")\n",
       "\n",
       "#other options to tweak includes:\n",
       "#saaTest, use.fold, rank.test, etc. Check arguments section above for\n",
       "#details and the vignette for more examples.\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>gage</em> version 2.34.0 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{gage}{GAGE (Generally Applicable Gene-set Enrichment) analysis}{gage}\n",
       "\\aliasA{gagePrep}{gage}{gagePrep}\n",
       "\\aliasA{gageSum}{gage}{gageSum}\n",
       "\\keyword{htest}{gage}\n",
       "\\keyword{multivariate}{gage}\n",
       "\\keyword{manip}{gage}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Run GAGE analysis to infer gene sets (or pathways, functional groups\n",
       "etc) that are signficantly perturbed relative to all genes\n",
       "considered. GAGE is generally applicable to essentially all microarray\n",
       "dta independent of data attributes including sample size, experimental\n",
       "layout, study design, and all types of heterogeneity in data\n",
       "generation.\n",
       "\n",
       "gage is the main function; gagePrep is the functions for the initial\n",
       "data preparation, especially sample pairing; gageSum carries out the\n",
       "final meta-test summarization.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "gage(exprs, gsets, ref = NULL, samp = NULL, set.size = c(10, 500),\n",
       "same.dir = TRUE, compare = \"paired\", rank.test = FALSE, use.fold = TRUE,\n",
       "FDR.adj = TRUE, weights = NULL, full.table = FALSE, saaPrep = gagePrep,\n",
       "saaTest = gs.tTest, saaSum = gageSum, use.stouffer=TRUE, ...)\n",
       "\n",
       "gagePrep(exprs, ref = NULL, samp = NULL, same.dir = TRUE, compare =\n",
       "\"paired\", rank.test = FALSE, use.fold = TRUE, weights = NULL, full.table =\n",
       "FALSE, ...)\n",
       "\n",
       "gageSum(rawRes, ref = NULL, test4up = TRUE, same.dir =\n",
       "TRUE, compare = \"paired\", use.fold = TRUE, weights = NULL, full.table =\n",
       "FALSE, use.stouffer=TRUE, ...)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{exprs}] \n",
       "an expression matrix or matrix-like data structure, with genes as\n",
       "rows and samples as columns.\n",
       "\n",
       "\\item[\\code{gsets}] \n",
       "a named list, each element contains a gene set that is a character\n",
       "vector of gene IDs or symbols. For example, type \\code{head(kegg.gs)}. A\n",
       "gene set can also be a \"smc\" object defined in PGSEA package. Please\n",
       "make sure that the same gene ID system is used for both \\code{gsets}\n",
       "and \\code{exprs}.\n",
       "\n",
       "\\item[\\code{ref}] \n",
       "a numeric vector of column numbers for the reference condition or\n",
       "phenotype (i.e. the control group) in the exprs data matrix. Default\n",
       "ref = NULL, all columns are considered as target experiments.\n",
       "\n",
       "\\item[\\code{samp}] \n",
       "a numeric vector of column numbers for the target condition or\n",
       "phenotype (i.e. the experiment group) in the exprs data\n",
       "matrix. Default samp = NULL, all columns other than ref are\n",
       "considered as target experiments.\n",
       "\n",
       "\\item[\\code{set.size}] \n",
       "gene set size (number of genes) range to be considered for\n",
       "enrichment test. Tests for too small or too big gene sets are not\n",
       "robust statistically or informative biologically. Default to be\n",
       "set.size = c(10, 500).\n",
       "\n",
       "\\item[\\code{same.dir}] \n",
       "boolean, whether to test for changes in a gene set toward a single direction\n",
       "(all genes up or down regulated) or changes towards both directions\n",
       "simultaneously. For experimentally derived gene sets, GO term\n",
       "groups, etc, coregulation is commonly the case, hence same.dir =\n",
       "TRUE (default); In KEGG, BioCarta pathways, genes frequently are not\n",
       "coregulated, hence it could be informative to let same.dir =\n",
       "FALSE. Although same.dir = TRUE could also be interesting for\n",
       "pathways.\n",
       "\n",
       "\\item[\\code{compare}] \n",
       "character, which comparison scheme to be used: 'paired', 'unpaired',\n",
       "'1ongroup', 'as.group'. 'paired' is the default, ref and samp are of\n",
       "equal length and one-on-one paired by the original experimental\n",
       "design; 'as.group', group-on-group comparison between ref and samp;\n",
       "'unpaired' (used to be '1on1'), one-on-one comparison between all\n",
       "possible ref and samp combinations, although the original\n",
       "experimental design may not be one-on-one paired; '1ongroup',\n",
       "comparison between one samp column at a time vs the average of all\n",
       "ref columns.\n",
       "\n",
       "For PAGE-like analysis, the default is compare='as.group', which is\n",
       "the only option provided in the original PAGE method. All other\n",
       "comparison schemas are set here for direct comparison to gage.\n",
       "\n",
       "\\item[\\code{rank.test}] \n",
       "rank.test: Boolean, whether do the optional rank based two-sample\n",
       "t-test (equivalent to the non-parametric Wilcoxon Mann-Whitney test)\n",
       "instead of parametric two-sample t-test. Default rank.test =\n",
       "FALSE. This argument should be used with respect to argument\n",
       "saaTest.\n",
       "\n",
       "\\item[\\code{use.fold}] \n",
       "Boolean, whether to use fold changes or t-test statistics as per\n",
       "gene statistics. Default use.fold= TRUE.\n",
       "\n",
       "\\item[\\code{FDR.adj}] \n",
       "Boolean, whether to do adjust for multiple testing as to control FDR\n",
       "(False dicovery rate). Default to be TRUE.\n",
       "\n",
       "\\item[\\code{weights}] \n",
       "a numeric vector to specify the weights assigned to pairs of\n",
       "ref-samp. This is needed for data with both technical replicates and\n",
       "biological replicates as to count for the different contributions\n",
       "from the two types of replicates. This argument is also useful in\n",
       "manually paring ref-samp for unpaired data, as in pairData function.\n",
       "function. Default to be NULL.\n",
       "\n",
       "\\item[\\code{full.table}] \n",
       "This option is obsolete. Boolean, whether to output the full table of all individual p-values\n",
       "from the pairwise comparisons of ref and samp. Default to\n",
       "be FALSE.\n",
       "\n",
       "\\item[\\code{saaPrep}] \n",
       "function used for data preparation for single array based analysis,\n",
       "including sanity check, sample pairing, per gene statistics\n",
       "calculation etc. Default to be gagePrep, i.e. the default data\n",
       "preparation routine for gage analysis.\n",
       "\n",
       "\\item[\\code{saaTest}] \n",
       "function used for gene set tests for single array based\n",
       "analysis. Default to be gs.tTest, which features a two-sample t-test\n",
       "for differential expression of gene sets. Other options includes:\n",
       "gs.zTest, using one-sample z-test as in PAGE, or gs.KSTest, using\n",
       "the non-parametric Kolmogorov-Smirnov tests as in GSEA. The two\n",
       "non-default options should only be used when rank.test = FALSE.\n",
       "\n",
       "\\item[\\code{saaSum}] \n",
       "function used for summarization of the results from single array\n",
       "analysis (i.e. pairwise comparison between ref and samp). This\n",
       "function should include a meta-test for a global p-value or summary\n",
       "statistis and a FDR adjustment for multi-testing issue. Default to\n",
       "be gageSum, i.e. the default data summarization routine for gage\n",
       "analysis.\n",
       "\n",
       "\\item[\\code{rawRes}] \n",
       "a named list, the raw results of gene set tests. Check the help\n",
       "information of gene set test functions like \\code{gs.tTest} for\n",
       "details. \n",
       "\n",
       "\\item[\\code{test4up}] \n",
       "boolean, whether to summarize the p-value results for up-regulation\n",
       "test (p.results) or not (ps.results for down-regulation).\n",
       "This argument is only needed when\n",
       "the argument same.dir=TRUE in the main \\code{gage} function,\n",
       "i.e. when test for one-directional changes.\n",
       "\n",
       "\n",
       "\\item[\\code{use.stouffer}] \n",
       "Boolean, whether to use Stouffer's method when summarizing\n",
       "individual p-values into a global p-value. Default to\n",
       "be TRUE. This default setting is recommended as to avoid the \"dual\n",
       "significance\", i.e. a gene set is significant for both up-regulation\n",
       "and down-regulation tests simultaneously. Dual signficance  occurs\n",
       "sometimes for data with large sample size, due to extremely small\n",
       "p-values in a few pair-wise comparison. More robust p-value\n",
       "summarization using Stouffer's method is a\n",
       "important new feature added to GAGE since version 2.2.0\n",
       "(Bioconductor 2.8). This new argument is set as to provide a option\n",
       "to the original summarization based on Gamma distribution (FALSE).\n",
       "\n",
       "\\item[\\code{...}] \n",
       "other arguments to be passed into the optional functions for\n",
       "saaPrep, saaTest and saaSum.\n",
       "\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "We proposed a single array analysis (i.e. the one-on-one comparison)\n",
       "approach with GAGE. Here we made single array analysis a general\n",
       "workflow for gene set analysis.  Single array analysis has 4 major\n",
       "steps: Step 1 sample pairing, Step 2 per gene tests, Step 3 gene set\n",
       "tests and Step 4 meta-test summarization. Correspondingly, this new main\n",
       "function, gage, is divided into 3 relatively independent modules. Module\n",
       "1 input preparation covers step 1-2 of single array analysis. Module 2\n",
       "corresponds to step 3 gene set test, and module 3 to step 4 meta-test\n",
       "summarization. These 3 modules become 3 argument functions to gage,\n",
       "saaPrep, saaTest and saaSum. The modulization made gage open to\n",
       "customization or plug-in routines at each steps and fully realize the\n",
       "general applicability of single array analysis. More examples will be\n",
       "included in a second vignette to demonstrate the customization with\n",
       "these modules.\n",
       "\n",
       "some important updates has been made to gage package since version 2.2.0\n",
       "(Bioconductor 2.8):\n",
       "First,  more robust p-value summarization using Stouffer's method\n",
       "through argument use.stouffer=TRUE. The original p-value\n",
       "summarization, i.e. negative log sum following a Gamma distribution as\n",
       "the Null hypothesis, may produce less stable global p-values for large\n",
       "or heterogenous datasets. In other words, the global p-value could be\n",
       "heavily affected by a small subset of extremely small individual\n",
       "p-values from pair-wise comparisons. Such sensitive global p-value\n",
       "leads to the \"dual signficance\" phenomenon. Dual-signficant means a gene set is called\n",
       "significant simultaneously in both 1-direction tests (up- and\n",
       "down-regulated). \"Dual\n",
       "signficance\" could be informative in revealing the sub-types or\n",
       "sub-classes in big clinical or disease studies, but may not be\n",
       "desirable in other cases.\n",
       "Second, output of gage function now includes the gene set test\n",
       "statistics from pair-wise comparisons for all proper gene sets. The\n",
       "output is always a named list now, with either 3 elements\n",
       "(\"greater\", \"less\", \"stats\") for one-directional test or 2 elements\n",
       "(\"greater\", \"stats\") for two-directional test. \n",
       "Third, the individual p-value (and test statistics)from dependent pair-wise\n",
       "comparisions, i.e. comparisions between the same experiment vs\n",
       "different controls, are now summarized into a single value. In other\n",
       "words, the column number of individual p-values or statistics is\n",
       "always the same as the sample number in the experiment (or disease)\n",
       "group. This change made the argument value compare=\"1ongroup\"\n",
       "and argument full.table less useful. It also became easier to check the\n",
       "perturbations at gene-set level for individual samples. \n",
       "Fourth, whole gene-set level changes (either p-values or statistics)\n",
       "can now be visualized using heatmaps due to the third change above.\n",
       "Correspondingly, functions \\code{sigGeneSet} and \\code{gagePipe} have\n",
       "been revised to plot heatmaps for whole gene sets. \n",
       "\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "The result returned by gage function is a named list, with either 3 elements\n",
       "(\"greater\", \"less\", \"stats\") for one-directional test (same.dir =\n",
       "TRUE) or 2 elements (\"greater\", \"stats\") for two-directional test\n",
       "(same.dir = FALSE). Elements \"greater\" and \"less\" are two data\n",
       "matrices of the same structure, mainly the p-values, element \"stats\"\n",
       "contains the test statistics. Each data matrix here has gene sets as rows sorted by\n",
       "global p- or q-values. Test signficance or statistics columns include:\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{p.geomean }] geometric mean of the individual p-values from\n",
       "multiple single array based gene set tests\n",
       "\\item[\\code{stat.mean }] mean of the individual statistics from multiple\n",
       "single array based gene set tests. Normally, its absoluate value\n",
       "measures the magnitude of gene-set level changes, and its sign\n",
       "indicates direction of the changes. When saaTest=gs.KSTest,\n",
       "stat.mean is always positive.\n",
       "\\item[\\code{p.val }] gloal p-value or summary of the individual p-values from\n",
       "multiple single array based gene set tests. This is the default\n",
       "p-value being used.\n",
       "\\item[\\code{q.val }] FDR q-value adjustment of the global p-value using\n",
       "the Benjamini \\& Hochberg procedure implemented in multtest\n",
       "package. This is the default q-value being used.\n",
       "\\item[\\code{set.size }] the effective gene set size, i.e. the number of genes\n",
       "included in the gene set test\n",
       "\\item[\\code{other columns }] columns of the individual p-values or\n",
       "statistics, each measures the gene set perturbation in a single\n",
       "experiment (vs its control or all controls, depends on the\n",
       "\"compare argument value)\n",
       "\n",
       "\\end{ldescription}\n",
       "The result returned by \\code{gagePrep} is a data matrix derived from \n",
       "\\code{exprs}, but ready for column-wise gene est tests. In the\n",
       "matrix, genes are rows, and columns are the per gene test\n",
       "statistics from the ref-samp pairwise comparison.\n",
       "\n",
       "The result returned by \\code{gageSum} is almost identical to the results\n",
       "of \\code{gage} function, it is also a named list but has only 2\n",
       "elements, \"p.glob\" and \"results\", with one round of test results. \n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "Weijun Luo <luo\\_weijun@yahoo.com>\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "Luo, W., Friedman, M., Shedden K., Hankenson, K. and Woolf, P GAGE:\n",
       "Generally Applicable Gene Set Enrichment for Pathways Analysis. BMC\n",
       "Bioinformatics 2009, 10:161\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\code{\\LinkA{gs.tTest}{gs.tTest}}, \\code{\\LinkA{gs.zTest}{gs.zTest}}, and\n",
       "\\code{\\LinkA{gs.KSTest}{gs.KSTest}} functions used for gene set test;\n",
       "\\code{\\LinkA{gagePipe}{gagePipe}} and \\code{\\LinkA{heter.gage}{heter.gage}} function used for\n",
       "multiple GAGE analysis in a batch or combined GAGE analysis on\n",
       "heterogeneous data\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "data(gse16873)\n",
       "cn=colnames(gse16873)\n",
       "hn=grep('HN',cn, ignore.case =TRUE)\n",
       "dcis=grep('DCIS',cn, ignore.case =TRUE)\n",
       "data(kegg.gs)\n",
       "data(go.gs)\n",
       "\n",
       "#kegg test for 1-directional changes\n",
       "gse16873.kegg.p <- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis)\n",
       "#go.gs with the first 1000 entries as a fast example.\n",
       "gse16873.go.p <- gage(gse16873, gsets = go.gs, \n",
       "    ref = hn, samp = dcis)\n",
       "str(gse16873.kegg.p)\n",
       "head(gse16873.kegg.p$greater)\n",
       "head(gse16873.kegg.p$less)\n",
       "head(gse16873.kegg.p$stats)\n",
       "#kegg test for 2-directional changes\n",
       "gse16873.kegg.2d.p <- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis, same.dir = FALSE)\n",
       "head(gse16873.kegg.2d.p$greater)\n",
       "head(gse16873.kegg.2d.p$stats)\n",
       "\n",
       "###alternative ways to do GAGE analysis###\n",
       "#with unpaired samples\n",
       "gse16873.kegg.unpaired.p <- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis, compare = \"unpaired\")\n",
       "\n",
       "#other options to tweak includes:\n",
       "#saaTest, use.fold, rank.test, etc. Check arguments section above for\n",
       "#details and the vignette for more examples.\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "gage                   package:gage                    R Documentation\n",
       "\n",
       "_\bG_\bA_\bG_\bE (_\bG_\be_\bn_\be_\br_\ba_\bl_\bl_\by _\bA_\bp_\bp_\bl_\bi_\bc_\ba_\bb_\bl_\be _\bG_\be_\bn_\be-_\bs_\be_\bt _\bE_\bn_\br_\bi_\bc_\bh_\bm_\be_\bn_\bt) _\ba_\bn_\ba_\bl_\by_\bs_\bi_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Run GAGE analysis to infer gene sets (or pathways, functional\n",
       "     groups etc) that are signficantly perturbed relative to all genes\n",
       "     considered. GAGE is generally applicable to essentially all\n",
       "     microarray dta independent of data attributes including sample\n",
       "     size, experimental layout, study design, and all types of\n",
       "     heterogeneity in data generation.\n",
       "\n",
       "     gage is the main function; gagePrep is the functions for the\n",
       "     initial data preparation, especially sample pairing; gageSum\n",
       "     carries out the final meta-test summarization.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     gage(exprs, gsets, ref = NULL, samp = NULL, set.size = c(10, 500),\n",
       "     same.dir = TRUE, compare = \"paired\", rank.test = FALSE, use.fold = TRUE,\n",
       "     FDR.adj = TRUE, weights = NULL, full.table = FALSE, saaPrep = gagePrep,\n",
       "     saaTest = gs.tTest, saaSum = gageSum, use.stouffer=TRUE, ...)\n",
       "     \n",
       "     gagePrep(exprs, ref = NULL, samp = NULL, same.dir = TRUE, compare =\n",
       "     \"paired\", rank.test = FALSE, use.fold = TRUE, weights = NULL, full.table =\n",
       "     FALSE, ...)\n",
       "     \n",
       "     gageSum(rawRes, ref = NULL, test4up = TRUE, same.dir =\n",
       "     TRUE, compare = \"paired\", use.fold = TRUE, weights = NULL, full.table =\n",
       "     FALSE, use.stouffer=TRUE, ...)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "   exprs: an expression matrix or matrix-like data structure, with\n",
       "          genes as rows and samples as columns.\n",
       "\n",
       "   gsets: a named list, each element contains a gene set that is a\n",
       "          character vector of gene IDs or symbols. For example, type\n",
       "          ‘head(kegg.gs)’. A gene set can also be a \"smc\" object\n",
       "          defined in PGSEA package. Please make sure that the same gene\n",
       "          ID system is used for both ‘gsets’ and ‘exprs’.\n",
       "\n",
       "     ref: a numeric vector of column numbers for the reference\n",
       "          condition or phenotype (i.e. the control group) in the exprs\n",
       "          data matrix. Default ref = NULL, all columns are considered\n",
       "          as target experiments.\n",
       "\n",
       "    samp: a numeric vector of column numbers for the target condition\n",
       "          or phenotype (i.e. the experiment group) in the exprs data\n",
       "          matrix. Default samp = NULL, all columns other than ref are\n",
       "          considered as target experiments.\n",
       "\n",
       "set.size: gene set size (number of genes) range to be considered for\n",
       "          enrichment test. Tests for too small or too big gene sets are\n",
       "          not robust statistically or informative biologically. Default\n",
       "          to be set.size = c(10, 500).\n",
       "\n",
       "same.dir: boolean, whether to test for changes in a gene set toward a\n",
       "          single direction (all genes up or down regulated) or changes\n",
       "          towards both directions simultaneously. For experimentally\n",
       "          derived gene sets, GO term groups, etc, coregulation is\n",
       "          commonly the case, hence same.dir = TRUE (default); In KEGG,\n",
       "          BioCarta pathways, genes frequently are not coregulated,\n",
       "          hence it could be informative to let same.dir = FALSE.\n",
       "          Although same.dir = TRUE could also be interesting for\n",
       "          pathways.\n",
       "\n",
       " compare: character, which comparison scheme to be used: 'paired',\n",
       "          'unpaired', '1ongroup', 'as.group'. 'paired' is the default,\n",
       "          ref and samp are of equal length and one-on-one paired by the\n",
       "          original experimental design; 'as.group', group-on-group\n",
       "          comparison between ref and samp; 'unpaired' (used to be\n",
       "          '1on1'), one-on-one comparison between all possible ref and\n",
       "          samp combinations, although the original experimental design\n",
       "          may not be one-on-one paired; '1ongroup', comparison between\n",
       "          one samp column at a time vs the average of all ref columns.\n",
       "\n",
       "          For PAGE-like analysis, the default is compare='as.group',\n",
       "          which is the only option provided in the original PAGE\n",
       "          method. All other comparison schemas are set here for direct\n",
       "          comparison to gage.\n",
       "\n",
       "rank.test: rank.test: Boolean, whether do the optional rank based\n",
       "          two-sample t-test (equivalent to the non-parametric Wilcoxon\n",
       "          Mann-Whitney test) instead of parametric two-sample t-test.\n",
       "          Default rank.test = FALSE. This argument should be used with\n",
       "          respect to argument saaTest.\n",
       "\n",
       "use.fold: Boolean, whether to use fold changes or t-test statistics as\n",
       "          per gene statistics. Default use.fold= TRUE.\n",
       "\n",
       " FDR.adj: Boolean, whether to do adjust for multiple testing as to\n",
       "          control FDR (False dicovery rate). Default to be TRUE.\n",
       "\n",
       " weights: a numeric vector to specify the weights assigned to pairs of\n",
       "          ref-samp. This is needed for data with both technical\n",
       "          replicates and biological replicates as to count for the\n",
       "          different contributions from the two types of replicates.\n",
       "          This argument is also useful in manually paring ref-samp for\n",
       "          unpaired data, as in pairData function.  function. Default to\n",
       "          be NULL.\n",
       "\n",
       "full.table: This option is obsolete. Boolean, whether to output the\n",
       "          full table of all individual p-values from the pairwise\n",
       "          comparisons of ref and samp. Default to be FALSE.\n",
       "\n",
       " saaPrep: function used for data preparation for single array based\n",
       "          analysis, including sanity check, sample pairing, per gene\n",
       "          statistics calculation etc. Default to be gagePrep, i.e. the\n",
       "          default data preparation routine for gage analysis.\n",
       "\n",
       " saaTest: function used for gene set tests for single array based\n",
       "          analysis. Default to be gs.tTest, which features a two-sample\n",
       "          t-test for differential expression of gene sets. Other\n",
       "          options includes: gs.zTest, using one-sample z-test as in\n",
       "          PAGE, or gs.KSTest, using the non-parametric\n",
       "          Kolmogorov-Smirnov tests as in GSEA. The two non-default\n",
       "          options should only be used when rank.test = FALSE.\n",
       "\n",
       "  saaSum: function used for summarization of the results from single\n",
       "          array analysis (i.e. pairwise comparison between ref and\n",
       "          samp). This function should include a meta-test for a global\n",
       "          p-value or summary statistis and a FDR adjustment for\n",
       "          multi-testing issue. Default to be gageSum, i.e. the default\n",
       "          data summarization routine for gage analysis.\n",
       "\n",
       "  rawRes: a named list, the raw results of gene set tests. Check the\n",
       "          help information of gene set test functions like ‘gs.tTest’\n",
       "          for details.\n",
       "\n",
       " test4up: boolean, whether to summarize the p-value results for\n",
       "          up-regulation test (p.results) or not (ps.results for\n",
       "          down-regulation).  This argument is only needed when the\n",
       "          argument same.dir=TRUE in the main ‘gage’ function, i.e. when\n",
       "          test for one-directional changes.\n",
       "\n",
       "use.stouffer: Boolean, whether to use Stouffer's method when\n",
       "          summarizing individual p-values into a global p-value.\n",
       "          Default to be TRUE. This default setting is recommended as to\n",
       "          avoid the \"dual significance\", i.e. a gene set is significant\n",
       "          for both up-regulation and down-regulation tests\n",
       "          simultaneously. Dual signficance occurs sometimes for data\n",
       "          with large sample size, due to extremely small p-values in a\n",
       "          few pair-wise comparison. More robust p-value summarization\n",
       "          using Stouffer's method is a important new feature added to\n",
       "          GAGE since version 2.2.0 (Bioconductor 2.8). This new\n",
       "          argument is set as to provide a option to the original\n",
       "          summarization based on Gamma distribution (FALSE).\n",
       "\n",
       "     ...: other arguments to be passed into the optional functions for\n",
       "          saaPrep, saaTest and saaSum.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     We proposed a single array analysis (i.e. the one-on-one\n",
       "     comparison) approach with GAGE. Here we made single array analysis\n",
       "     a general workflow for gene set analysis.  Single array analysis\n",
       "     has 4 major steps: Step 1 sample pairing, Step 2 per gene tests,\n",
       "     Step 3 gene set tests and Step 4 meta-test summarization.\n",
       "     Correspondingly, this new main function, gage, is divided into 3\n",
       "     relatively independent modules. Module 1 input preparation covers\n",
       "     step 1-2 of single array analysis. Module 2 corresponds to step 3\n",
       "     gene set test, and module 3 to step 4 meta-test summarization.\n",
       "     These 3 modules become 3 argument functions to gage, saaPrep,\n",
       "     saaTest and saaSum. The modulization made gage open to\n",
       "     customization or plug-in routines at each steps and fully realize\n",
       "     the general applicability of single array analysis. More examples\n",
       "     will be included in a second vignette to demonstrate the\n",
       "     customization with these modules.\n",
       "\n",
       "     some important updates has been made to gage package since version\n",
       "     2.2.0 (Bioconductor 2.8): First, more robust p-value summarization\n",
       "     using Stouffer's method through argument use.stouffer=TRUE. The\n",
       "     original p-value summarization, i.e. negative log sum following a\n",
       "     Gamma distribution as the Null hypothesis, may produce less stable\n",
       "     global p-values for large or heterogenous datasets. In other\n",
       "     words, the global p-value could be heavily affected by a small\n",
       "     subset of extremely small individual p-values from pair-wise\n",
       "     comparisons. Such sensitive global p-value leads to the \"dual\n",
       "     signficance\" phenomenon. Dual-signficant means a gene set is\n",
       "     called significant simultaneously in both 1-direction tests (up-\n",
       "     and down-regulated). \"Dual signficance\" could be informative in\n",
       "     revealing the sub-types or sub-classes in big clinical or disease\n",
       "     studies, but may not be desirable in other cases.  Second, output\n",
       "     of gage function now includes the gene set test statistics from\n",
       "     pair-wise comparisons for all proper gene sets. The output is\n",
       "     always a named list now, with either 3 elements (\"greater\",\n",
       "     \"less\", \"stats\") for one-directional test or 2 elements\n",
       "     (\"greater\", \"stats\") for two-directional test.  Third, the\n",
       "     individual p-value (and test statistics)from dependent pair-wise\n",
       "     comparisions, i.e. comparisions between the same experiment vs\n",
       "     different controls, are now summarized into a single value. In\n",
       "     other words, the column number of individual p-values or\n",
       "     statistics is always the same as the sample number in the\n",
       "     experiment (or disease) group. This change made the argument value\n",
       "     compare=\"1ongroup\" and argument full.table less useful. It also\n",
       "     became easier to check the perturbations at gene-set level for\n",
       "     individual samples.  Fourth, whole gene-set level changes (either\n",
       "     p-values or statistics) can now be visualized using heatmaps due\n",
       "     to the third change above.  Correspondingly, functions\n",
       "     ‘sigGeneSet’ and ‘gagePipe’ have been revised to plot heatmaps for\n",
       "     whole gene sets.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     The result returned by gage function is a named list, with either\n",
       "     3 elements (\"greater\", \"less\", \"stats\") for one-directional test\n",
       "     (same.dir = TRUE) or 2 elements (\"greater\", \"stats\") for\n",
       "     two-directional test (same.dir = FALSE). Elements \"greater\" and\n",
       "     \"less\" are two data matrices of the same structure, mainly the\n",
       "     p-values, element \"stats\" contains the test statistics. Each data\n",
       "     matrix here has gene sets as rows sorted by global p- or q-values.\n",
       "     Test signficance or statistics columns include:\n",
       "\n",
       "p.geomean : geometric mean of the individual p-values from multiple\n",
       "          single array based gene set tests\n",
       "\n",
       "stat.mean : mean of the individual statistics from multiple single\n",
       "          array based gene set tests. Normally, its absoluate value\n",
       "          measures the magnitude of gene-set level changes, and its\n",
       "          sign indicates direction of the changes. When\n",
       "          saaTest=gs.KSTest, stat.mean is always positive.\n",
       "\n",
       "  p.val : gloal p-value or summary of the individual p-values from\n",
       "          multiple single array based gene set tests. This is the\n",
       "          default p-value being used.\n",
       "\n",
       "  q.val : FDR q-value adjustment of the global p-value using the\n",
       "          Benjamini & Hochberg procedure implemented in multtest\n",
       "          package. This is the default q-value being used.\n",
       "\n",
       "set.size : the effective gene set size, i.e. the number of genes\n",
       "          included in the gene set test\n",
       "\n",
       "other columns : columns of the individual p-values or statistics, each\n",
       "          measures the gene set perturbation in a single experiment (vs\n",
       "          its control or all controls, depends on the \"compare argument\n",
       "          value)\n",
       "     The result returned by ‘gagePrep’ is a data matrix derived from\n",
       "     ‘exprs’, but ready for column-wise gene est tests. In the matrix,\n",
       "     genes are rows, and columns are the per gene test statistics from\n",
       "     the ref-samp pairwise comparison.\n",
       "\n",
       "     The result returned by ‘gageSum’ is almost identical to the\n",
       "     results of ‘gage’ function, it is also a named list but has only 2\n",
       "     elements, \"p.glob\" and \"results\", with one round of test results.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Weijun Luo <luo_weijun@yahoo.com>\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     Luo, W., Friedman, M., Shedden K., Hankenson, K. and Woolf, P\n",
       "     GAGE: Generally Applicable Gene Set Enrichment for Pathways\n",
       "     Analysis. BMC Bioinformatics 2009, 10:161\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘gs.tTest’, ‘gs.zTest’, and ‘gs.KSTest’ functions used for gene\n",
       "     set test; ‘gagePipe’ and ‘heter.gage’ function used for multiple\n",
       "     GAGE analysis in a batch or combined GAGE analysis on\n",
       "     heterogeneous data\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     data(gse16873)\n",
       "     cn=colnames(gse16873)\n",
       "     hn=grep('HN',cn, ignore.case =TRUE)\n",
       "     dcis=grep('DCIS',cn, ignore.case =TRUE)\n",
       "     data(kegg.gs)\n",
       "     data(go.gs)\n",
       "     \n",
       "     #kegg test for 1-directional changes\n",
       "     gse16873.kegg.p <- gage(gse16873, gsets = kegg.gs, \n",
       "         ref = hn, samp = dcis)\n",
       "     #go.gs with the first 1000 entries as a fast example.\n",
       "     gse16873.go.p <- gage(gse16873, gsets = go.gs, \n",
       "         ref = hn, samp = dcis)\n",
       "     str(gse16873.kegg.p)\n",
       "     head(gse16873.kegg.p$greater)\n",
       "     head(gse16873.kegg.p$less)\n",
       "     head(gse16873.kegg.p$stats)\n",
       "     #kegg test for 2-directional changes\n",
       "     gse16873.kegg.2d.p <- gage(gse16873, gsets = kegg.gs, \n",
       "         ref = hn, samp = dcis, same.dir = FALSE)\n",
       "     head(gse16873.kegg.2d.p$greater)\n",
       "     head(gse16873.kegg.2d.p$stats)\n",
       "     \n",
       "     ###alternative ways to do GAGE analysis###\n",
       "     #with unpaired samples\n",
       "     gse16873.kegg.unpaired.p <- gage(gse16873, gsets = kegg.gs, \n",
       "         ref = hn, samp = dcis, compare = \"unpaired\")\n",
       "     \n",
       "     #other options to tweak includes:\n",
       "     #saaTest, use.fold, rank.test, etc. Check arguments section above for\n",
       "     #details and the vignette for more examples.\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?gage::gage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
