{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GAGE Scratch workbook\n",
    "\n",
    "Figuring out GAGE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"pacman\")\n",
    "\n",
    "pacman::p_load(\"gage\", \"gageData\", \"tidyverse\", \"globaltest\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load sample dataset\n",
    "data(gse16873)\n",
    "hn_idx = (1:6)*2-1 # hn = \"histologically normal\"\n",
    "dcis_idx = (1:6)*2 # dcis = \"ductal carcinoma in situ\"\n",
    "\n",
    "# Load (attach) gene set from kegg\n",
    "data(kegg.gs) # named list of 177 hsa pathways. Each value is a list of gene Ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for gage {gage}\"><tr><td>gage {gage}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>\n",
       "GAGE (Generally Applicable Gene-set Enrichment) analysis\n",
       "</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Run GAGE analysis to infer gene sets (or pathways, functional groups\n",
       "etc) that are signficantly perturbed relative to all genes\n",
       "considered. GAGE is generally applicable to essentially all microarray\n",
       "dta independent of data attributes including sample size, experimental\n",
       "layout, study design, and all types of heterogeneity in data\n",
       "generation.\n",
       "</p>\n",
       "<p>gage is the main function; gagePrep is the functions for the initial\n",
       "data preparation, especially sample pairing; gageSum carries out the\n",
       "final meta-test summarization.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "gage(exprs, gsets, ref = NULL, samp = NULL, set.size = c(10, 500),\n",
       "same.dir = TRUE, compare = \"paired\", rank.test = FALSE, use.fold = TRUE,\n",
       "FDR.adj = TRUE, weights = NULL, full.table = FALSE, saaPrep = gagePrep,\n",
       "saaTest = gs.tTest, saaSum = gageSum, use.stouffer=TRUE, ...)\n",
       "\n",
       "gagePrep(exprs, ref = NULL, samp = NULL, same.dir = TRUE, compare =\n",
       "\"paired\", rank.test = FALSE, use.fold = TRUE, weights = NULL, full.table =\n",
       "FALSE, ...)\n",
       "\n",
       "gageSum(rawRes, ref = NULL, test4up = TRUE, same.dir =\n",
       "TRUE, compare = \"paired\", use.fold = TRUE, weights = NULL, full.table =\n",
       "FALSE, use.stouffer=TRUE, ...)\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>exprs</code></td>\n",
       "<td>\n",
       "\n",
       "<p>an expression matrix or matrix-like data structure, with genes as\n",
       "rows and samples as columns.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>gsets</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a named list, each element contains a gene set that is a character\n",
       "vector of gene IDs or symbols. For example, type <code>head(kegg.gs)</code>. A\n",
       "gene set can also be a &quot;smc&quot; object defined in PGSEA package. Please\n",
       "make sure that the same gene ID system is used for both <code>gsets</code>\n",
       "and <code>exprs</code>.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>ref</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a numeric vector of column numbers for the reference condition or\n",
       "phenotype (i.e. the control group) in the exprs data matrix. Default\n",
       "ref = NULL, all columns are considered as target experiments.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>samp</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a numeric vector of column numbers for the target condition or\n",
       "phenotype (i.e. the experiment group) in the exprs data\n",
       "matrix. Default samp = NULL, all columns other than ref are\n",
       "considered as target experiments.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>set.size</code></td>\n",
       "<td>\n",
       "\n",
       "<p>gene set size (number of genes) range to be considered for\n",
       "enrichment test. Tests for too small or too big gene sets are not\n",
       "robust statistically or informative biologically. Default to be\n",
       "set.size = c(10, 500).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>same.dir</code></td>\n",
       "<td>\n",
       "\n",
       "<p>boolean, whether to test for changes in a gene set toward a single direction\n",
       "(all genes up or down regulated) or changes towards both directions\n",
       "simultaneously. For experimentally derived gene sets, GO term\n",
       "groups, etc, coregulation is commonly the case, hence same.dir =\n",
       "TRUE (default); In KEGG, BioCarta pathways, genes frequently are not\n",
       "coregulated, hence it could be informative to let same.dir =\n",
       "FALSE. Although same.dir = TRUE could also be interesting for\n",
       "pathways.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>compare</code></td>\n",
       "<td>\n",
       "\n",
       "<p>character, which comparison scheme to be used: 'paired', 'unpaired',\n",
       "'1ongroup', 'as.group'. 'paired' is the default, ref and samp are of\n",
       "equal length and one-on-one paired by the original experimental\n",
       "design; 'as.group', group-on-group comparison between ref and samp;\n",
       "'unpaired' (used to be '1on1'), one-on-one comparison between all\n",
       "possible ref and samp combinations, although the original\n",
       "experimental design may not be one-on-one paired; '1ongroup',\n",
       "comparison between one samp column at a time vs the average of all\n",
       "ref columns.\n",
       "</p>\n",
       "<p>For PAGE-like analysis, the default is compare='as.group', which is\n",
       "the only option provided in the original PAGE method. All other\n",
       "comparison schemas are set here for direct comparison to gage.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>rank.test</code></td>\n",
       "<td>\n",
       "\n",
       "<p>rank.test: Boolean, whether do the optional rank based two-sample\n",
       "t-test (equivalent to the non-parametric Wilcoxon Mann-Whitney test)\n",
       "instead of parametric two-sample t-test. Default rank.test =\n",
       "FALSE. This argument should be used with respect to argument\n",
       "saaTest.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>use.fold</code></td>\n",
       "<td>\n",
       "\n",
       "<p>Boolean, whether to use fold changes or t-test statistics as per\n",
       "gene statistics. Default use.fold= TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>FDR.adj</code></td>\n",
       "<td>\n",
       "\n",
       "<p>Boolean, whether to do adjust for multiple testing as to control FDR\n",
       "(False dicovery rate). Default to be TRUE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>weights</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a numeric vector to specify the weights assigned to pairs of\n",
       "ref-samp. This is needed for data with both technical replicates and\n",
       "biological replicates as to count for the different contributions\n",
       "from the two types of replicates. This argument is also useful in\n",
       "manually paring ref-samp for unpaired data, as in pairData function.\n",
       "function. Default to be NULL.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>full.table</code></td>\n",
       "<td>\n",
       "\n",
       "<p>This option is obsolete. Boolean, whether to output the full table of all individual p-values\n",
       "from the pairwise comparisons of ref and samp. Default to\n",
       "be FALSE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>saaPrep</code></td>\n",
       "<td>\n",
       "\n",
       "<p>function used for data preparation for single array based analysis,\n",
       "including sanity check, sample pairing, per gene statistics\n",
       "calculation etc. Default to be gagePrep, i.e. the default data\n",
       "preparation routine for gage analysis.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>saaTest</code></td>\n",
       "<td>\n",
       "\n",
       "<p>function used for gene set tests for single array based\n",
       "analysis. Default to be gs.tTest, which features a two-sample t-test\n",
       "for differential expression of gene sets. Other options includes:\n",
       "gs.zTest, using one-sample z-test as in PAGE, or gs.KSTest, using\n",
       "the non-parametric Kolmogorov-Smirnov tests as in GSEA. The two\n",
       "non-default options should only be used when rank.test = FALSE.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>saaSum</code></td>\n",
       "<td>\n",
       "\n",
       "<p>function used for summarization of the results from single array\n",
       "analysis (i.e. pairwise comparison between ref and samp). This\n",
       "function should include a meta-test for a global p-value or summary\n",
       "statistis and a FDR adjustment for multi-testing issue. Default to\n",
       "be gageSum, i.e. the default data summarization routine for gage\n",
       "analysis.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>rawRes</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a named list, the raw results of gene set tests. Check the help\n",
       "information of gene set test functions like <code>gs.tTest</code> for\n",
       "details. \n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>test4up</code></td>\n",
       "<td>\n",
       "\n",
       "<p>boolean, whether to summarize the p-value results for up-regulation\n",
       "test (p.results) or not (ps.results for down-regulation).\n",
       "This argument is only needed when\n",
       "the argument same.dir=TRUE in the main <code>gage</code> function,\n",
       "i.e. when test for one-directional changes.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>use.stouffer</code></td>\n",
       "<td>\n",
       "\n",
       "<p>Boolean, whether to use Stouffer's method when summarizing\n",
       "individual p-values into a global p-value. Default to\n",
       "be TRUE. This default setting is recommended as to avoid the &quot;dual\n",
       "significance&quot;, i.e. a gene set is significant for both up-regulation\n",
       "and down-regulation tests simultaneously. Dual signficance  occurs\n",
       "sometimes for data with large sample size, due to extremely small\n",
       "p-values in a few pair-wise comparison. More robust p-value\n",
       "summarization using Stouffer's method is a\n",
       "important new feature added to GAGE since version 2.2.0\n",
       "(Bioconductor 2.8). This new argument is set as to provide a option\n",
       "to the original summarization based on Gamma distribution (FALSE).\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>...</code></td>\n",
       "<td>\n",
       "\n",
       "<p>other arguments to be passed into the optional functions for\n",
       "saaPrep, saaTest and saaSum.\n",
       "</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>We proposed a single array analysis (i.e. the one-on-one comparison)\n",
       "approach with GAGE. Here we made single array analysis a general\n",
       "workflow for gene set analysis.  Single array analysis has 4 major\n",
       "steps: Step 1 sample pairing, Step 2 per gene tests, Step 3 gene set\n",
       "tests and Step 4 meta-test summarization. Correspondingly, this new main\n",
       "function, gage, is divided into 3 relatively independent modules. Module\n",
       "1 input preparation covers step 1-2 of single array analysis. Module 2\n",
       "corresponds to step 3 gene set test, and module 3 to step 4 meta-test\n",
       "summarization. These 3 modules become 3 argument functions to gage,\n",
       "saaPrep, saaTest and saaSum. The modulization made gage open to\n",
       "customization or plug-in routines at each steps and fully realize the\n",
       "general applicability of single array analysis. More examples will be\n",
       "included in a second vignette to demonstrate the customization with\n",
       "these modules.\n",
       "</p>\n",
       "<p>some important updates has been made to gage package since version 2.2.0\n",
       "(Bioconductor 2.8):\n",
       "First,  more robust p-value summarization using Stouffer's method\n",
       "through argument use.stouffer=TRUE. The original p-value\n",
       "summarization, i.e. negative log sum following a Gamma distribution as\n",
       "the Null hypothesis, may produce less stable global p-values for large\n",
       "or heterogenous datasets. In other words, the global p-value could be\n",
       "heavily affected by a small subset of extremely small individual\n",
       "p-values from pair-wise comparisons. Such sensitive global p-value\n",
       "leads to the &quot;dual signficance&quot; phenomenon. Dual-signficant means a gene set is called\n",
       "significant simultaneously in both 1-direction tests (up- and\n",
       "down-regulated). &quot;Dual\n",
       "signficance&quot; could be informative in revealing the sub-types or\n",
       "sub-classes in big clinical or disease studies, but may not be\n",
       "desirable in other cases.\n",
       "Second, output of gage function now includes the gene set test\n",
       "statistics from pair-wise comparisons for all proper gene sets. The\n",
       "output is always a named list now, with either 3 elements\n",
       "(&quot;greater&quot;, &quot;less&quot;, &quot;stats&quot;) for one-directional test or 2 elements\n",
       "(&quot;greater&quot;, &quot;stats&quot;) for two-directional test. \n",
       "Third, the individual p-value (and test statistics)from dependent pair-wise\n",
       "comparisions, i.e. comparisions between the same experiment vs\n",
       "different controls, are now summarized into a single value. In other\n",
       "words, the column number of individual p-values or statistics is\n",
       "always the same as the sample number in the experiment (or disease)\n",
       "group. This change made the argument value compare=&quot;1ongroup&quot;\n",
       "and argument full.table less useful. It also became easier to check the\n",
       "perturbations at gene-set level for individual samples. \n",
       "Fourth, whole gene-set level changes (either p-values or statistics)\n",
       "can now be visualized using heatmaps due to the third change above.\n",
       "Correspondingly, functions <code>sigGeneSet</code> and <code>gagePipe</code> have\n",
       "been revised to plot heatmaps for whole gene sets. \n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>The result returned by gage function is a named list, with either 3 elements\n",
       "(&quot;greater&quot;, &quot;less&quot;, &quot;stats&quot;) for one-directional test (same.dir =\n",
       "TRUE) or 2 elements (&quot;greater&quot;, &quot;stats&quot;) for two-directional test\n",
       "(same.dir = FALSE). Elements &quot;greater&quot; and &quot;less&quot; are two data\n",
       "matrices of the same structure, mainly the p-values, element &quot;stats&quot;\n",
       "contains the test statistics. Each data matrix here has gene sets as rows sorted by\n",
       "global p- or q-values. Test signficance or statistics columns include:\n",
       "</p>\n",
       "<table summary=\"R valueblock\">\n",
       "<tr valign=\"top\"><td><code>p.geomean </code></td>\n",
       "<td>\n",
       "<p>geometric mean of the individual p-values from\n",
       "multiple single array based gene set tests</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>stat.mean </code></td>\n",
       "<td>\n",
       "<p>mean of the individual statistics from multiple\n",
       "single array based gene set tests. Normally, its absoluate value\n",
       "measures the magnitude of gene-set level changes, and its sign\n",
       "indicates direction of the changes. When saaTest=gs.KSTest,\n",
       "stat.mean is always positive.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>p.val </code></td>\n",
       "<td>\n",
       "<p>gloal p-value or summary of the individual p-values from\n",
       "multiple single array based gene set tests. This is the default\n",
       "p-value being used.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>q.val </code></td>\n",
       "<td>\n",
       "<p>FDR q-value adjustment of the global p-value using\n",
       "the Benjamini &amp; Hochberg procedure implemented in multtest\n",
       "package. This is the default q-value being used.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>set.size </code></td>\n",
       "<td>\n",
       "<p>the effective gene set size, i.e. the number of genes\n",
       "included in the gene set test</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>other columns </code></td>\n",
       "<td>\n",
       "<p>columns of the individual p-values or\n",
       "statistics, each measures the gene set perturbation in a single\n",
       "experiment (vs its control or all controls, depends on the\n",
       "&quot;compare argument value)</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "<p>The result returned by <code>gagePrep</code> is a data matrix derived from \n",
       "<code>exprs</code>, but ready for column-wise gene est tests. In the\n",
       "matrix, genes are rows, and columns are the per gene test\n",
       "statistics from the ref-samp pairwise comparison.\n",
       "</p>\n",
       "<p>The result returned by <code>gageSum</code> is almost identical to the results\n",
       "of <code>gage</code> function, it is also a named list but has only 2\n",
       "elements, &quot;p.glob&quot; and &quot;results&quot;, with one round of test results. \n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Weijun Luo &lt;luo_weijun@yahoo.com&gt;\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>Luo, W., Friedman, M., Shedden K., Hankenson, K. and Woolf, P GAGE:\n",
       "Generally Applicable Gene Set Enrichment for Pathways Analysis. BMC\n",
       "Bioinformatics 2009, 10:161\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>gs.tTest</code>, <code>gs.zTest</code>, and\n",
       "<code>gs.KSTest</code> functions used for gene set test;\n",
       "<code>gagePipe</code> and <code>heter.gage</code> function used for\n",
       "multiple GAGE analysis in a batch or combined GAGE analysis on\n",
       "heterogeneous data\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "data(gse16873)\n",
       "cn=colnames(gse16873)\n",
       "hn=grep('HN',cn, ignore.case =TRUE)\n",
       "dcis=grep('DCIS',cn, ignore.case =TRUE)\n",
       "data(kegg.gs)\n",
       "data(go.gs)\n",
       "\n",
       "#kegg test for 1-directional changes\n",
       "gse16873.kegg.p &lt;- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis)\n",
       "#go.gs with the first 1000 entries as a fast example.\n",
       "gse16873.go.p &lt;- gage(gse16873, gsets = go.gs, \n",
       "    ref = hn, samp = dcis)\n",
       "str(gse16873.kegg.p)\n",
       "head(gse16873.kegg.p$greater)\n",
       "head(gse16873.kegg.p$less)\n",
       "head(gse16873.kegg.p$stats)\n",
       "#kegg test for 2-directional changes\n",
       "gse16873.kegg.2d.p &lt;- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis, same.dir = FALSE)\n",
       "head(gse16873.kegg.2d.p$greater)\n",
       "head(gse16873.kegg.2d.p$stats)\n",
       "\n",
       "###alternative ways to do GAGE analysis###\n",
       "#with unpaired samples\n",
       "gse16873.kegg.unpaired.p &lt;- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis, compare = \"unpaired\")\n",
       "\n",
       "#other options to tweak includes:\n",
       "#saaTest, use.fold, rank.test, etc. Check arguments section above for\n",
       "#details and the vignette for more examples.\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>gage</em> version 2.34.0 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{gage}{GAGE (Generally Applicable Gene-set Enrichment) analysis}{gage}\n",
       "\\aliasA{gagePrep}{gage}{gagePrep}\n",
       "\\aliasA{gageSum}{gage}{gageSum}\n",
       "\\keyword{htest}{gage}\n",
       "\\keyword{multivariate}{gage}\n",
       "\\keyword{manip}{gage}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Run GAGE analysis to infer gene sets (or pathways, functional groups\n",
       "etc) that are signficantly perturbed relative to all genes\n",
       "considered. GAGE is generally applicable to essentially all microarray\n",
       "dta independent of data attributes including sample size, experimental\n",
       "layout, study design, and all types of heterogeneity in data\n",
       "generation.\n",
       "\n",
       "gage is the main function; gagePrep is the functions for the initial\n",
       "data preparation, especially sample pairing; gageSum carries out the\n",
       "final meta-test summarization.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "gage(exprs, gsets, ref = NULL, samp = NULL, set.size = c(10, 500),\n",
       "same.dir = TRUE, compare = \"paired\", rank.test = FALSE, use.fold = TRUE,\n",
       "FDR.adj = TRUE, weights = NULL, full.table = FALSE, saaPrep = gagePrep,\n",
       "saaTest = gs.tTest, saaSum = gageSum, use.stouffer=TRUE, ...)\n",
       "\n",
       "gagePrep(exprs, ref = NULL, samp = NULL, same.dir = TRUE, compare =\n",
       "\"paired\", rank.test = FALSE, use.fold = TRUE, weights = NULL, full.table =\n",
       "FALSE, ...)\n",
       "\n",
       "gageSum(rawRes, ref = NULL, test4up = TRUE, same.dir =\n",
       "TRUE, compare = \"paired\", use.fold = TRUE, weights = NULL, full.table =\n",
       "FALSE, use.stouffer=TRUE, ...)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{exprs}] \n",
       "an expression matrix or matrix-like data structure, with genes as\n",
       "rows and samples as columns.\n",
       "\n",
       "\\item[\\code{gsets}] \n",
       "a named list, each element contains a gene set that is a character\n",
       "vector of gene IDs or symbols. For example, type \\code{head(kegg.gs)}. A\n",
       "gene set can also be a \"smc\" object defined in PGSEA package. Please\n",
       "make sure that the same gene ID system is used for both \\code{gsets}\n",
       "and \\code{exprs}.\n",
       "\n",
       "\\item[\\code{ref}] \n",
       "a numeric vector of column numbers for the reference condition or\n",
       "phenotype (i.e. the control group) in the exprs data matrix. Default\n",
       "ref = NULL, all columns are considered as target experiments.\n",
       "\n",
       "\\item[\\code{samp}] \n",
       "a numeric vector of column numbers for the target condition or\n",
       "phenotype (i.e. the experiment group) in the exprs data\n",
       "matrix. Default samp = NULL, all columns other than ref are\n",
       "considered as target experiments.\n",
       "\n",
       "\\item[\\code{set.size}] \n",
       "gene set size (number of genes) range to be considered for\n",
       "enrichment test. Tests for too small or too big gene sets are not\n",
       "robust statistically or informative biologically. Default to be\n",
       "set.size = c(10, 500).\n",
       "\n",
       "\\item[\\code{same.dir}] \n",
       "boolean, whether to test for changes in a gene set toward a single direction\n",
       "(all genes up or down regulated) or changes towards both directions\n",
       "simultaneously. For experimentally derived gene sets, GO term\n",
       "groups, etc, coregulation is commonly the case, hence same.dir =\n",
       "TRUE (default); In KEGG, BioCarta pathways, genes frequently are not\n",
       "coregulated, hence it could be informative to let same.dir =\n",
       "FALSE. Although same.dir = TRUE could also be interesting for\n",
       "pathways.\n",
       "\n",
       "\\item[\\code{compare}] \n",
       "character, which comparison scheme to be used: 'paired', 'unpaired',\n",
       "'1ongroup', 'as.group'. 'paired' is the default, ref and samp are of\n",
       "equal length and one-on-one paired by the original experimental\n",
       "design; 'as.group', group-on-group comparison between ref and samp;\n",
       "'unpaired' (used to be '1on1'), one-on-one comparison between all\n",
       "possible ref and samp combinations, although the original\n",
       "experimental design may not be one-on-one paired; '1ongroup',\n",
       "comparison between one samp column at a time vs the average of all\n",
       "ref columns.\n",
       "\n",
       "For PAGE-like analysis, the default is compare='as.group', which is\n",
       "the only option provided in the original PAGE method. All other\n",
       "comparison schemas are set here for direct comparison to gage.\n",
       "\n",
       "\\item[\\code{rank.test}] \n",
       "rank.test: Boolean, whether do the optional rank based two-sample\n",
       "t-test (equivalent to the non-parametric Wilcoxon Mann-Whitney test)\n",
       "instead of parametric two-sample t-test. Default rank.test =\n",
       "FALSE. This argument should be used with respect to argument\n",
       "saaTest.\n",
       "\n",
       "\\item[\\code{use.fold}] \n",
       "Boolean, whether to use fold changes or t-test statistics as per\n",
       "gene statistics. Default use.fold= TRUE.\n",
       "\n",
       "\\item[\\code{FDR.adj}] \n",
       "Boolean, whether to do adjust for multiple testing as to control FDR\n",
       "(False dicovery rate). Default to be TRUE.\n",
       "\n",
       "\\item[\\code{weights}] \n",
       "a numeric vector to specify the weights assigned to pairs of\n",
       "ref-samp. This is needed for data with both technical replicates and\n",
       "biological replicates as to count for the different contributions\n",
       "from the two types of replicates. This argument is also useful in\n",
       "manually paring ref-samp for unpaired data, as in pairData function.\n",
       "function. Default to be NULL.\n",
       "\n",
       "\\item[\\code{full.table}] \n",
       "This option is obsolete. Boolean, whether to output the full table of all individual p-values\n",
       "from the pairwise comparisons of ref and samp. Default to\n",
       "be FALSE.\n",
       "\n",
       "\\item[\\code{saaPrep}] \n",
       "function used for data preparation for single array based analysis,\n",
       "including sanity check, sample pairing, per gene statistics\n",
       "calculation etc. Default to be gagePrep, i.e. the default data\n",
       "preparation routine for gage analysis.\n",
       "\n",
       "\\item[\\code{saaTest}] \n",
       "function used for gene set tests for single array based\n",
       "analysis. Default to be gs.tTest, which features a two-sample t-test\n",
       "for differential expression of gene sets. Other options includes:\n",
       "gs.zTest, using one-sample z-test as in PAGE, or gs.KSTest, using\n",
       "the non-parametric Kolmogorov-Smirnov tests as in GSEA. The two\n",
       "non-default options should only be used when rank.test = FALSE.\n",
       "\n",
       "\\item[\\code{saaSum}] \n",
       "function used for summarization of the results from single array\n",
       "analysis (i.e. pairwise comparison between ref and samp). This\n",
       "function should include a meta-test for a global p-value or summary\n",
       "statistis and a FDR adjustment for multi-testing issue. Default to\n",
       "be gageSum, i.e. the default data summarization routine for gage\n",
       "analysis.\n",
       "\n",
       "\\item[\\code{rawRes}] \n",
       "a named list, the raw results of gene set tests. Check the help\n",
       "information of gene set test functions like \\code{gs.tTest} for\n",
       "details. \n",
       "\n",
       "\\item[\\code{test4up}] \n",
       "boolean, whether to summarize the p-value results for up-regulation\n",
       "test (p.results) or not (ps.results for down-regulation).\n",
       "This argument is only needed when\n",
       "the argument same.dir=TRUE in the main \\code{gage} function,\n",
       "i.e. when test for one-directional changes.\n",
       "\n",
       "\n",
       "\\item[\\code{use.stouffer}] \n",
       "Boolean, whether to use Stouffer's method when summarizing\n",
       "individual p-values into a global p-value. Default to\n",
       "be TRUE. This default setting is recommended as to avoid the \"dual\n",
       "significance\", i.e. a gene set is significant for both up-regulation\n",
       "and down-regulation tests simultaneously. Dual signficance  occurs\n",
       "sometimes for data with large sample size, due to extremely small\n",
       "p-values in a few pair-wise comparison. More robust p-value\n",
       "summarization using Stouffer's method is a\n",
       "important new feature added to GAGE since version 2.2.0\n",
       "(Bioconductor 2.8). This new argument is set as to provide a option\n",
       "to the original summarization based on Gamma distribution (FALSE).\n",
       "\n",
       "\\item[\\code{...}] \n",
       "other arguments to be passed into the optional functions for\n",
       "saaPrep, saaTest and saaSum.\n",
       "\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "We proposed a single array analysis (i.e. the one-on-one comparison)\n",
       "approach with GAGE. Here we made single array analysis a general\n",
       "workflow for gene set analysis.  Single array analysis has 4 major\n",
       "steps: Step 1 sample pairing, Step 2 per gene tests, Step 3 gene set\n",
       "tests and Step 4 meta-test summarization. Correspondingly, this new main\n",
       "function, gage, is divided into 3 relatively independent modules. Module\n",
       "1 input preparation covers step 1-2 of single array analysis. Module 2\n",
       "corresponds to step 3 gene set test, and module 3 to step 4 meta-test\n",
       "summarization. These 3 modules become 3 argument functions to gage,\n",
       "saaPrep, saaTest and saaSum. The modulization made gage open to\n",
       "customization or plug-in routines at each steps and fully realize the\n",
       "general applicability of single array analysis. More examples will be\n",
       "included in a second vignette to demonstrate the customization with\n",
       "these modules.\n",
       "\n",
       "some important updates has been made to gage package since version 2.2.0\n",
       "(Bioconductor 2.8):\n",
       "First,  more robust p-value summarization using Stouffer's method\n",
       "through argument use.stouffer=TRUE. The original p-value\n",
       "summarization, i.e. negative log sum following a Gamma distribution as\n",
       "the Null hypothesis, may produce less stable global p-values for large\n",
       "or heterogenous datasets. In other words, the global p-value could be\n",
       "heavily affected by a small subset of extremely small individual\n",
       "p-values from pair-wise comparisons. Such sensitive global p-value\n",
       "leads to the \"dual signficance\" phenomenon. Dual-signficant means a gene set is called\n",
       "significant simultaneously in both 1-direction tests (up- and\n",
       "down-regulated). \"Dual\n",
       "signficance\" could be informative in revealing the sub-types or\n",
       "sub-classes in big clinical or disease studies, but may not be\n",
       "desirable in other cases.\n",
       "Second, output of gage function now includes the gene set test\n",
       "statistics from pair-wise comparisons for all proper gene sets. The\n",
       "output is always a named list now, with either 3 elements\n",
       "(\"greater\", \"less\", \"stats\") for one-directional test or 2 elements\n",
       "(\"greater\", \"stats\") for two-directional test. \n",
       "Third, the individual p-value (and test statistics)from dependent pair-wise\n",
       "comparisions, i.e. comparisions between the same experiment vs\n",
       "different controls, are now summarized into a single value. In other\n",
       "words, the column number of individual p-values or statistics is\n",
       "always the same as the sample number in the experiment (or disease)\n",
       "group. This change made the argument value compare=\"1ongroup\"\n",
       "and argument full.table less useful. It also became easier to check the\n",
       "perturbations at gene-set level for individual samples. \n",
       "Fourth, whole gene-set level changes (either p-values or statistics)\n",
       "can now be visualized using heatmaps due to the third change above.\n",
       "Correspondingly, functions \\code{sigGeneSet} and \\code{gagePipe} have\n",
       "been revised to plot heatmaps for whole gene sets. \n",
       "\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "The result returned by gage function is a named list, with either 3 elements\n",
       "(\"greater\", \"less\", \"stats\") for one-directional test (same.dir =\n",
       "TRUE) or 2 elements (\"greater\", \"stats\") for two-directional test\n",
       "(same.dir = FALSE). Elements \"greater\" and \"less\" are two data\n",
       "matrices of the same structure, mainly the p-values, element \"stats\"\n",
       "contains the test statistics. Each data matrix here has gene sets as rows sorted by\n",
       "global p- or q-values. Test signficance or statistics columns include:\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{p.geomean }] geometric mean of the individual p-values from\n",
       "multiple single array based gene set tests\n",
       "\\item[\\code{stat.mean }] mean of the individual statistics from multiple\n",
       "single array based gene set tests. Normally, its absoluate value\n",
       "measures the magnitude of gene-set level changes, and its sign\n",
       "indicates direction of the changes. When saaTest=gs.KSTest,\n",
       "stat.mean is always positive.\n",
       "\\item[\\code{p.val }] gloal p-value or summary of the individual p-values from\n",
       "multiple single array based gene set tests. This is the default\n",
       "p-value being used.\n",
       "\\item[\\code{q.val }] FDR q-value adjustment of the global p-value using\n",
       "the Benjamini \\& Hochberg procedure implemented in multtest\n",
       "package. This is the default q-value being used.\n",
       "\\item[\\code{set.size }] the effective gene set size, i.e. the number of genes\n",
       "included in the gene set test\n",
       "\\item[\\code{other columns }] columns of the individual p-values or\n",
       "statistics, each measures the gene set perturbation in a single\n",
       "experiment (vs its control or all controls, depends on the\n",
       "\"compare argument value)\n",
       "\n",
       "\\end{ldescription}\n",
       "The result returned by \\code{gagePrep} is a data matrix derived from \n",
       "\\code{exprs}, but ready for column-wise gene est tests. In the\n",
       "matrix, genes are rows, and columns are the per gene test\n",
       "statistics from the ref-samp pairwise comparison.\n",
       "\n",
       "The result returned by \\code{gageSum} is almost identical to the results\n",
       "of \\code{gage} function, it is also a named list but has only 2\n",
       "elements, \"p.glob\" and \"results\", with one round of test results. \n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "Weijun Luo <luo\\_weijun@yahoo.com>\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "Luo, W., Friedman, M., Shedden K., Hankenson, K. and Woolf, P GAGE:\n",
       "Generally Applicable Gene Set Enrichment for Pathways Analysis. BMC\n",
       "Bioinformatics 2009, 10:161\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\code{\\LinkA{gs.tTest}{gs.tTest}}, \\code{\\LinkA{gs.zTest}{gs.zTest}}, and\n",
       "\\code{\\LinkA{gs.KSTest}{gs.KSTest}} functions used for gene set test;\n",
       "\\code{\\LinkA{gagePipe}{gagePipe}} and \\code{\\LinkA{heter.gage}{heter.gage}} function used for\n",
       "multiple GAGE analysis in a batch or combined GAGE analysis on\n",
       "heterogeneous data\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "data(gse16873)\n",
       "cn=colnames(gse16873)\n",
       "hn=grep('HN',cn, ignore.case =TRUE)\n",
       "dcis=grep('DCIS',cn, ignore.case =TRUE)\n",
       "data(kegg.gs)\n",
       "data(go.gs)\n",
       "\n",
       "#kegg test for 1-directional changes\n",
       "gse16873.kegg.p <- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis)\n",
       "#go.gs with the first 1000 entries as a fast example.\n",
       "gse16873.go.p <- gage(gse16873, gsets = go.gs, \n",
       "    ref = hn, samp = dcis)\n",
       "str(gse16873.kegg.p)\n",
       "head(gse16873.kegg.p$greater)\n",
       "head(gse16873.kegg.p$less)\n",
       "head(gse16873.kegg.p$stats)\n",
       "#kegg test for 2-directional changes\n",
       "gse16873.kegg.2d.p <- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis, same.dir = FALSE)\n",
       "head(gse16873.kegg.2d.p$greater)\n",
       "head(gse16873.kegg.2d.p$stats)\n",
       "\n",
       "###alternative ways to do GAGE analysis###\n",
       "#with unpaired samples\n",
       "gse16873.kegg.unpaired.p <- gage(gse16873, gsets = kegg.gs, \n",
       "    ref = hn, samp = dcis, compare = \"unpaired\")\n",
       "\n",
       "#other options to tweak includes:\n",
       "#saaTest, use.fold, rank.test, etc. Check arguments section above for\n",
       "#details and the vignette for more examples.\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "gage                   package:gage                    R Documentation\n",
       "\n",
       "_\bG_\bA_\bG_\bE (_\bG_\be_\bn_\be_\br_\ba_\bl_\bl_\by _\bA_\bp_\bp_\bl_\bi_\bc_\ba_\bb_\bl_\be _\bG_\be_\bn_\be-_\bs_\be_\bt _\bE_\bn_\br_\bi_\bc_\bh_\bm_\be_\bn_\bt) _\ba_\bn_\ba_\bl_\by_\bs_\bi_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Run GAGE analysis to infer gene sets (or pathways, functional\n",
       "     groups etc) that are signficantly perturbed relative to all genes\n",
       "     considered. GAGE is generally applicable to essentially all\n",
       "     microarray dta independent of data attributes including sample\n",
       "     size, experimental layout, study design, and all types of\n",
       "     heterogeneity in data generation.\n",
       "\n",
       "     gage is the main function; gagePrep is the functions for the\n",
       "     initial data preparation, especially sample pairing; gageSum\n",
       "     carries out the final meta-test summarization.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     gage(exprs, gsets, ref = NULL, samp = NULL, set.size = c(10, 500),\n",
       "     same.dir = TRUE, compare = \"paired\", rank.test = FALSE, use.fold = TRUE,\n",
       "     FDR.adj = TRUE, weights = NULL, full.table = FALSE, saaPrep = gagePrep,\n",
       "     saaTest = gs.tTest, saaSum = gageSum, use.stouffer=TRUE, ...)\n",
       "     \n",
       "     gagePrep(exprs, ref = NULL, samp = NULL, same.dir = TRUE, compare =\n",
       "     \"paired\", rank.test = FALSE, use.fold = TRUE, weights = NULL, full.table =\n",
       "     FALSE, ...)\n",
       "     \n",
       "     gageSum(rawRes, ref = NULL, test4up = TRUE, same.dir =\n",
       "     TRUE, compare = \"paired\", use.fold = TRUE, weights = NULL, full.table =\n",
       "     FALSE, use.stouffer=TRUE, ...)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "   exprs: an expression matrix or matrix-like data structure, with\n",
       "          genes as rows and samples as columns.\n",
       "\n",
       "   gsets: a named list, each element contains a gene set that is a\n",
       "          character vector of gene IDs or symbols. For example, type\n",
       "          ‘head(kegg.gs)’. A gene set can also be a \"smc\" object\n",
       "          defined in PGSEA package. Please make sure that the same gene\n",
       "          ID system is used for both ‘gsets’ and ‘exprs’.\n",
       "\n",
       "     ref: a numeric vector of column numbers for the reference\n",
       "          condition or phenotype (i.e. the control group) in the exprs\n",
       "          data matrix. Default ref = NULL, all columns are considered\n",
       "          as target experiments.\n",
       "\n",
       "    samp: a numeric vector of column numbers for the target condition\n",
       "          or phenotype (i.e. the experiment group) in the exprs data\n",
       "          matrix. Default samp = NULL, all columns other than ref are\n",
       "          considered as target experiments.\n",
       "\n",
       "set.size: gene set size (number of genes) range to be considered for\n",
       "          enrichment test. Tests for too small or too big gene sets are\n",
       "          not robust statistically or informative biologically. Default\n",
       "          to be set.size = c(10, 500).\n",
       "\n",
       "same.dir: boolean, whether to test for changes in a gene set toward a\n",
       "          single direction (all genes up or down regulated) or changes\n",
       "          towards both directions simultaneously. For experimentally\n",
       "          derived gene sets, GO term groups, etc, coregulation is\n",
       "          commonly the case, hence same.dir = TRUE (default); In KEGG,\n",
       "          BioCarta pathways, genes frequently are not coregulated,\n",
       "          hence it could be informative to let same.dir = FALSE.\n",
       "          Although same.dir = TRUE could also be interesting for\n",
       "          pathways.\n",
       "\n",
       " compare: character, which comparison scheme to be used: 'paired',\n",
       "          'unpaired', '1ongroup', 'as.group'. 'paired' is the default,\n",
       "          ref and samp are of equal length and one-on-one paired by the\n",
       "          original experimental design; 'as.group', group-on-group\n",
       "          comparison between ref and samp; 'unpaired' (used to be\n",
       "          '1on1'), one-on-one comparison between all possible ref and\n",
       "          samp combinations, although the original experimental design\n",
       "          may not be one-on-one paired; '1ongroup', comparison between\n",
       "          one samp column at a time vs the average of all ref columns.\n",
       "\n",
       "          For PAGE-like analysis, the default is compare='as.group',\n",
       "          which is the only option provided in the original PAGE\n",
       "          method. All other comparison schemas are set here for direct\n",
       "          comparison to gage.\n",
       "\n",
       "rank.test: rank.test: Boolean, whether do the optional rank based\n",
       "          two-sample t-test (equivalent to the non-parametric Wilcoxon\n",
       "          Mann-Whitney test) instead of parametric two-sample t-test.\n",
       "          Default rank.test = FALSE. This argument should be used with\n",
       "          respect to argument saaTest.\n",
       "\n",
       "use.fold: Boolean, whether to use fold changes or t-test statistics as\n",
       "          per gene statistics. Default use.fold= TRUE.\n",
       "\n",
       " FDR.adj: Boolean, whether to do adjust for multiple testing as to\n",
       "          control FDR (False dicovery rate). Default to be TRUE.\n",
       "\n",
       " weights: a numeric vector to specify the weights assigned to pairs of\n",
       "          ref-samp. This is needed for data with both technical\n",
       "          replicates and biological replicates as to count for the\n",
       "          different contributions from the two types of replicates.\n",
       "          This argument is also useful in manually paring ref-samp for\n",
       "          unpaired data, as in pairData function.  function. Default to\n",
       "          be NULL.\n",
       "\n",
       "full.table: This option is obsolete. Boolean, whether to output the\n",
       "          full table of all individual p-values from the pairwise\n",
       "          comparisons of ref and samp. Default to be FALSE.\n",
       "\n",
       " saaPrep: function used for data preparation for single array based\n",
       "          analysis, including sanity check, sample pairing, per gene\n",
       "          statistics calculation etc. Default to be gagePrep, i.e. the\n",
       "          default data preparation routine for gage analysis.\n",
       "\n",
       " saaTest: function used for gene set tests for single array based\n",
       "          analysis. Default to be gs.tTest, which features a two-sample\n",
       "          t-test for differential expression of gene sets. Other\n",
       "          options includes: gs.zTest, using one-sample z-test as in\n",
       "          PAGE, or gs.KSTest, using the non-parametric\n",
       "          Kolmogorov-Smirnov tests as in GSEA. The two non-default\n",
       "          options should only be used when rank.test = FALSE.\n",
       "\n",
       "  saaSum: function used for summarization of the results from single\n",
       "          array analysis (i.e. pairwise comparison between ref and\n",
       "          samp). This function should include a meta-test for a global\n",
       "          p-value or summary statistis and a FDR adjustment for\n",
       "          multi-testing issue. Default to be gageSum, i.e. the default\n",
       "          data summarization routine for gage analysis.\n",
       "\n",
       "  rawRes: a named list, the raw results of gene set tests. Check the\n",
       "          help information of gene set test functions like ‘gs.tTest’\n",
       "          for details.\n",
       "\n",
       " test4up: boolean, whether to summarize the p-value results for\n",
       "          up-regulation test (p.results) or not (ps.results for\n",
       "          down-regulation).  This argument is only needed when the\n",
       "          argument same.dir=TRUE in the main ‘gage’ function, i.e. when\n",
       "          test for one-directional changes.\n",
       "\n",
       "use.stouffer: Boolean, whether to use Stouffer's method when\n",
       "          summarizing individual p-values into a global p-value.\n",
       "          Default to be TRUE. This default setting is recommended as to\n",
       "          avoid the \"dual significance\", i.e. a gene set is significant\n",
       "          for both up-regulation and down-regulation tests\n",
       "          simultaneously. Dual signficance occurs sometimes for data\n",
       "          with large sample size, due to extremely small p-values in a\n",
       "          few pair-wise comparison. More robust p-value summarization\n",
       "          using Stouffer's method is a important new feature added to\n",
       "          GAGE since version 2.2.0 (Bioconductor 2.8). This new\n",
       "          argument is set as to provide a option to the original\n",
       "          summarization based on Gamma distribution (FALSE).\n",
       "\n",
       "     ...: other arguments to be passed into the optional functions for\n",
       "          saaPrep, saaTest and saaSum.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     We proposed a single array analysis (i.e. the one-on-one\n",
       "     comparison) approach with GAGE. Here we made single array analysis\n",
       "     a general workflow for gene set analysis.  Single array analysis\n",
       "     has 4 major steps: Step 1 sample pairing, Step 2 per gene tests,\n",
       "     Step 3 gene set tests and Step 4 meta-test summarization.\n",
       "     Correspondingly, this new main function, gage, is divided into 3\n",
       "     relatively independent modules. Module 1 input preparation covers\n",
       "     step 1-2 of single array analysis. Module 2 corresponds to step 3\n",
       "     gene set test, and module 3 to step 4 meta-test summarization.\n",
       "     These 3 modules become 3 argument functions to gage, saaPrep,\n",
       "     saaTest and saaSum. The modulization made gage open to\n",
       "     customization or plug-in routines at each steps and fully realize\n",
       "     the general applicability of single array analysis. More examples\n",
       "     will be included in a second vignette to demonstrate the\n",
       "     customization with these modules.\n",
       "\n",
       "     some important updates has been made to gage package since version\n",
       "     2.2.0 (Bioconductor 2.8): First, more robust p-value summarization\n",
       "     using Stouffer's method through argument use.stouffer=TRUE. The\n",
       "     original p-value summarization, i.e. negative log sum following a\n",
       "     Gamma distribution as the Null hypothesis, may produce less stable\n",
       "     global p-values for large or heterogenous datasets. In other\n",
       "     words, the global p-value could be heavily affected by a small\n",
       "     subset of extremely small individual p-values from pair-wise\n",
       "     comparisons. Such sensitive global p-value leads to the \"dual\n",
       "     signficance\" phenomenon. Dual-signficant means a gene set is\n",
       "     called significant simultaneously in both 1-direction tests (up-\n",
       "     and down-regulated). \"Dual signficance\" could be informative in\n",
       "     revealing the sub-types or sub-classes in big clinical or disease\n",
       "     studies, but may not be desirable in other cases.  Second, output\n",
       "     of gage function now includes the gene set test statistics from\n",
       "     pair-wise comparisons for all proper gene sets. The output is\n",
       "     always a named list now, with either 3 elements (\"greater\",\n",
       "     \"less\", \"stats\") for one-directional test or 2 elements\n",
       "     (\"greater\", \"stats\") for two-directional test.  Third, the\n",
       "     individual p-value (and test statistics)from dependent pair-wise\n",
       "     comparisions, i.e. comparisions between the same experiment vs\n",
       "     different controls, are now summarized into a single value. In\n",
       "     other words, the column number of individual p-values or\n",
       "     statistics is always the same as the sample number in the\n",
       "     experiment (or disease) group. This change made the argument value\n",
       "     compare=\"1ongroup\" and argument full.table less useful. It also\n",
       "     became easier to check the perturbations at gene-set level for\n",
       "     individual samples.  Fourth, whole gene-set level changes (either\n",
       "     p-values or statistics) can now be visualized using heatmaps due\n",
       "     to the third change above.  Correspondingly, functions\n",
       "     ‘sigGeneSet’ and ‘gagePipe’ have been revised to plot heatmaps for\n",
       "     whole gene sets.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     The result returned by gage function is a named list, with either\n",
       "     3 elements (\"greater\", \"less\", \"stats\") for one-directional test\n",
       "     (same.dir = TRUE) or 2 elements (\"greater\", \"stats\") for\n",
       "     two-directional test (same.dir = FALSE). Elements \"greater\" and\n",
       "     \"less\" are two data matrices of the same structure, mainly the\n",
       "     p-values, element \"stats\" contains the test statistics. Each data\n",
       "     matrix here has gene sets as rows sorted by global p- or q-values.\n",
       "     Test signficance or statistics columns include:\n",
       "\n",
       "p.geomean : geometric mean of the individual p-values from multiple\n",
       "          single array based gene set tests\n",
       "\n",
       "stat.mean : mean of the individual statistics from multiple single\n",
       "          array based gene set tests. Normally, its absoluate value\n",
       "          measures the magnitude of gene-set level changes, and its\n",
       "          sign indicates direction of the changes. When\n",
       "          saaTest=gs.KSTest, stat.mean is always positive.\n",
       "\n",
       "  p.val : gloal p-value or summary of the individual p-values from\n",
       "          multiple single array based gene set tests. This is the\n",
       "          default p-value being used.\n",
       "\n",
       "  q.val : FDR q-value adjustment of the global p-value using the\n",
       "          Benjamini & Hochberg procedure implemented in multtest\n",
       "          package. This is the default q-value being used.\n",
       "\n",
       "set.size : the effective gene set size, i.e. the number of genes\n",
       "          included in the gene set test\n",
       "\n",
       "other columns : columns of the individual p-values or statistics, each\n",
       "          measures the gene set perturbation in a single experiment (vs\n",
       "          its control or all controls, depends on the \"compare argument\n",
       "          value)\n",
       "     The result returned by ‘gagePrep’ is a data matrix derived from\n",
       "     ‘exprs’, but ready for column-wise gene est tests. In the matrix,\n",
       "     genes are rows, and columns are the per gene test statistics from\n",
       "     the ref-samp pairwise comparison.\n",
       "\n",
       "     The result returned by ‘gageSum’ is almost identical to the\n",
       "     results of ‘gage’ function, it is also a named list but has only 2\n",
       "     elements, \"p.glob\" and \"results\", with one round of test results.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Weijun Luo <luo_weijun@yahoo.com>\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     Luo, W., Friedman, M., Shedden K., Hankenson, K. and Woolf, P\n",
       "     GAGE: Generally Applicable Gene Set Enrichment for Pathways\n",
       "     Analysis. BMC Bioinformatics 2009, 10:161\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘gs.tTest’, ‘gs.zTest’, and ‘gs.KSTest’ functions used for gene\n",
       "     set test; ‘gagePipe’ and ‘heter.gage’ function used for multiple\n",
       "     GAGE analysis in a batch or combined GAGE analysis on\n",
       "     heterogeneous data\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     data(gse16873)\n",
       "     cn=colnames(gse16873)\n",
       "     hn=grep('HN',cn, ignore.case =TRUE)\n",
       "     dcis=grep('DCIS',cn, ignore.case =TRUE)\n",
       "     data(kegg.gs)\n",
       "     data(go.gs)\n",
       "     \n",
       "     #kegg test for 1-directional changes\n",
       "     gse16873.kegg.p <- gage(gse16873, gsets = kegg.gs, \n",
       "         ref = hn, samp = dcis)\n",
       "     #go.gs with the first 1000 entries as a fast example.\n",
       "     gse16873.go.p <- gage(gse16873, gsets = go.gs, \n",
       "         ref = hn, samp = dcis)\n",
       "     str(gse16873.kegg.p)\n",
       "     head(gse16873.kegg.p$greater)\n",
       "     head(gse16873.kegg.p$less)\n",
       "     head(gse16873.kegg.p$stats)\n",
       "     #kegg test for 2-directional changes\n",
       "     gse16873.kegg.2d.p <- gage(gse16873, gsets = kegg.gs, \n",
       "         ref = hn, samp = dcis, same.dir = FALSE)\n",
       "     head(gse16873.kegg.2d.p$greater)\n",
       "     head(gse16873.kegg.2d.p$stats)\n",
       "     \n",
       "     ###alternative ways to do GAGE analysis###\n",
       "     #with unpaired samples\n",
       "     gse16873.kegg.unpaired.p <- gage(gse16873, gsets = kegg.gs, \n",
       "         ref = hn, samp = dcis, compare = \"unpaired\")\n",
       "     \n",
       "     #other options to tweak includes:\n",
       "     #saaTest, use.fold, rank.test, etc. Check arguments section above for\n",
       "     #details and the vignette for more examples.\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?gage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
